<!DOCTYPE html>
<html lang="zh-CN,en,zh-TW,default">
    <head hexo-theme='https://github.com/volantis-x/hexo-theme-volantis/#6.0.0-alpha.1'>
  <meta name="generator" content="Hexo 6.3.0">
  <meta name="Volantis" content="6.0.0-alpha.1">
  <meta charset="utf-8">
  <!-- SEO相关 -->
  
  <link rel="canonical" href="https://snowman12137.github.io/2024/03/25/生信分析-通过共表达网络分析（wgcna）/"/>
  <!-- 渲染优化 -->
    <meta http-equiv='x-dns-prefetch-control' content='on' />
      <link rel='dns-prefetch' href='https://unpkg.com'>
      <link rel="preconnect" href="https://unpkg.com" crossorigin>
      <link rel='dns-prefetch' href='https://cdn.bootcdn.net'>
      <link rel="preconnect" href="https://cdn.bootcdn.net" crossorigin>
  <meta name="renderer" content="webkit">
  <meta name="force-rendering" content="webkit">
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1">
    <meta http-equiv="Content-Security-Policy" content=" default-src 'self' https:; block-all-mixed-content; base-uri 'self' https:; form-action 'self' https:; worker-src 'self' https:; connect-src 'self' https: *; img-src 'self' data: https: *; media-src 'self' https: *; font-src 'self' data: https: *; frame-src 'self' https: *; manifest-src 'self' https: *; child-src https:; script-src 'self' https: 'unsafe-inline' *; style-src 'self' https: 'unsafe-inline' *; ">
  <meta name="HandheldFriendly" content="True" >
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=5">
  <meta content="black-translucent" name="apple-mobile-web-app-status-bar-style">
  <meta content="telephone=no" name="format-detection">
  <!-- import head_begin begin -->
  <!-- import head_begin end -->
  <!-- Custom Files headBegin begin-->
  
  <!-- Custom Files headBegin end-->
  <link rel="preload" href="/css/style.css" as="style">
  <link rel="preload" href="https://unpkg.com/volantis-static@0.0.1660614606622/media/fonts/VarelaRound/VarelaRound-Regular.ttf" as="font" type="font/ttf" crossorigin="anonymous">
<link rel="preload" href="https://unpkg.com/volantis-static@0.0.1660614606622/media/fonts/UbuntuMono/UbuntuMono-Regular.ttf" as="font" type="font/ttf" crossorigin="anonymous">

  <!-- feed -->
  <!-- 页面元数据 -->
  <title>通过共表达网络分析（WGCNA） - 小吴の博客</title>
  <meta name="keywords" content="WGCAN网络,null">
  <meta desc name="description" content="我的博客 - Snowman - 小吴の博客">
  
<meta property="og:type" content="article">
<meta property="og:title" content="通过共表达网络分析（WGCNA）">
<meta property="og:url" content="https://snowman12137.github.io/2024/03/25/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90-%E9%80%9A%E8%BF%87%E5%85%B1%E8%A1%A8%E8%BE%BE%E7%BD%91%E7%BB%9C%E5%88%86%E6%9E%90%EF%BC%88WGCNA%EF%BC%89/index.html">
<meta property="og:site_name" content="小吴の博客">
<meta property="og:description" content="通过共表达网络分析（WGCNA）文章解析：谭永飞. 通过共表达网络分析 (WGCNA) 探究前列腺癌发生与进展相关功能基因模块. MS thesis. 吉林大学, 2018. 数据来源：TCGA（The Cancer Genome Atlas） (https:&#x2F;&#x2F;portal.gdc.cancer.gov&#x2F;)  一、数据下载与处理文章使用了TCGA-Assembler的R程序包下载TCGA规范化处">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://unpkg.com/volantis-static@0.0.1660614606622/media/org.volantis/blog/favicon/android-chrome-192x192.png">
<meta property="article:published_time" content="2024-03-24T16:00:00.000Z">
<meta property="article:modified_time" content="2024-03-25T14:51:16.314Z">
<meta property="article:author" content="Snowman">
<meta property="article:tag" content="WGCAN网络">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://unpkg.com/volantis-static@0.0.1660614606622/media/org.volantis/blog/favicon/android-chrome-192x192.png">
  <style>
    /* 首屏样式 */
    #safearea {
  display: none;
}
:root {
  --color-site-body: #f4f4f4;
  --color-site-bg: #f4f4f4;
  --color-site-inner: #fff;
  --color-site-footer: #666;
  --color-card: #fff;
  --color-text: #444;
  --color-block: #f6f6f6;
  --color-inlinecode: #c74f00;
  --color-codeblock: #fff7ea;
  --color-h1: #3a3a3a;
  --color-h2: #3a3a3a;
  --color-h3: #333;
  --color-h4: #444;
  --color-h5: #555;
  --color-h6: #666;
  --color-p: #444;
  --color-list: #666;
  --color-list-hl: #30ad91;
  --color-meta: #888;
  --color-read-bkg: #e0d8c8;
  --color-read-post: #f8f1e2;
  --color-copyright-bkg: #f5f5f5;
}
* {
  box-sizing: border-box;
  -webkit-box-sizing: border-box;
  -moz-box-sizing: border-box;
  outline: none;
  margin: 0;
  padding: 0;
}
*::-webkit-scrollbar {
  height: 4px;
  width: 4px;
}
*::-webkit-scrollbar-track-piece {
  background: transparent;
}
*::-webkit-scrollbar-thumb {
  background: #3dd9b6;
  cursor: pointer;
  border-radius: 2px;
  -webkit-border-radius: 2px;
}
*::-webkit-scrollbar-thumb:hover {
  background: #ff5722;
}
html {
  color: var(--color-text);
  width: 100%;
  height: 100%;
  font-family: UbuntuMono, "Varela Round", "PingFang SC", "Microsoft YaHei", Helvetica, Arial, Menlo, Monaco, monospace, sans-serif;
  font-size: 16px;
}
html >::-webkit-scrollbar {
  height: 4px;
  width: 4px;
}
html >::-webkit-scrollbar-track-piece {
  background: transparent;
}
html >::-webkit-scrollbar-thumb {
  background: #3dd9b6;
  cursor: pointer;
  border-radius: 2px;
  -webkit-border-radius: 2px;
}
html >::-webkit-scrollbar-thumb:hover {
  background: #ff5722;
}
body {
  background-color: var(--color-site-body);
  text-rendering: optimizelegibility;
  -webkit-tap-highlight-color: rgba(0,0,0,0);
  line-height: 1.6;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
}
body.modal-active {
  overflow: hidden;
}
@media screen and (max-width: 680px) {
  body.modal-active {
    position: fixed;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
  }
}
a {
  color: #2092ec;
  cursor: pointer;
  text-decoration: none;
  transition: all 0.28s ease;
  -webkit-transition: all 0.28s ease;
  -khtml-transition: all 0.28s ease;
  -moz-transition: all 0.28s ease;
  -o-transition: all 0.28s ease;
  -ms-transition: all 0.28s ease;
}
a:hover {
  color: #ff5722;
}
a:active,
a:hover {
  outline: 0;
}
ul,
ol {
  padding-left: 0;
}
ul li,
ol li {
  list-style: none;
}
header {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: block;
}
img {
  border: 0;
  background: none;
  max-width: 100%;
}
svg:not(:root) {
  overflow: hidden;
}
hr {
  -moz-box-sizing: content-box;
  box-sizing: content-box;
  -webkit-box-sizing: content-box;
  -moz-box-sizing: content-box;
  height: 0;
  border: 0;
  border-radius: 1px;
  -webkit-border-radius: 1px;
  border-bottom: 1px solid rgba(68,68,68,0.1);
}
button,
input {
  color: inherit;
  font: inherit;
  margin: 0;
}
button {
  overflow: visible;
  text-transform: none;
  -webkit-appearance: button;
  cursor: pointer;
}
@supports (backdrop-filter: blur(20px)) {
  .blur {
    background: rgba(255,255,255,0.9) !important;
    backdrop-filter: saturate(200%) blur(20px);
  }
}
.shadow {
  box-shadow: 0 1px 2px 0px rgba(0,0,0,0.1);
  -webkit-box-shadow: 0 1px 2px 0px rgba(0,0,0,0.1);
}
.shadow.floatable {
  transition: all 0.28s ease;
  -webkit-transition: all 0.28s ease;
  -khtml-transition: all 0.28s ease;
  -moz-transition: all 0.28s ease;
  -o-transition: all 0.28s ease;
  -ms-transition: all 0.28s ease;
}
.shadow.floatable:hover {
  box-shadow: 0 2px 4px 0px rgba(0,0,0,0.1), 0 4px 8px 0px rgba(0,0,0,0.1), 0 8px 16px 0px rgba(0,0,0,0.1);
  -webkit-box-shadow: 0 2px 4px 0px rgba(0,0,0,0.1), 0 4px 8px 0px rgba(0,0,0,0.1), 0 8px 16px 0px rgba(0,0,0,0.1);
}
#l_cover {
  min-height: 64px;
}
.cover-wrapper {
  top: 0;
  left: 0;
  max-width: 100%;
  height: 100vh;
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: -ms-flexbox /* TWEENER - IE 10 */;
  display: -webkit-flex /* NEW - Chrome */;
  display: flex /* NEW, Spec - Opera 12.1, Firefox 20+ */;
  display: flex;
  flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -khtml-flex-wrap: nowrap;
  -moz-flex-wrap: nowrap;
  -o-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  -webkit-box-direction: normal;
  -moz-box-direction: normal;
  -webkit-box-orient: vertical;
  -moz-box-orient: vertical;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  align-items: center;
  align-self: center;
  align-content: center;
  color: var(--color-site-inner);
  padding: 0 16px;
  user-select: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  position: relative;
  overflow: hidden;
  margin-bottom: -100px;
}
.cover-wrapper .cover-bg {
  position: absolute;
  width: 100%;
  height: 100%;
  background-position: center;
  background-size: cover;
  -webkit-background-size: cover;
  -moz-background-size: cover;
}
.cover-wrapper .cover-bg.lazyload:not(.loaded) {
  opacity: 0;
  -webkit-opacity: 0;
  -moz-opacity: 0;
}
.cover-wrapper .cover-bg.lazyload.loaded {
  animation-delay: 0s;
  animation-duration: 0.5s;
  animation-fill-mode: forwards;
  animation-timing-function: ease-out;
  animation-name: fadeIn;
}
@-moz-keyframes fadeIn {
  0% {
    opacity: 0;
    -webkit-opacity: 0;
    -moz-opacity: 0;
    filter: blur(12px);
    transform: scale(1.02);
    -webkit-transform: scale(1.02);
    -khtml-transform: scale(1.02);
    -moz-transform: scale(1.02);
    -o-transform: scale(1.02);
    -ms-transform: scale(1.02);
  }
  100% {
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
}
@-webkit-keyframes fadeIn {
  0% {
    opacity: 0;
    -webkit-opacity: 0;
    -moz-opacity: 0;
    filter: blur(12px);
    transform: scale(1.02);
    -webkit-transform: scale(1.02);
    -khtml-transform: scale(1.02);
    -moz-transform: scale(1.02);
    -o-transform: scale(1.02);
    -ms-transform: scale(1.02);
  }
  100% {
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
}
@-o-keyframes fadeIn {
  0% {
    opacity: 0;
    -webkit-opacity: 0;
    -moz-opacity: 0;
    filter: blur(12px);
    transform: scale(1.02);
    -webkit-transform: scale(1.02);
    -khtml-transform: scale(1.02);
    -moz-transform: scale(1.02);
    -o-transform: scale(1.02);
    -ms-transform: scale(1.02);
  }
  100% {
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
}
@keyframes fadeIn {
  0% {
    opacity: 0;
    -webkit-opacity: 0;
    -moz-opacity: 0;
    filter: blur(12px);
    transform: scale(1.02);
    -webkit-transform: scale(1.02);
    -khtml-transform: scale(1.02);
    -moz-transform: scale(1.02);
    -o-transform: scale(1.02);
    -ms-transform: scale(1.02);
  }
  100% {
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
}
.cover-wrapper .cover-body {
  z-index: 1;
  position: relative;
  width: 100%;
  height: 100%;
}
.cover-wrapper#full {
  height: calc(100vh + 100px);
  padding-bottom: 100px;
}
.cover-wrapper#half {
  max-height: 640px;
  min-height: 400px;
  height: calc(36vh - 64px + 200px);
}
.cover-wrapper #scroll-down {
  width: 100%;
  height: 64px;
  position: absolute;
  bottom: 100px;
  text-align: center;
  cursor: pointer;
}
.cover-wrapper #scroll-down .scroll-down-effects {
  color: #fff;
  font-size: 24px;
  line-height: 64px;
  position: absolute;
  width: 24px;
  left: calc(50% - 12px);
  text-shadow: 0 1px 2px rgba(0,0,0,0.1);
  animation: scroll-down-effect 1.5s infinite;
  -webkit-animation: scroll-down-effect 1.5s infinite;
  -khtml-animation: scroll-down-effect 1.5s infinite;
  -moz-animation: scroll-down-effect 1.5s infinite;
  -o-animation: scroll-down-effect 1.5s infinite;
  -ms-animation: scroll-down-effect 1.5s infinite;
}
@-moz-keyframes scroll-down-effect {
  0% {
    top: 0;
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
  50% {
    top: -16px;
    opacity: 0.4;
    -webkit-opacity: 0.4;
    -moz-opacity: 0.4;
  }
  100% {
    top: 0;
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
}
@-webkit-keyframes scroll-down-effect {
  0% {
    top: 0;
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
  50% {
    top: -16px;
    opacity: 0.4;
    -webkit-opacity: 0.4;
    -moz-opacity: 0.4;
  }
  100% {
    top: 0;
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
}
@-o-keyframes scroll-down-effect {
  0% {
    top: 0;
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
  50% {
    top: -16px;
    opacity: 0.4;
    -webkit-opacity: 0.4;
    -moz-opacity: 0.4;
  }
  100% {
    top: 0;
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
}
@keyframes scroll-down-effect {
  0% {
    top: 0;
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
  50% {
    top: -16px;
    opacity: 0.4;
    -webkit-opacity: 0.4;
    -moz-opacity: 0.4;
  }
  100% {
    top: 0;
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
}
.cover-wrapper .cover-body {
  margin-top: 64px;
  margin-bottom: 100px;
}
.cover-wrapper .cover-body,
.cover-wrapper .cover-body .top,
.cover-wrapper .cover-body .bottom {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: -ms-flexbox /* TWEENER - IE 10 */;
  display: -webkit-flex /* NEW - Chrome */;
  display: flex /* NEW, Spec - Opera 12.1, Firefox 20+ */;
  display: flex;
  -webkit-box-direction: normal;
  -moz-box-direction: normal;
  -webkit-box-orient: vertical;
  -moz-box-orient: vertical;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  -webkit-justify-content: center;
  -khtml-justify-content: center;
  -moz-justify-content: center;
  -o-justify-content: center;
  -ms-justify-content: center;
  max-width: 100%;
}
.cover-wrapper .cover-body .bottom {
  margin-top: 32px;
}
.cover-wrapper .cover-body .title {
  font-family: "Varela Round", "PingFang SC", "Microsoft YaHei", Helvetica, Arial, Helvetica, monospace;
  font-size: 3.125rem;
  line-height: 1.2;
  text-shadow: 0 1px 2px rgba(0,0,0,0.1);
}
.cover-wrapper .cover-body .subtitle {
  font-size: 20px;
}
.cover-wrapper .cover-body .logo {
  max-height: 120px;
  max-width: calc(100% - 4 * 16px);
}
@media screen and (min-height: 1024px) {
  .cover-wrapper .cover-body .title {
    font-size: 3rem;
  }
  .cover-wrapper .cover-body .subtitle {
    font-size: 1.05rem;
  }
  .cover-wrapper .cover-body .logo {
    max-height: 150px;
  }
}
.cover-wrapper .cover-body .m_search {
  position: relative;
  max-width: calc(100% - 16px);
  width: 320px;
  vertical-align: middle;
}
.cover-wrapper .cover-body .m_search .form {
  position: relative;
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: block;
  width: 100%;
}
.cover-wrapper .cover-body .m_search .icon,
.cover-wrapper .cover-body .m_search .input {
  transition: all 0.28s ease;
  -webkit-transition: all 0.28s ease;
  -khtml-transition: all 0.28s ease;
  -moz-transition: all 0.28s ease;
  -o-transition: all 0.28s ease;
  -ms-transition: all 0.28s ease;
}
.cover-wrapper .cover-body .m_search .icon {
  position: absolute;
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: block;
  line-height: 2.5rem;
  width: 32px;
  top: 0;
  left: 5px;
  color: rgba(68,68,68,0.75);
}
.cover-wrapper .cover-body .m_search .input {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: block;
  height: 2.5rem;
  width: 100%;
  box-shadow: none;
  -webkit-box-shadow: none;
  box-sizing: border-box;
  -webkit-box-sizing: border-box;
  -moz-box-sizing: border-box;
  font-size: 0.875rem;
  -webkit-appearance: none;
  padding-left: 36px;
  border-radius: 1.4rem;
  -webkit-border-radius: 1.4rem;
  background: rgba(255,255,255,0.6);
  backdrop-filter: blur(10px);
  border: none;
  color: var(--color-text);
}
@media screen and (max-width: 500px) {
  .cover-wrapper .cover-body .m_search .input {
    padding-left: 36px;
  }
}
.cover-wrapper .cover-body .m_search .input:hover {
  background: rgba(255,255,255,0.8);
}
.cover-wrapper .cover-body .m_search .input:focus {
  background: #fff;
}
.cover-wrapper .list-h {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: -ms-flexbox /* TWEENER - IE 10 */;
  display: -webkit-flex /* NEW - Chrome */;
  display: flex /* NEW, Spec - Opera 12.1, Firefox 20+ */;
  display: flex;
  -webkit-box-direction: normal;
  -moz-box-direction: normal;
  -webkit-box-orient: horizontal;
  -moz-box-orient: horizontal;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -khtml-flex-wrap: wrap;
  -moz-flex-wrap: wrap;
  -o-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  align-items: stretch;
  border-radius: 4px;
  -webkit-border-radius: 4px;
  user-select: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
}
.cover-wrapper .list-h a {
  -webkit-box-flex: 1;
  -moz-box-flex: 1;
  -webkit-flex: 1 0;
  -ms-flex: 1 0;
  flex: 1 0;
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: -ms-flexbox /* TWEENER - IE 10 */;
  display: -webkit-flex /* NEW - Chrome */;
  display: flex /* NEW, Spec - Opera 12.1, Firefox 20+ */;
  display: flex;
  font-weight: 600;
}
.cover-wrapper .list-h a img {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: block;
  border-radius: 2px;
  -webkit-border-radius: 2px;
  margin: 4px;
  min-width: 40px;
  max-width: 44px;
}
@media screen and (max-width: 768px) {
  .cover-wrapper .list-h a img {
    min-width: 36px;
    max-width: 40px;
  }
}
@media screen and (max-width: 500px) {
  .cover-wrapper .list-h a img {
    margin: 2px 4px;
    min-width: 32px;
    max-width: 36px;
  }
}
@media screen and (max-width: 375px) {
  .cover-wrapper .list-h a img {
    min-width: 28px;
    max-width: 32px;
  }
}
.cover-wrapper {
  max-width: 100%;
}
.cover-wrapper.search .bottom .menu {
  margin-top: 16px;
}
.cover-wrapper.search .bottom .menu .list-h a {
  white-space: nowrap;
  -webkit-box-direction: normal;
  -moz-box-direction: normal;
  -webkit-box-orient: horizontal;
  -moz-box-orient: horizontal;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  align-items: baseline;
  padding: 2px;
  margin: 4px;
  color: var(--color-site-inner);
  opacity: 0.75;
  -webkit-opacity: 0.75;
  -moz-opacity: 0.75;
  text-shadow: 0 1px 2px rgba(0,0,0,0.05);
  border-bottom: 2px solid transparent;
}
.cover-wrapper.search .bottom .menu .list-h a i {
  margin-right: 4px;
}
.cover-wrapper.search .bottom .menu .list-h a p {
  font-size: 0.9375rem;
}
.cover-wrapper.search .bottom .menu .list-h a:hover,
.cover-wrapper.search .bottom .menu .list-h a.active,
.cover-wrapper.search .bottom .menu .list-h a:active {
  opacity: 1;
  -webkit-opacity: 1;
  -moz-opacity: 1;
  border-bottom: 2px solid var(--color-site-inner);
}
.cover-wrapper.dock .menu,
.cover-wrapper.featured .menu,
.cover-wrapper.focus .menu {
  border-radius: 6px;
  -webkit-border-radius: 6px;
}
.cover-wrapper.dock .menu .list-h a,
.cover-wrapper.featured .menu .list-h a,
.cover-wrapper.focus .menu .list-h a {
  -webkit-box-direction: normal;
  -moz-box-direction: normal;
  -webkit-box-orient: vertical;
  -moz-box-orient: vertical;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  align-items: center;
  padding: 12px;
  line-height: 24px;
  border-radius: 4px;
  -webkit-border-radius: 4px;
  border-bottom: none;
  text-align: center;
  align-content: flex-end;
  color: rgba(68,68,68,0.7);
  font-size: 1.5rem;
}
@media screen and (max-width: 500px) {
  .cover-wrapper.dock .menu .list-h a,
  .cover-wrapper.featured .menu .list-h a,
  .cover-wrapper.focus .menu .list-h a {
    padding: 12px 8px;
  }
}
.cover-wrapper.dock .menu .list-h a i,
.cover-wrapper.featured .menu .list-h a i,
.cover-wrapper.focus .menu .list-h a i {
  margin: 8px;
}
.cover-wrapper.dock .menu .list-h a p,
.cover-wrapper.featured .menu .list-h a p,
.cover-wrapper.focus .menu .list-h a p {
  font-size: 0.875rem;
}
.cover-wrapper.dock .menu .list-h a.active,
.cover-wrapper.featured .menu .list-h a.active,
.cover-wrapper.focus .menu .list-h a.active {
  background: var(--color-card);
  backdrop-filter: none;
}
.cover-wrapper.dock .menu .list-h a.active i,
.cover-wrapper.featured .menu .list-h a.active i,
.cover-wrapper.focus .menu .list-h a.active i,
.cover-wrapper.dock .menu .list-h a.active i+p,
.cover-wrapper.featured .menu .list-h a.active i+p,
.cover-wrapper.focus .menu .list-h a.active i+p {
  color: #3dd9b6;
}
.cover-wrapper.dock .menu .list-h a.active img+p,
.cover-wrapper.featured .menu .list-h a.active img+p,
.cover-wrapper.focus .menu .list-h a.active img+p {
  color: var(--color-text);
}
.cover-wrapper.dock .menu .list-h a:hover,
.cover-wrapper.featured .menu .list-h a:hover,
.cover-wrapper.focus .menu .list-h a:hover {
  background: var(--color-card);
}
.cover-wrapper.dock .top {
  margin-bottom: 48px;
}
.cover-wrapper.dock .menu {
  background: rgba(255,255,255,0.5);
  position: absolute;
  bottom: 0;
  max-width: 100%;
}
.cover-wrapper.dock .menu .list-h {
  flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -khtml-flex-wrap: nowrap;
  -moz-flex-wrap: nowrap;
  -o-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  margin: 4px;
}
.cover-wrapper.dock .menu .list-h a+a {
  margin-left: 4px;
}
@media screen and (max-width: 500px) {
  .cover-wrapper.dock .menu .list-h {
    overflow-x: scroll;
  }
  .cover-wrapper.dock .menu .list-h::-webkit-scrollbar {
    height: 0;
    width: 0;
  }
  .cover-wrapper.dock .menu .list-h::-webkit-scrollbar-track-piece {
    background: transparent;
  }
  .cover-wrapper.dock .menu .list-h::-webkit-scrollbar-thumb {
    background: #3dd9b6;
    cursor: pointer;
    border-radius: 0;
    -webkit-border-radius: 0;
  }
  .cover-wrapper.dock .menu .list-h::-webkit-scrollbar-thumb:hover {
    background: #ff5722;
  }
}
@supports (backdrop-filter: blur(20px)) {
  .cover-wrapper.dock .menu {
    background: rgba(255,255,255,0.5);
    backdrop-filter: saturate(200%) blur(20px);
  }
}
.cover-wrapper #parallax-window {
  position: absolute;
  width: 100%;
  height: 100%;
  background: transparent;
}
.parallax-mirror {
  animation-delay: 0s;
  animation-duration: 0.5s;
  animation-fill-mode: forwards;
  animation-timing-function: ease-out;
  animation-name: fadeIn;
}
@-moz-keyframes fadeIn {
  0% {
    opacity: 0;
    -webkit-opacity: 0;
    -moz-opacity: 0;
    filter: blur(12px);
  }
  100% {
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
}
@-webkit-keyframes fadeIn {
  0% {
    opacity: 0;
    -webkit-opacity: 0;
    -moz-opacity: 0;
    filter: blur(12px);
  }
  100% {
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
}
@-o-keyframes fadeIn {
  0% {
    opacity: 0;
    -webkit-opacity: 0;
    -moz-opacity: 0;
    filter: blur(12px);
  }
  100% {
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
}
@keyframes fadeIn {
  0% {
    opacity: 0;
    -webkit-opacity: 0;
    -moz-opacity: 0;
    filter: blur(12px);
  }
  100% {
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
}
@font-face {
  font-family: 'UbuntuMono';
  src: url("https://unpkg.com/volantis-static@0.0.1660614606622/media/fonts/UbuntuMono/UbuntuMono-Regular.ttf");
  font-weight: 'normal';
  font-style: 'normal';
  font-display: swap;
}
@font-face {
  font-family: 'Varela Round';
  src: url("https://unpkg.com/volantis-static@0.0.1660614606622/media/fonts/VarelaRound/VarelaRound-Regular.ttf");
  font-weight: 'normal';
  font-style: 'normal';
  font-display: swap;
}
.l_header {
  position: fixed;
  z-index: 1000;
  top: 0;
  width: 100%;
  height: 64px;
  background: var(--color-card);
  box-shadow: 0 1px 2px 0px rgba(0,0,0,0.1);
  -webkit-box-shadow: 0 1px 2px 0px rgba(0,0,0,0.1);
}
.l_header.auto {
  transition: opacity 0.4s ease;
  -webkit-transition: opacity 0.4s ease;
  -khtml-transition: opacity 0.4s ease;
  -moz-transition: opacity 0.4s ease;
  -o-transition: opacity 0.4s ease;
  -ms-transition: opacity 0.4s ease;
  visibility: hidden;
}
.l_header.auto.show {
  opacity: 1 !important;
  -webkit-opacity: 1 !important;
  -moz-opacity: 1 !important;
  visibility: visible;
}
.l_header .container {
  margin-left: 16px;
  margin-right: 16px;
}
.l_header #wrapper {
  height: 100%;
  user-select: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
}
.l_header #wrapper .nav-main,
.l_header #wrapper .nav-sub {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: -ms-flexbox /* TWEENER - IE 10 */;
  display: -webkit-flex /* NEW - Chrome */;
  display: flex /* NEW, Spec - Opera 12.1, Firefox 20+ */;
  display: flex;
  flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -khtml-flex-wrap: nowrap;
  -moz-flex-wrap: nowrap;
  -o-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  justify-content: space-between;
  -webkit-justify-content: space-between;
  -khtml-justify-content: space-between;
  -moz-justify-content: space-between;
  -o-justify-content: space-between;
  -ms-justify-content: space-between;
  align-items: center;
}
.l_header #wrapper .nav-main {
  transition: all 0.28s ease;
  -webkit-transition: all 0.28s ease;
  -khtml-transition: all 0.28s ease;
  -moz-transition: all 0.28s ease;
  -o-transition: all 0.28s ease;
  -ms-transition: all 0.28s ease;
}
.l_header #wrapper.sub .nav-main {
  transform: translateY(-64px);
  -webkit-transform: translateY(-64px);
  -khtml-transform: translateY(-64px);
  -moz-transform: translateY(-64px);
  -o-transform: translateY(-64px);
  -ms-transform: translateY(-64px);
}
.l_header #wrapper .nav-sub {
  transition: all 0.28s ease;
  -webkit-transition: all 0.28s ease;
  -khtml-transition: all 0.28s ease;
  -moz-transition: all 0.28s ease;
  -o-transition: all 0.28s ease;
  -ms-transition: all 0.28s ease;
  opacity: 0;
  -webkit-opacity: 0;
  -moz-opacity: 0;
  height: 64px;
  width: calc(100% - 2 * 16px);
  position: absolute;
}
.l_header #wrapper .nav-sub ::-webkit-scrollbar {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: none;
}
@media screen and (min-width: 2048px) {
  .l_header #wrapper .nav-sub {
    max-width: 55vw;
    margin: auto;
  }
}
.l_header #wrapper.sub .nav-sub {
  opacity: 1;
  -webkit-opacity: 1;
  -moz-opacity: 1;
}
.l_header #wrapper .title {
  position: relative;
  color: var(--color-text);
  padding-left: 24px;
  max-height: 64px;
}
.l_header #wrapper .nav-main .title {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  flex-shrink: 0;
  line-height: 64px;
  padding: 0 24px;
  font-size: 1.25rem;
  font-family: "Varela Round", "PingFang SC", "Microsoft YaHei", Helvetica, Arial, Helvetica, monospace;
}
.l_header #wrapper .nav-main .title img {
  height: 64px;
}
.l_header .nav-sub {
  max-width: 1080px;
  margin: auto;
}
.l_header .nav-sub .title {
  font-weight: bold;
  font-family: UbuntuMono, "Varela Round", "PingFang SC", "Microsoft YaHei", Helvetica, Arial, Menlo, Monaco, monospace, sans-serif;
  line-height: 1.2;
  max-height: 64px;
  white-space: normal;
  flex-shrink: 1;
}
.l_header .switcher {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: none;
  line-height: 64px;
  align-items: center;
}
.l_header .switcher .s-toc {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: none;
}
@media screen and (max-width: 768px) {
  .l_header .switcher .s-toc {
    display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
    display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
    display: -ms-flexbox /* TWEENER - IE 10 */;
    display: -webkit-flex /* NEW - Chrome */;
    display: flex /* NEW, Spec - Opera 12.1, Firefox 20+ */;
    display: flex;
  }
}
.l_header .switcher >li {
  height: 48px;
  transition: all 0.28s ease;
  -webkit-transition: all 0.28s ease;
  -khtml-transition: all 0.28s ease;
  -moz-transition: all 0.28s ease;
  -o-transition: all 0.28s ease;
  -ms-transition: all 0.28s ease;
  margin: 2px;
}
@media screen and (max-width: 500px) {
  .l_header .switcher >li {
    margin: 0 1px;
    height: 48px;
  }
}
.l_header .switcher >li >a {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: -ms-flexbox /* TWEENER - IE 10 */;
  display: -webkit-flex /* NEW - Chrome */;
  display: flex /* NEW, Spec - Opera 12.1, Firefox 20+ */;
  display: flex;
  justify-content: center;
  -webkit-justify-content: center;
  -khtml-justify-content: center;
  -moz-justify-content: center;
  -o-justify-content: center;
  -ms-justify-content: center;
  align-items: center;
  width: 48px;
  height: 48px;
  padding: 0.85em 1.1em;
  border-radius: 100px;
  -webkit-border-radius: 100px;
  border: none;
  transition: all 0.28s ease;
  -webkit-transition: all 0.28s ease;
  -khtml-transition: all 0.28s ease;
  -moz-transition: all 0.28s ease;
  -o-transition: all 0.28s ease;
  -ms-transition: all 0.28s ease;
  color: #3dd9b6;
}
.l_header .switcher >li >a:hover {
  border: none;
}
.l_header .switcher >li >a.active,
.l_header .switcher >li >a:active {
  border: none;
  background: var(--color-site-bg);
}
@media screen and (max-width: 500px) {
  .l_header .switcher >li >a {
    width: 36px;
    height: 48px;
  }
}
.l_header .nav-sub .switcher {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: -ms-flexbox /* TWEENER - IE 10 */;
  display: -webkit-flex /* NEW - Chrome */;
  display: flex /* NEW, Spec - Opera 12.1, Firefox 20+ */;
  display: flex;
}
.l_header .m_search {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: -ms-flexbox /* TWEENER - IE 10 */;
  display: -webkit-flex /* NEW - Chrome */;
  display: flex /* NEW, Spec - Opera 12.1, Firefox 20+ */;
  display: flex;
  height: 64px;
  width: 240px;
  transition: all 0.28s ease;
  -webkit-transition: all 0.28s ease;
  -khtml-transition: all 0.28s ease;
  -moz-transition: all 0.28s ease;
  -o-transition: all 0.28s ease;
  -ms-transition: all 0.28s ease;
}
@media screen and (max-width: 1024px) {
  .l_header .m_search {
    width: 44px;
    min-width: 44px;
  }
  .l_header .m_search input::placeholder {
    opacity: 0;
    -webkit-opacity: 0;
    -moz-opacity: 0;
  }
  .l_header .m_search:hover {
    width: 240px;
  }
  .l_header .m_search:hover input::placeholder {
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
}
@media screen and (min-width: 500px) {
  .l_header .m_search:hover .input {
    width: 100%;
  }
  .l_header .m_search:hover .input::placeholder {
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
}
@media screen and (max-width: 500px) {
  .l_header .m_search {
    min-width: 0;
  }
  .l_header .m_search input::placeholder {
    opacity: 1;
    -webkit-opacity: 1;
    -moz-opacity: 1;
  }
}
.l_header .m_search .form {
  position: relative;
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: -ms-flexbox /* TWEENER - IE 10 */;
  display: -webkit-flex /* NEW - Chrome */;
  display: flex /* NEW, Spec - Opera 12.1, Firefox 20+ */;
  display: flex;
  width: 100%;
  align-items: center;
}
.l_header .m_search .icon {
  position: absolute;
  width: 36px;
  left: 5px;
  color: var(--color-meta);
}
@media screen and (max-width: 500px) {
  .l_header .m_search .icon {
    display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
    display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
    display: none;
  }
}
.l_header .m_search .input {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: block;
  padding-top: 8px;
  padding-bottom: 8px;
  line-height: 1.3;
  width: 100%;
  color: var(--color-text);
  background: #fafafa;
  box-shadow: none;
  -webkit-box-shadow: none;
  box-sizing: border-box;
  -webkit-box-sizing: border-box;
  -moz-box-sizing: border-box;
  padding-left: 40px;
  font-size: 0.875rem;
  border-radius: 8px;
  -webkit-border-radius: 8px;
  border: none;
  transition: all 0.28s ease;
  -webkit-transition: all 0.28s ease;
  -khtml-transition: all 0.28s ease;
  -moz-transition: all 0.28s ease;
  -o-transition: all 0.28s ease;
  -ms-transition: all 0.28s ease;
}
@media screen and (min-width: 500px) {
  .l_header .m_search .input:focus {
    box-shadow: 0 4px 8px 0px rgba(0,0,0,0.1);
    -webkit-box-shadow: 0 4px 8px 0px rgba(0,0,0,0.1);
  }
}
@media screen and (max-width: 500px) {
  .l_header .m_search .input {
    background: var(--color-block);
    padding-left: 8px;
    border: none;
  }
  .l_header .m_search .input:hover,
  .l_header .m_search .input:focus {
    border: none;
  }
}
@media (max-width: 500px) {
  .l_header .m_search {
    left: 0;
    width: 0;
    overflow: hidden;
    position: absolute;
    background: #fff;
    transition: all 0.28s ease;
    -webkit-transition: all 0.28s ease;
    -khtml-transition: all 0.28s ease;
    -moz-transition: all 0.28s ease;
    -o-transition: all 0.28s ease;
    -ms-transition: all 0.28s ease;
  }
  .l_header .m_search .input {
    border-radius: 32px;
    -webkit-border-radius: 32px;
    margin-left: 16px;
    padding-left: 16px;
  }
  .l_header.z_search-open .m_search {
    width: 100%;
  }
  .l_header.z_search-open .m_search .input {
    width: calc(100% - 120px);
  }
}
ul.m-pc >li>a {
  color: inherit;
  border-bottom: 2px solid transparent;
}
ul.m-pc >li>a:active,
ul.m-pc >li>a.active {
  border-bottom: 2px solid #3dd9b6;
}
ul.m-pc li:hover >ul.list-v,
ul.list-v li:hover >ul.list-v {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: block;
}
ul.nav-list-h {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: -ms-flexbox /* TWEENER - IE 10 */;
  display: -webkit-flex /* NEW - Chrome */;
  display: flex /* NEW, Spec - Opera 12.1, Firefox 20+ */;
  display: flex;
  align-items: stretch;
}
ul.nav-list-h>li {
  position: relative;
  justify-content: center;
  -webkit-justify-content: center;
  -khtml-justify-content: center;
  -moz-justify-content: center;
  -o-justify-content: center;
  -ms-justify-content: center;
  height: 100%;
  line-height: 2.4;
  border-radius: 4px;
  -webkit-border-radius: 4px;
}
ul.nav-list-h>li >a {
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 600;
}
ul.list-v {
  z-index: 1;
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: none;
  position: absolute;
  background: var(--color-card);
  box-shadow: 0 2px 4px 0px rgba(0,0,0,0.08), 0 4px 8px 0px rgba(0,0,0,0.08), 0 8px 16px 0px rgba(0,0,0,0.08);
  -webkit-box-shadow: 0 2px 4px 0px rgba(0,0,0,0.08), 0 4px 8px 0px rgba(0,0,0,0.08), 0 8px 16px 0px rgba(0,0,0,0.08);
  margin-top: -6px;
  border-radius: 4px;
  -webkit-border-radius: 4px;
  padding: 8px 0;
}
ul.list-v.show {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: block;
}
ul.list-v hr {
  margin-top: 8px;
  margin-bottom: 8px;
}
ul.list-v >li {
  white-space: nowrap;
  word-break: keep-all;
}
ul.list-v >li.header {
  font-size: 0.78125rem;
  font-weight: bold;
  line-height: 2em;
  color: var(--color-meta);
  margin: 8px 16px 4px;
}
ul.list-v >li.header i {
  margin-right: 8px;
}
ul.list-v >li ul {
  margin-left: 0;
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: none;
  margin-top: -40px;
}
ul.list-v .aplayer-container {
  min-height: 64px;
  padding: 6px 16px;
}
ul.list-v >li>a {
  transition: all 0.28s ease;
  -webkit-transition: all 0.28s ease;
  -khtml-transition: all 0.28s ease;
  -moz-transition: all 0.28s ease;
  -o-transition: all 0.28s ease;
  -ms-transition: all 0.28s ease;
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: block;
  color: var(--color-list);
  font-size: 0.875rem;
  font-weight: bold;
  line-height: 36px;
  padding: 0 20px 0 16px;
  text-overflow: ellipsis;
  margin: 0 4px;
  border-radius: 4px;
  -webkit-border-radius: 4px;
}
@media screen and (max-width: 1024px) {
  ul.list-v >li>a {
    line-height: 40px;
  }
}
ul.list-v >li>a >i {
  margin-right: 8px;
}
ul.list-v >li>a:active,
ul.list-v >li>a.active {
  color: var(--color-list-hl);
}
ul.list-v >li>a:hover {
  color: var(--color-list-hl);
  background: var(--color-site-bg);
}
.l_header .menu >ul>li>a {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: block;
  padding: 0 8px;
}
.l_header .menu >ul>li>a >i {
  margin-right: 4px;
}
.l_header ul.nav-list-h>li {
  color: var(--color-list);
  line-height: 64px;
}
.l_header ul.nav-list-h>li >a {
  max-height: 64px;
  overflow: hidden;
  color: inherit;
}
.l_header ul.nav-list-h>li >a:active,
.l_header ul.nav-list-h>li >a.active {
  color: #3dd9b6;
}
.l_header ul.nav-list-h>li:hover>a {
  color: var(--color-list-hl);
}
.l_header ul.nav-list-h>li i.music {
  animation: rotate-effect 1.5s linear infinite;
  -webkit-animation: rotate-effect 1.5s linear infinite;
  -khtml-animation: rotate-effect 1.5s linear infinite;
  -moz-animation: rotate-effect 1.5s linear infinite;
  -o-animation: rotate-effect 1.5s linear infinite;
  -ms-animation: rotate-effect 1.5s linear infinite;
}
@-moz-keyframes rotate-effect {
  0% {
    transform: rotate(0);
    -webkit-transform: rotate(0);
    -khtml-transform: rotate(0);
    -moz-transform: rotate(0);
    -o-transform: rotate(0);
    -ms-transform: rotate(0);
  }
  25% {
    transform: rotate(90deg);
    -webkit-transform: rotate(90deg);
    -khtml-transform: rotate(90deg);
    -moz-transform: rotate(90deg);
    -o-transform: rotate(90deg);
    -ms-transform: rotate(90deg);
  }
  50% {
    transform: rotate(180deg);
    -webkit-transform: rotate(180deg);
    -khtml-transform: rotate(180deg);
    -moz-transform: rotate(180deg);
    -o-transform: rotate(180deg);
    -ms-transform: rotate(180deg);
  }
  75% {
    transform: rotate(270deg);
    -webkit-transform: rotate(270deg);
    -khtml-transform: rotate(270deg);
    -moz-transform: rotate(270deg);
    -o-transform: rotate(270deg);
    -ms-transform: rotate(270deg);
  }
  100% {
    transform: rotate(360deg);
    -webkit-transform: rotate(360deg);
    -khtml-transform: rotate(360deg);
    -moz-transform: rotate(360deg);
    -o-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
  }
}
@-webkit-keyframes rotate-effect {
  0% {
    transform: rotate(0);
    -webkit-transform: rotate(0);
    -khtml-transform: rotate(0);
    -moz-transform: rotate(0);
    -o-transform: rotate(0);
    -ms-transform: rotate(0);
  }
  25% {
    transform: rotate(90deg);
    -webkit-transform: rotate(90deg);
    -khtml-transform: rotate(90deg);
    -moz-transform: rotate(90deg);
    -o-transform: rotate(90deg);
    -ms-transform: rotate(90deg);
  }
  50% {
    transform: rotate(180deg);
    -webkit-transform: rotate(180deg);
    -khtml-transform: rotate(180deg);
    -moz-transform: rotate(180deg);
    -o-transform: rotate(180deg);
    -ms-transform: rotate(180deg);
  }
  75% {
    transform: rotate(270deg);
    -webkit-transform: rotate(270deg);
    -khtml-transform: rotate(270deg);
    -moz-transform: rotate(270deg);
    -o-transform: rotate(270deg);
    -ms-transform: rotate(270deg);
  }
  100% {
    transform: rotate(360deg);
    -webkit-transform: rotate(360deg);
    -khtml-transform: rotate(360deg);
    -moz-transform: rotate(360deg);
    -o-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
  }
}
@-o-keyframes rotate-effect {
  0% {
    transform: rotate(0);
    -webkit-transform: rotate(0);
    -khtml-transform: rotate(0);
    -moz-transform: rotate(0);
    -o-transform: rotate(0);
    -ms-transform: rotate(0);
  }
  25% {
    transform: rotate(90deg);
    -webkit-transform: rotate(90deg);
    -khtml-transform: rotate(90deg);
    -moz-transform: rotate(90deg);
    -o-transform: rotate(90deg);
    -ms-transform: rotate(90deg);
  }
  50% {
    transform: rotate(180deg);
    -webkit-transform: rotate(180deg);
    -khtml-transform: rotate(180deg);
    -moz-transform: rotate(180deg);
    -o-transform: rotate(180deg);
    -ms-transform: rotate(180deg);
  }
  75% {
    transform: rotate(270deg);
    -webkit-transform: rotate(270deg);
    -khtml-transform: rotate(270deg);
    -moz-transform: rotate(270deg);
    -o-transform: rotate(270deg);
    -ms-transform: rotate(270deg);
  }
  100% {
    transform: rotate(360deg);
    -webkit-transform: rotate(360deg);
    -khtml-transform: rotate(360deg);
    -moz-transform: rotate(360deg);
    -o-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
  }
}
@keyframes rotate-effect {
  0% {
    transform: rotate(0);
    -webkit-transform: rotate(0);
    -khtml-transform: rotate(0);
    -moz-transform: rotate(0);
    -o-transform: rotate(0);
    -ms-transform: rotate(0);
  }
  25% {
    transform: rotate(90deg);
    -webkit-transform: rotate(90deg);
    -khtml-transform: rotate(90deg);
    -moz-transform: rotate(90deg);
    -o-transform: rotate(90deg);
    -ms-transform: rotate(90deg);
  }
  50% {
    transform: rotate(180deg);
    -webkit-transform: rotate(180deg);
    -khtml-transform: rotate(180deg);
    -moz-transform: rotate(180deg);
    -o-transform: rotate(180deg);
    -ms-transform: rotate(180deg);
  }
  75% {
    transform: rotate(270deg);
    -webkit-transform: rotate(270deg);
    -khtml-transform: rotate(270deg);
    -moz-transform: rotate(270deg);
    -o-transform: rotate(270deg);
    -ms-transform: rotate(270deg);
  }
  100% {
    transform: rotate(360deg);
    -webkit-transform: rotate(360deg);
    -khtml-transform: rotate(360deg);
    -moz-transform: rotate(360deg);
    -o-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
  }
}
.menu-phone li ul.list-v {
  right: calc(100% - 0.5 * 16px);
}
.menu-phone li ul.list-v ul {
  right: calc(100% - 0.5 * 16px);
}
#wrapper {
  max-width: 1080px;
  margin: auto;
}
@media screen and (min-width: 2048px) {
  #wrapper {
    max-width: 55vw;
  }
}
#wrapper .menu {
  -webkit-box-flex: 1;
  -moz-box-flex: 1;
  -webkit-flex: 1 1;
  -ms-flex: 1 1;
  flex: 1 1;
  margin: 0 16px 0 0;
}
#wrapper .menu .list-v ul {
  left: calc(100% - 0.5 * 16px);
}
.menu-phone {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: none;
  margin-top: 16px;
  right: 8px;
  transition: all 0.28s ease;
  -webkit-transition: all 0.28s ease;
  -khtml-transition: all 0.28s ease;
  -moz-transition: all 0.28s ease;
  -o-transition: all 0.28s ease;
  -ms-transition: all 0.28s ease;
}
.menu-phone ul {
  right: calc(100% - 0.5 * 16px);
}
@media screen and (max-width: 500px) {
  .menu-phone {
    display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
    display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
    display: block;
  }
}
.l_header {
  max-width: 65vw;
  left: calc((100% - 65vw) * 0.5);
  border-bottom-left-radius: 8px;
  border-bottom-right-radius: 8px;
}
@media screen and (max-width: 2048px) {
  .l_header {
    max-width: 1112px;
    left: calc((100% - 1112px) * 0.5);
  }
}
@media screen and (max-width: 1112px) {
  .l_header {
    left: 0;
    border-radius: 0;
    -webkit-border-radius: 0;
    max-width: 100%;
  }
}
@media screen and (max-width: 500px) {
  .l_header .container {
    margin-left: 0;
    margin-right: 0;
  }
  .l_header #wrapper .nav-main .title {
    padding-left: 16px;
    padding-right: 16px;
  }
  .l_header #wrapper .nav-sub {
    width: 100%;
  }
  .l_header #wrapper .nav-sub .title {
    overflow-y: scroll;
    margin-top: 2px;
    padding: 8px 16px;
  }
  .l_header #wrapper .switcher {
    display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
    display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
    display: -ms-flexbox /* TWEENER - IE 10 */;
    display: -webkit-flex /* NEW - Chrome */;
    display: flex /* NEW, Spec - Opera 12.1, Firefox 20+ */;
    display: flex;
    margin-right: 8px;
  }
  .l_header .menu {
    display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
    display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
    display: none;
  }
}
@media screen and (max-width: 500px) {
  .list-v li {
    max-width: 270px;
  }
}
#u-search {
  display: -webkit-box /* OLD - iOS 6-, Safari 3.1-6 */;
  display: -moz-box /* OLD - Firefox 19- (buggy but mostly works) */;
  display: none;
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  padding: 60px 20px;
  z-index: 1001;
}
@media screen and (max-width: 680px) {
  #u-search {
    padding: 0px;
  }
}

  </style>
  <link rel="stylesheet" href="/css/style.css" media="print" onload="this.media='all';this.onload=null">
  <noscript><link rel="stylesheet" href="/css/style.css"></noscript>
  
<script>
if (/*@cc_on!@*/false || (!!window.MSInputMethodContext && !!document.documentMode))
    document.write(
	'<style>'+
		'html{'+
			'overflow-x: hidden !important;'+
			'overflow-y: hidden !important;'+
		'}'+
		'.kill-ie{'+
			'text-align:center;'+
			'height: 100%;'+
			'margin-top: 15%;'+
			'margin-bottom: 5500%;'+
		'}'+
    '.kill-t{'+
      'font-size: 2rem;'+
    '}'+
    '.kill-c{'+
      'font-size: 1.2rem;'+
    '}'+
		'#l_header,#l_body{'+
			'display: none;'+
		'}'+
	'</style>'+
    '<div class="kill-ie">'+
        `<span class="kill-t"><b>抱歉，您的浏览器无法访问本站</b></span><br/>`+
        `<span class="kill-c">微软已经于2016年终止了对 Internet Explorer (IE) 10 及更早版本的支持，<br/>继续使用存在极大的安全隐患，请使用当代主流的浏览器进行访问。</span><br/>`+
        `<a target="_blank" rel="noopener" href="https://blogs.windows.com/windowsexperience/2021/05/19/the-future-of-internet-explorer-on-windows-10-is-in-microsoft-edge/"><strong>了解详情 ></strong></a>`+
    '</div>');
</script>


<noscript>
	<style>
		html{
			overflow-x: hidden !important;
			overflow-y: hidden !important;
		}
		.kill-noscript{
			text-align:center;
			height: 100%;
			margin-top: 15%;
			margin-bottom: 5500%;
		}
    .kill-t{
      font-size: 2rem;
    }
    .kill-c{
      font-size: 1.2rem;
    }
		#l_header,#l_body{
			display: none;
		}
	</style>
    <div class="kill-noscript">
        <span class="kill-t"><b>抱歉，您的浏览器无法访问本站</b></span><br/>
        <span class="kill-c">本页面需要浏览器支持（启用）JavaScript</span><br/>
        <a target="_blank" rel="noopener" href="https://www.baidu.com/s?wd=启用JavaScript"><strong>了解详情 ></strong></a>
    </div>
</noscript>


  <script>
  /************这个文件存放不需要重载的全局变量和全局函数*********/
  window.volantis = {}; // volantis 全局变量
  volantis.debug = "env"; // 调试模式
  volantis.dom = {}; // 页面Dom see: /source/js/app.js etc.
  /******************** volantis.dom.$ ********************************/
  // 注：这里没有选择器，也没有forEach一次只处理一个dom，这里重新封装主题常用的dom方法，返回的是dom对象，对象包含了以下方法，同时保留dom的原生API
  function volantisDom(ele) {
    if (!ele) ele = document.createElement("div")
    this.ele = ele;
    // ==============================================================
    this.ele.find = (c) => {
      let q = this.ele.querySelector(c)
      if (q)
        return new volantisDom(q)
    }
    // ==============================================================
    this.ele.hasClass = (c) => {
      return this.ele.className.match(new RegExp('(\\s|^)' + c + '(\\s|$)'));
    }
    this.ele.addClass = (c) => {
      this.ele.classList.add(c);
      return this.ele
    }
    this.ele.removeClass = (c) => {
      this.ele.classList.remove(c);
      return this.ele
    }
    this.ele.toggleClass = (c) => {
      if (this.ele.hasClass(c)) {
        this.ele.removeClass(c)
      } else {
        this.ele.addClass(c)
      }
      return this.ele
    }
    // ==============================================================
    this.ele.on = (c, f) => {
      this.ele.addEventListener(c, f, false)
      return this.ele
    }
    this.ele.click = (f, r) => {
      this.ele.on("click", f, r)
      return this.ele
    }
    this.ele.scroll = (f, r) => {
      this.ele.on("scroll", f, r)
      return this.ele
    }
    // ==============================================================
    this.ele.html = (c) => {
      // if(c=== undefined){
      //   return this.ele.innerHTML
      // }else{
      this.ele.innerHTML = c
      return this.ele
      // }
    }
    // ==============================================================
    this.ele.hide = (c) => {
      this.ele.style.display = "none"
      return this.ele
    }
    this.ele.show = (c) => {
      this.ele.style.display = "block"
      return this.ele
    }
    // ==============================================================
    return this.ele
  }
  volantis.dom.$ = (ele) => {
    return !!ele ? new volantisDom(ele) : null;
  }
  /******************** RunItem ********************************/
  function RunItem() {
    this.list = []; // 存放回调函数
    this.start = () => {
      for (var i = 0; i < this.list.length; i++) {
        this.list[i].run();
      }
    };
    this.push = (fn, name, setRequestAnimationFrame = true) => {
      let myfn = fn
      if (setRequestAnimationFrame) {
        myfn = () => {
          volantis.requestAnimationFrame(fn)
        }
      }
      var f = new Item(myfn, name);
      this.list.push(f);
    };
    this.remove = (name) => {
      for (let index = 0; index < this.list.length; index++) {
        const e = this.list[index];
        if (e.name == name) {
          this.list.splice(index, 1);
        }
      }
    }
    // 构造一个可以run的对象
    function Item(fn, name) {
      // 函数名称
      this.name = name || fn.name;
      // run方法
      this.run = () => {
        try {
          fn()
        } catch (error) {
          console.log(error);
        }
      };
    }
  }
  /******************** RightMenu ********************************/
  // volantis.rightmenu.handle(callBack[,"callBackName"]) 外部菜单项控制
  // 可在 volantis.mouseEvent 处获取右键事件
  volantis.rightmenu = {};
  volantis.rightmenu.method = {
    handle: new RunItem(),
  }
  volantis.rightmenu = Object.assign(volantis.rightmenu, {
    handle: volantis.rightmenu.method.handle.push,
  });
  /********************  Dark Mode  ********************************/
  // /layout/_partial/scripts/darkmode.ejs
  // volantis.dark.mode 当前模式 dark or light
  // volantis.dark.toggle() 暗黑模式触发器
  // volantis.dark.push(callBack[,"callBackName"]) 传入触发器回调函数
  volantis.dark = {};
  volantis.dark.method = {
    toggle: new RunItem(),
  };
  volantis.dark = Object.assign(volantis.dark, {
    push: volantis.dark.method.toggle.push,
  });
  /********************  Message  ********************************/
  // VolantisApp.message
  /********************  isMobile  ********************************/
  // /source/js/app.js
  // volantis.isMobile
  // volantis.isMobileOld
  /********************脚本动态加载函数********************************/
  // volantis.js(src, cb)  cb 可以传入onload回调函数 或者 JSON对象 例如: volantis.js("src", ()=>{}) 或 volantis.js("src", {defer:true,onload:()=>{}})
  // volantis.css(src)

  // 返回Promise对象，如下方法同步加载资源，这利于处理文件资源之间的依赖关系，例如：APlayer 需要在 MetingJS 之前加载
  // (async () => {
  //     await volantis.js("...theme.plugins.aplayer.js.aplayer...")
  //     await volantis.js("...theme.plugins.aplayer.js.meting...")
  // })();

  // 已经加入了setTimeout
  volantis.js = (src, cb) => {
    return new Promise(resolve => {
      setTimeout(function () {
        var HEAD = document.getElementsByTagName("head")[0] || document.documentElement;
        var script = document.createElement("script");
        script.setAttribute("type", "text/javascript");
        if (cb) {
          if (JSON.stringify(cb)) {
            for (let p in cb) {
              if (p == "onload") {
                script[p] = () => {
                  cb[p]()
                  resolve()
                }
              } else {
                script[p] = cb[p]
                script.onload = resolve
              }
            }
          } else {
            script.onload = () => {
              cb()
              resolve()
            };
          }
        } else {
          script.onload = resolve
        }
        script.setAttribute("src", src);
        HEAD.appendChild(script);
      });
    });
  }
  volantis.css = (src) => {
    return new Promise(resolve => {
      setTimeout(function () {
        var link = document.createElement('link');
        link.rel = "stylesheet";
        link.href = src;
        link.onload = resolve;
        document.getElementsByTagName("head")[0].appendChild(link);
      });
    });
  }
  /********************按需加载的插件********************************/
  // volantis.import.jQuery().then(()=>{})
  volantis.import = {
    jQuery: () => {
      if (typeof jQuery == "undefined") {
        return volantis.js("https://cdn.bootcdn.net/ajax/libs/jquery/3.6.0/jquery.min.js")
      } else {
        return new Promise(resolve => {
          resolve()
        });
      }
    }
  }
  /********************** requestAnimationFrame ********************************/
  // 1、requestAnimationFrame 会把每一帧中的所有 DOM 操作集中起来，在一次重绘或回流中就完成，并且重绘或回流的时间间隔紧紧跟随浏览器的刷新频率，一般来说，这个频率为每秒60帧。
  // 2、在隐藏或不可见的元素中，requestAnimationFrame 将不会进行重绘或回流，这当然就意味着更少的的 cpu，gpu 和内存使用量。
  volantis.requestAnimationFrame = (fn) => {
    if (!window.requestAnimationFrame) {
      window.requestAnimationFrame = window.requestAnimationFrame || window.mozRequestAnimationFrame || window.webkitRequestAnimationFrame;
    }
    window.requestAnimationFrame(fn)
  }
  /************************ layoutHelper *****************************************/
  volantis.layoutHelper = (helper, html, opt) => {
    opt = Object.assign({ clean: false}, opt)
    function myhelper(helper, html, clean) {
      volantis.tempDiv = document.createElement("div");
      volantis.tempDiv.innerHTML = html;
      let layoutHelper = document.querySelector("#layoutHelper-" + helper)
      if (layoutHelper) {
        if (clean) {
          layoutHelper.innerHTML = ""
        }
        layoutHelper.append(volantis.tempDiv);
      }
    }
    myhelper(helper, html, opt.clean)
  }
  /****************************** 滚动事件处理 ****************************************/
  volantis.scroll = {
    engine: new RunItem(),
    unengine: new RunItem(),
  };
  volantis.scroll = Object.assign(volantis.scroll, {
    push: volantis.scroll.engine.push,
  });
  // 滚动条距离顶部的距离
  volantis.scroll.getScrollTop = () => {
    let scrollPos;
    if (window.pageYOffset) {
      scrollPos = window.pageYOffset;
    } else if (document.compatMode && document.compatMode != 'BackCompat') {
      scrollPos = document.documentElement.scrollTop;
    } else if (document.body) {
      scrollPos = document.body.scrollTop;
    }
    return scrollPos;
  }
  volantis.scroll.scrollHeight = function () {
    return Math.max(
      document.body.scrollHeight,
      document.documentElement.scrollHeight,
    )
  }
  volantis.scroll.offsetHeight = function () {
    return Math.max(
      document.body.offsetHeight,
      document.documentElement.offsetHeight,
      document.body.clientHeight,
      document.documentElement.clientHeight
    )
  }
  volantis.scroll.progress = function () {
    return volantis.scroll.getScrollTop() / (volantis.scroll.scrollHeight() - volantis.scroll.offsetHeight())
  }
  // 使用 requestAnimationFrame 处理滚动事件
  // `volantis.scroll.del` 中存储了一个数值, 该数值检测一定时间间隔内滚动条滚动的位移, 数值的检测频率是浏览器的刷新频率. 数值为正数时, 表示向下滚动. 数值为负数时, 表示向上滚动.
  volantis.scroll.handleScrollEvents = () => {
    volantis.scroll.lastScrollTop = volantis.scroll.getScrollTop()
    function loop() {
      const scrollTop = volantis.scroll.getScrollTop();
      if (volantis.scroll.lastScrollTop !== scrollTop) {
        volantis.scroll.del = scrollTop - volantis.scroll.lastScrollTop;
        volantis.scroll.lastScrollTop = scrollTop;
        // if (volantis.scroll.del > 0) {
        //   console.log("向下滚动");
        // } else {
        //   console.log("向上滚动");
        // }
        // 注销过期的unengine未滚动事件
        volantis.scroll.unengine.list = []
        volantis.scroll.engine.start();
      } else {
        volantis.scroll.unengine.start();
      }
      volantis.requestAnimationFrame(loop)
    }
    volantis.requestAnimationFrame(loop)
  }
  volantis.scroll.handleScrollEvents()
  volantis.scroll.ele = null;
  // 触发页面滚动至目标元素位置
  volantis.scroll.to = (ele, option = {}) => {
    if (!ele) return;
    volantis.scroll.ele = ele;
    // 默认配置
    opt = {
      top: ele.getBoundingClientRect().top + document.documentElement.scrollTop,
      behavior: "smooth"
    }
    // 定义配置
    if ("top" in option) {
      opt.top = option.top
    }
    if ("behavior" in option) {
      opt.behavior = option.behavior
    }
    if ("addTop" in option) {
      opt.top += option.addTop
    }
    if (!("observerDic" in option)) {
      option.observerDic = 100
    }
    // 滚动
    window.scrollTo(opt);
    // 监视器
    // 监视并矫正元素滚动到指定位置
    // 用于处理 lazyload 引起的 cls 导致的定位失败问题
    // option.observer = false
    if (option.observer) {
      setTimeout(() => {
        if (volantis.scroll.ele != ele) {
          return
        }
        volantis.scroll.unengine.push(() => {
          let me = ele.getBoundingClientRect().top
          if (!(me >= -option.observerDic && me <= option.observerDic)) {
            volantis.scroll.to(ele, option)
          }
          volantis.scroll.unengine.remove("unengineObserver")
        }, "unengineObserver")
      }, 1000)
    }
  }
  /********************** Content Visibility ********************************/
  // 见 source/css/first.styl 如果遇到任何问题 删除 .post-story 即可
  // 一个元素被声明 content-visibility 属性后 如果元素不在 viewport 中 浏览器不会计算其后代元素样式和属性 从而节省 Style & Layout 耗时
  // content-visibility 的副作用: 锚点失效 等等(实验初期 暂不明确), 使用此方法清除样式
  volantis.cleanContentVisibility = () => {
    if (document.querySelector(".post-story")) {
      console.log("cleanContentVisibility");
      document.querySelectorAll(".post-story").forEach(e => {
        e.classList.remove("post-story")
      })
    }
  }
  /******************************************************************************/
  /******************************************************************************/
  /******************************************************************************/
  //图像加载出错时的处理
  function errorImgAvatar(img) {
    img.src = "https://unpkg.com/volantis-static@0.0.1660614606622/media/placeholder/avatar/round/3442075.svg";
    img.onerror = null;
  }
  function errorImgCover(img) {
    img.src = "https://unpkg.com/volantis-static@0.0.1660614606622/media/placeholder/cover/76b86c0226ffd.svg";
    img.onerror = null;
  }
  /******************************************************************************/
</script>

<script>
  window.pdata={}
  pdata.ispage=true;
  pdata.commentPath="";
  pdata.commentPlaceholder="";
  pdata.commentConfig={};
  volantis.GLOBAL_CONFIG ={
    root: '/',
    debug: "env",
    default: {"avatar":"https://unpkg.com/volantis-static@0.0.1660614606622/media/placeholder/avatar/round/3442075.svg","link":"https://unpkg.com/volantis-static@0.0.1660614606622/media/placeholder/link/8f277b4ee0ecd.svg","cover":"https://unpkg.com/volantis-static@0.0.1660614606622/media/placeholder/cover/76b86c0226ffd.svg","image":"https://unpkg.com/volantis-static@0.0.1660614606622/media/placeholder/image/2659360.svg"},
    lastupdate: new Date(1716188511136),
    cdn: {
      izitoast_css: 'https://cdn.bootcdn.net/ajax/libs/izitoast/1.4.0/css/iziToast.min.css',
      izitoast_js: 'https://cdn.bootcdn.net/ajax/libs/izitoast/1.4.0/js/iziToast.min.js',
      fancybox_css: 'https://cdn.bootcdn.net/ajax/libs/fancyapps-ui/4.0.31/fancybox.min.css',
      fancybox_js: 'https://cdn.bootcdn.net/ajax/libs/fancyapps-ui/4.0.31/fancybox.umd.min.js'
    },
    sidebar: {
      for_page: ["blogger","category","tagcloud","donate"],
      for_post: ["toc"],
      webinfo: {
        lastupd: {
          enable: true,
          friendlyShow: true
        },
        runtime: {
          data: "2020/01/01",
          unit: "天"
        }
      }
    },
    plugins: {
      message: {"enable":true,"icon":{"default":"fa-solid fa-info-circle light-blue","quection":"fa-solid fa-question-circle light-blue"},"time":{"default":5000,"quection":20000},"position":"topRight","transitionIn":"bounceInLeft","transitionOut":"fadeOutRight","titleColor":"var(--color-text)","messageColor":"var(--color-text)","backgroundColor":"var(--color-card)","zindex":2147483647,"copyright":{"enable":true,"title":"知识共享许可协议","message":"请遵守 CC BY-NC-SA 4.0 协议。","icon":"far fa-copyright light-blue"},"aplayer":{"enable":true,"play":"fa-solid fa-play","pause":"fa-solid fa-pause"},"rightmenu":{"enable":true,"notice":true}},
      
      
      rightmenus: {"enable":"ture","order":["plugins.navigation","hr","plugins.inputBox","plugins.seletctText","plugins.elementCheck","plugins.elementImage","menus.link","hr","menus.darkMode","plugins.articlePage","music"],"options":{"iconPrefix":"fa-solid","articleShowLink":true,"musicAlwaysShow":true},"plugins":{"navigation":[{"id":"left","name":"转到上一页","icon":"fa-solid fa-arrow-left","event":"history.back()","group":"navigation"},{"id":"right","name":"转到下一页","icon":"fa-solid fa-arrow-right","event":"history.forward()","group":"navigation"},{"id":"redo","name":"刷新当前页面","icon":"fa-solid fa-redo","event":"window.location.reload()","group":"navigation"},{"id":"up","name":"回到顶部","icon":"fa-solid fa-arrow-up","event":"VolantisApp.scrolltoElement(volantis.dom.bodyAnchor)","group":"navigation"}],"inputBox":[{"id":"copyPaste","name":"粘贴文本","icon":"fa-solid fa-paste","event":"copyPaste","group":"inputBox"},{"id":"copyAll","name":"全选文本","icon":"fa-solid fa-object-ungroup","event":"copyAll","group":"inputBox"},{"id":"copyCut","name":"剪切文本","icon":"fa-solid fa-cut","event":"copyCut","group":"inputBox"}],"seletctText":[{"id":"copyText","name":"复制文本","icon":"fa-solid fa-copy","event":"copyText","group":"seletctText"},{"id":"searchWord","name":"站内搜索","icon":"fa-solid fa-search","event":"OpenSearch(__text__)","group":"seletctText"},{"id":"bingSearch","name":"必应搜索","icon":"fa-solid fa-search","event":"window.open(`https://cn.bing.com/search?q=${__text__}`)","group":"seletctText"}],"elementCheck":[{"id":"openTab","name":"新标签页打开","icon":"fa-solid fa-external-link-square-alt","event":"window.open(__link__)","group":"elementCheck"},{"id":"copyLink","name":"复制链接地址","icon":"fa-solid fa-link","event":"copyLink","group":"elementCheck"}],"elementImage":[{"id":"copyImg","name":"复制图片","icon":"fa-solid fa-image","event":"copyImg","group":"elementImage"},{"id":"googleImg","name":"谷歌识图","icon":"fa-solid fa-images","event":"window.open(`https://www.google.com.hk/searchbyimage?image_url=${__link__}`)","group":"elementImage"}],"articlePage":[{"id":"printMode","name":"打印页面","icon":"fa-solid fa-print","event":"printMode","group":"articlePage"},{"id":"readMode","name":"阅读模式","icon":"fa-solid fa-book-open","event":"readMode","group":"articlePage"}]},"menus":{"link":[{"id":"help","name":"常见问题","icon":"fa-solid fa-question","link":"https://volantis.js.org/faqs/","group":"link"},{"id":"examples","name":"示例博客","icon":"fa-solid fa-rss","link":"https://volantis.js.org/examples/","group":"link"},{"id":"contributors","name":"加入社区","icon":"fa-solid fa-fan","link":"https://volantis.js.org/contributors/","group":"link"},"hr",{"id":"source_docs","name":"本站源码","icon":"fa-solid fa-code-branch","link":"https://github.com/volantis-x/volantis-docs/","group":"link"},{"id":"source_theme","name":"主题源码","icon":"fa-solid fa-code-branch","link":"https://github.com/volantis-x/hexo-theme-volantis/","group":"link"}],"darkMode":[{"id":"darkMode","name":"暗黑模式","icon":"fa-solid fa-moon","event":"volantis.dark.toggle()","group":"darkMode"}]}}
      
    },
    search: {
      

      
        dataPath: (("/" || "/").endsWith("/") ? "/" || "/" : "//" || "/") + "content.json",
      

      
    },
    languages: {
      
        search: {
          hits_empty: "找不到您查询的内容：${query}",
          hits_stats: "找到 ${hits} 条结果，用时 ${time} 毫秒",
        }
      
    }
  }
</script>
  <!-- import head_end begin -->
  <!-- import head_end end -->
  <!-- Custom Files headEnd begin-->
  
  <!-- Custom Files headEnd end-->
</head>
  <body itemscope itemtype="http://schema.org/WebPage">
    <!-- import body_begin begin-->
    <!-- import body_begin end-->
    <!-- Custom Files bodyBegin begin-->
    
    <!-- Custom Files bodyBegin end-->
    <header itemscope itemtype="http://schema.org/WPHeader" id="l_header" class="l_header auto shadow blur show" style='opacity: 0' >
  <div class='container'>
  <div id='wrapper'>
    <div class='nav-sub'>
      <p class="title"></p>
      <ul class='switcher nav-list-h m-phone'>
        <li><a id="s-comment" class="fa-solid fa-comments fa-fw" target="_self" title="comment"></a></li>
        
          <li><a id="s-toc" class="s-toc fa-solid fa-list fa-fw" target="_self" title="toc"></a></li>
        
      </ul>
    </div>
		<div class="nav-main">
      
        
        <a class="title flat-box" target="_self" href='/'>
          
            <img no-lazy class='logo' src='https://unpkg.com/volantis-static@0.0.1660614606622/media/org.volantis/blog/Logo-NavBar@3x.png'/>
          
          
          
        </a>
      

			<div class='menu navigation'>
				<ul class='nav-list-h m-pc'>
          
          
          
            
            
              <li>
                <a class="menuitem flat-box faa-parent animated-hover"
                href="/" title="博客"
                  
                  
                  
                    active-action="action-home"
                  >
                  <i class='fa-solid fa-rss fa-fw'></i>博客
                </a>
                
              </li>
            
          
          
            
            
              <li>
                <a class="menuitem flat-box faa-parent animated-hover"
                href="/categories/" title="分类"
                  
                  
                  
                    active-action="action-categories"
                  >
                  <i class='fa-solid fa-folder-open fa-fw'></i>分类
                </a>
                
              </li>
            
          
          
            
            
              <li>
                <a class="menuitem flat-box faa-parent animated-hover"
                href="/tags/" title="标签"
                  
                  
                  
                    active-action="action-tags"
                  >
                  <i class='fa-solid fa-tags fa-fw'></i>标签
                </a>
                
              </li>
            
          
          
            
            
              <li>
                <a class="menuitem flat-box faa-parent animated-hover"
                href="/archives/" title="归档"
                  
                  
                  
                    active-action="action-archives"
                  >
                  <i class='fa-solid fa-archive fa-fw'></i>归档
                </a>
                
              </li>
            
          
          
            
            
              <li>
                <a class="menuitem flat-box faa-parent animated-hover"
                href="/friends/" title="友链"
                  
                  
                  
                    active-action="action-friends"
                  >
                  <i class='fa-solid fa-link fa-fw'></i>友链
                </a>
                
              </li>
            
          
          
            
            
              <li>
                <a class="menuitem flat-box faa-parent animated-hover"
                href="/about/" title="关于"
                  
                  
                  
                    active-action="action-about"
                  >
                  <i class='fa-solid fa-info-circle fa-fw'></i>关于
                </a>
                
              </li>
            
          
          
				</ul>
			</div>
      
      <div class="m_search">
        <form name="searchform" class="form u-search-form">
          <i class="icon fa-solid fa-search fa-fw"></i>
          <input type="text" class="input u-search-input" placeholder="Search..." />
        </form>
      </div>
      

			<ul class='switcher nav-list-h m-phone'>
				
					<li><a class="s-search fa-solid fa-search fa-fw" target="_self" title="search"></a></li>
				
				<li>
          <a class="s-menu fa-solid fa-bars fa-fw" target="_self" title="menu"></a>
          <ul class="menu-phone list-v navigation white-box">
            
              
            
              <li>
                <a class="menuitem flat-box faa-parent animated-hover"
                href="/" title="博客"
                  
                  
                  
                    active-action="action-home"
                  >
                  <i class='fa-solid fa-rss fa-fw'></i>博客
                </a>
                
              </li>
            
          
            
              
            
              <li>
                <a class="menuitem flat-box faa-parent animated-hover"
                href="/categories/" title="分类"
                  
                  
                  
                    active-action="action-categories"
                  >
                  <i class='fa-solid fa-folder-open fa-fw'></i>分类
                </a>
                
              </li>
            
          
            
              
            
              <li>
                <a class="menuitem flat-box faa-parent animated-hover"
                href="/tags/" title="标签"
                  
                  
                  
                    active-action="action-tags"
                  >
                  <i class='fa-solid fa-tags fa-fw'></i>标签
                </a>
                
              </li>
            
          
            
              
            
              <li>
                <a class="menuitem flat-box faa-parent animated-hover"
                href="/archives/" title="归档"
                  
                  
                  
                    active-action="action-archives"
                  >
                  <i class='fa-solid fa-archive fa-fw'></i>归档
                </a>
                
              </li>
            
          
            
              
            
              <li>
                <a class="menuitem flat-box faa-parent animated-hover"
                href="/friends/" title="友链"
                  
                  
                  
                    active-action="action-friends"
                  >
                  <i class='fa-solid fa-link fa-fw'></i>友链
                </a>
                
              </li>
            
          
            
              
            
              <li>
                <a class="menuitem flat-box faa-parent animated-hover"
                href="/about/" title="关于"
                  
                  
                  
                    active-action="action-about"
                  >
                  <i class='fa-solid fa-info-circle fa-fw'></i>关于
                </a>
                
              </li>
            
          
            
          </ul>
        </li>
			</ul>

      <!-- Custom Files header begin -->
      
      <!-- Custom Files header end -->
		</div>
	</div>
  </div>
</header>

    <div id="l_body">
      <div id="l_cover">
  
</div>

      <div id="safearea">
        <div class="body-wrapper">
          
<div id="l_main" class=''>
  <article itemscope itemtype="http://schema.org/Article" class="article post white-box reveal md shadow floatable blur article-type-post" id="post" itemscope itemprop="blogPost">
  <link itemprop="mainEntityOfPage" href="https://snowman12137.github.io/2024/03/25/生信分析-通过共表达网络分析（WGCNA）/">
  <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
    <meta itemprop="name" content="小吴の博客">
  </span>
  <span hidden itemprop="post" itemscope itemtype="http://schema.org/Post">
    <meta itemprop="name" content="小吴の博客">
    <meta itemprop="description" content="我的博客">
  </span>
  


  
    <span hidden>
      <meta itemprop="image" content="https://unpkg.com/volantis-static@0.0.1660614606622/media/org.volantis/blog/favicon/android-chrome-192x192.png">
    </span>
  
  <div class="article-meta" id="top">
    
    
    
      <h1 class="title" itemprop="name headline">
        通过共表达网络分析（WGCNA）
      </h1>
      <div class='new-meta-box'>
        
          
            
<div class='new-meta-item author' itemprop="author" itemscope itemtype="http://schema.org/Person">
  <a itemprop="url" class='author' href="/" rel="nofollow">
    <img itemprop="image" src="https://unpkg.com/volantis-static@0.0.1660614606622/media/org.volantis/blog/favicon/apple-touch-icon.png" class="lazyload" data-srcset="https://unpkg.com/volantis-static@0.0.1660614606622/media/org.volantis/blog/favicon/apple-touch-icon.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==">
    <p itemprop="name">Sn0wm1an</p>
  </a>
</div>

          
        
          
            
  <div class="new-meta-item wordcount">
    <a class='notlink'>
      <i class="fa-solid fa-keyboard fa-fw" aria-hidden="true"></i>
      <p>字数：9.6k 字</p>
    </a>
  </div>
  <div class="new-meta-item readtime">
    <a class='notlink'>
      <i class="fa-solid fa-hourglass-half fa-fw" aria-hidden="true"></i>
      <p>时长：43 分钟</p>
    </a>
  </div>


          
        
          
            
  
  
  
  <div class='new-meta-item category'>
    <i class="fa-solid fa-folder-open fa-fw" aria-hidden="true"></i>
    <a class="category-link" href="/categories/%E7%94%9F%E7%89%A9%E8%AE%A1%E7%AE%97%E6%9C%BA%E7%A7%91%E5%AD%A6/">生物计算机科学</a>
    
      <span hidden itemprop="about" itemscope itemtype="http://schema.org/Thing">
        <a href="/categories/%E7%94%9F%E7%89%A9%E8%AE%A1%E7%AE%97%E6%9C%BA%E7%A7%91%E5%AD%A6/" itemprop="url"><span itemprop="name">生物计算机科学</span></a>
      </span>
    
  </div>
  



          
        
          
            <div class="new-meta-item date" itemprop="dateCreated datePublished" datetime="2024-03-25T00:00:00+08:00">
  <a class='notlink'>
    <i class="fa-solid fa-calendar-alt fa-fw" aria-hidden="true"></i>
    <p>发布于：2024年3月25日</p>
  </a>
</div>

          
        
          
            



          
        
        <!-- Custom Files topMeta begin-->
        
        <!-- Custom Files topMeta end-->
      </div>
    
  </div>


  <div id="layoutHelper-page-plugins"></div>
  <div id="post-body" itemprop="articleBody">
    <h1 id="通过共表达网络分析（WGCNA）"><a href="#通过共表达网络分析（WGCNA）" class="headerlink" title="通过共表达网络分析（WGCNA）"></a>通过共表达网络分析（WGCNA）</h1><p>文章解析：谭永飞. <em>通过共表达网络分析 (WGCNA) 探究前列腺癌发生与进展相关功能基因模块</em>. MS thesis. 吉林大学, 2018.</p>
<p>数据来源：TCGA（The Cancer Genome Atlas） (<a target="_blank" rel="noopener" href="https://portal.gdc.cancer.gov/">https://portal.gdc.cancer.gov/</a>) </p>
<h1 id="一、数据下载与处理"><a href="#一、数据下载与处理" class="headerlink" title="一、数据下载与处理"></a>一、数据下载与处理</h1><p>文章使用了TCGA-Assembler的R程序包下载TCGA规范化处理过后的LV3开放前列腺癌数据和临床信息总共553个癌症样本文件，包括正常样本52个，癌症样本501个。分别下载这553个样本的count以及FPKM数据和临床信息。</p>
<p>参考安装文章：<a target="_blank" rel="noopener" href="https://cloud.tencent.com/developer/article/1481868">https://cloud.tencent.com/developer/article/1481868</a></p>
<p>注：</p>
<ul>
<li><strong>TCGA数据等级：</strong></li>
</ul>
<p>level1：原始数据</p>
<p>level2：处理过的数据</p>
<p>level3：经过分割、解释的数据</p>
<p>level4：感兴趣的区域或概要</p>
<p>总而言之，前面2个层级的数据一般是拿不到的，需要权限，一般也只有国外的PI才能申请到（听说的），我们一般拿到的open数据就属于那种已经标准化后的数据。</p>
<h2 id="1-源码下载"><a href="#1-源码下载" class="headerlink" title="1.源码下载"></a>1.源码下载</h2><h3 id="a-文件准备"><a href="#a-文件准备" class="headerlink" title="a.文件准备"></a>a.文件准备</h3><p>最重要的文件只有两个<code>Moudule_A.R和Moudule_B.R</code> 所以我们把它们复制到工作文件夹下,这个Module_A主要是用来下载数据的，而Module_B主要用来分析数据；</p>
<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252242829.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252242829.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#创建工作目录(注意不要有中文)</span></span><br><span class="line"><span class="built_in">mkdir</span> workspace</span><br><span class="line"><span class="built_in">cd</span> workspace</span><br><span class="line"><span class="comment">#克隆源码到工作文件夹下</span></span><br><span class="line">git <span class="built_in">clone</span> https://github.com/compgenome365/TCGA-Assembler-2.git</span><br><span class="line"><span class="built_in">cp</span> ./TCGA-Assembler-2/TCGA-Assembler/Module_A.R ./Module_A.R</span><br><span class="line"><span class="built_in">cp</span> ./TCGA-Assembler-2/TCGA-Assembler/Module_B.R ./Module_B.R</span><br><span class="line"><span class="comment">#创建下载后存放数据的文件夹</span></span><br><span class="line"><span class="built_in">mkdir</span> data</span><br></pre></td></tr></table></figure>

<p>然后打开R软件，设置工作目录，直接使用代码：setwd(“&#x2F;home&#x2F;gc&#x2F;workspace&#x2F;xue”) 来实现，输入这行代码后，可通过getwd()来获取当前工作目录，确认是否设置成功，要不然后面会出错。还可通过软件直接设置：文件&gt;&gt;改变工作目录，然后选择刚刚那个文件路径就行了，getwd()再验证一下。</p>
<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252242515.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252242515.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<h3 id="b-R包准备"><a href="#b-R包准备" class="headerlink" title="b.R包准备"></a>b.R包准备</h3><p>加载需要用的包，安装好下面几个包以后，后续载入TCGA_assemble文件夹中的两个模块(Module_A.R和Module_B.R)时，如果提示错误，一般是缺少包，提示缺什么安装什么就行了：</p>
<p>（原文版本太低了）</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">install.packages(<span class="string">&quot;BiocManager&quot;</span>)</span><br><span class="line">BiocManager::install(c(<span class="string">&quot;httr&quot;</span>, <span class="string">&quot;RCurl&quot;</span>, <span class="string">&quot;stringr&quot;</span>, <span class="string">&quot;HGNChelper&quot;</span>, <span class="string">&quot;rjson&quot;</span>))</span><br><span class="line">library(httr)</span><br><span class="line">library(bitops)</span><br><span class="line">library(RCurl)</span><br><span class="line">library(stringr)</span><br><span class="line">library(HGNChelper)</span><br><span class="line">library(rjson)</span><br></pre></td></tr></table></figure>

<h2 id="2-下载数据"><a href="#2-下载数据" class="headerlink" title="2.下载数据"></a>2.下载数据</h2><h3 id="a-下载生物样本临床数据"><a href="#a-下载生物样本临床数据" class="headerlink" title="a.下载生物样本临床数据"></a>a.<strong>下载生物样本临床数据</strong></h3><p>核心代码</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">DownloadBiospecimenClinicalData(cancerType, saveFolderName = <span class="string">&quot;./BiospecimenClinicalData&quot;</span>,outputFileName = <span class="string">&quot;&quot;</span>)  </span><br></pre></td></tr></table></figure>

<p><strong>参数说明：</strong></p>
<p><strong>cancerType:</strong>  一个字符串，指示应为其下载数据的指定癌症类型。它的值可以是任何癌症类型的缩写，可以查看 Manual.pdf文档，也可去官网<a target="_blank" rel="noopener" href="https://tcga-data.nci.nih.gov/docs/publications/tcga/">https://tcga-data.nci.nih.gov/docs/publications/tcga/</a>  查询各种癌症的缩写。这里列举部分出来。</p>
<table>
<thead>
<tr>
<th>cancerType</th>
<th>Canncer Type Full Name</th>
</tr>
</thead>
<tbody><tr>
<td>ACC</td>
<td>Adrenocortical carcinoma</td>
</tr>
<tr>
<td>BLCA</td>
<td>Bladder urothelial carcinoma</td>
</tr>
<tr>
<td>BRCA</td>
<td>Breast invasive carcinoma</td>
</tr>
<tr>
<td>CESC</td>
<td>Cervical squamous cell carcinoma and endocervical adenocarcinoma</td>
</tr>
</tbody></table>
<p>案例：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">source</span>(<span class="string">&quot;Module_A.R&quot;</span>)  </span><br><span class="line">filename_biosClin &lt;- DownloadBiospecimenClinicalData(cancerType = <span class="string">&quot;BLCA&quot;</span>, saveFolderName =<span class="string">&quot;./data/ManualExampleData/RawData.TCGA-Assembler/BiospecimenClinicalData&quot;</span>, outputFileName = <span class="string">&quot;test&quot;</span>)  </span><br></pre></td></tr></table></figure>

<p>报错</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">sh: 1: curl: not found</span><br><span class="line">Error in system2(&quot;curl&quot;, arg, stdout = T) : error in running command</span><br></pre></td></tr></table></figure>

<p>因为安装的curl在conda里面，而R包执行时是在系统里面执行，在系统里安装curl时发现有依赖问题，于是我们把conda的curl加入到环境变量里即可</p>
<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252242199.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252242199.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">whereis curl</span><br><span class="line">sudo vim ~/.bashrc</span><br><span class="line"><span class="comment">#写入你的curl路径</span></span><br><span class="line"><span class="built_in">export</span> PATH=<span class="string">&quot;/home/gc/anaconda3/bin/curl:<span class="variable">$PATH</span>&quot;</span></span><br><span class="line"><span class="built_in">source</span> ~/.bashrc</span><br></pre></td></tr></table></figure>

<p>或者解决依赖关系</p>
<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252242186.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252242186.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">apt-get purge libcurl4</span><br><span class="line">apt-get install curl</span><br></pre></td></tr></table></figure>

<h3 id="b-下载拷贝数据"><a href="#b-下载拷贝数据" class="headerlink" title="b 下载拷贝数据"></a>b 下载拷贝数据</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">DownloadCNAData(cancerType, assayPlatform = NULL, tissueType = NULL, saveFolderName = <span class="string">&quot;.&quot;</span>,outputFileName = <span class="string">&quot;&quot;</span>, inputPatientIDs = NULL)  </span><br></pre></td></tr></table></figure>

<p><strong>参数说明：</strong></p>
<p><strong>cancerType:</strong>  同上</p>
<p><strong>assayPlatform:</strong>  一个字符向量，指示应下载数据的分析平台。它的值可以是cna_cnv.hg18, cna_cnv.hg19, cna_nocnv.hg18, 和cna_nocnv.hg19中的一个或一个组合，它的默认值是NULL，这表示上面所有的测试平台(如果可用)。关于测序平台可参考下表：</p>
<table>
<thead>
<tr>
<th>assayPlatform</th>
<th>Assay</th>
<th>Reference Genome</th>
<th>Additional Description</th>
<th>Filename Pattern</th>
</tr>
</thead>
<tbody><tr>
<td>cna_cnv.hg18</td>
<td>Affymetrix SNP Array 6.0</td>
<td>Hg18</td>
<td>All probes</td>
<td></td>
</tr>
<tr>
<td>cna_cnv.hg19</td>
<td>Affymetrix SNP Array 6.0</td>
<td>Hg19</td>
<td></td>
<td></td>
</tr>
<tr>
<td>cna_nocnv.hg18</td>
<td>Affymetrix SNP Array 6.0</td>
<td>Hg18</td>
<td></td>
<td></td>
</tr>
<tr>
<td>cna_nocnv.hg19</td>
<td>Affymetrix SNP Array 6.0</td>
<td>Hg19</td>
<td></td>
<td></td>
</tr>
</tbody></table>
<p>下载失败，因为旧数据已被清除，于是只能到<a target="_blank" rel="noopener" href="https://portal.gdc.cancer.gov/analysis_page?app=Downloads">官网</a>下载，查找TCGA-PRAD得到498个病例</p>
<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252242668.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252242668.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<p>下载manifest文件，然后到服务器上下载拉取的客户端</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">wget https://gdc.cancer.gov/files/public/file/gdc-client_v1.6.1_Ubuntu_x64.zip</span><br><span class="line">unzip gdc-client_v1.6.1_Ubuntu_x64.zip </span><br><span class="line">./gdc-client download -m gdc_manifest.2024-03-05.txt</span><br></pre></td></tr></table></figure>

<p>数据下载完成</p>
<h2 id="3-数据预处理"><a href="#3-数据预处理" class="headerlink" title="3.数据预处理"></a>3.数据预处理</h2><h3 id="a-进行FPKM进行数据归一化"><a href="#a-进行FPKM进行数据归一化" class="headerlink" title="a. 进行FPKM进行数据归一化"></a>a. 进行FPKM进行数据归一化</h3><p>(每千个碱基的转录每百万映射读取的fragments)。通俗讲，把比对到的某个基因的Fragment数目，除以基因的长度，其比值再除以所有基因的总长度。FPKM可以排除因基因长度、测序深度等因素造成的干扰。</p>
<p>定义表达基因为：在所有肿瘤样本中，FPKM&gt;1的样本数&gt;80%的总肿瘤&#x2F;总瘤旁组织样本数</p>
<p>参考<a target="_blank" rel="noopener" href="https://blog.csdn.net/weixin_49878699/article/details/135373467">https://blog.csdn.net/weixin_49878699&#x2F;article&#x2F;details&#x2F;135373467</a></p>
<p>在学习过程中发现了有处理过后的数据网站，<a target="_blank" rel="noopener" href="https://xena.ucsc.edu/">XENA.UCSC</a>，其就是很直观强大的一个在线网站，里面收录了众多数据库的数据集，其中就包括了TCGA数据集，并且是整合好的，可以直接用于分析。</p>
<ul>
<li>进入官网 launch</li>
</ul>
<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252242617.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252242617.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<ul>
<li>点击dataset</li>
</ul>
<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252242221.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252242221.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<ul>
<li>直接搜索想要的数据</li>
</ul>
<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252242308.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252242308.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<ul>
<li>这里面有</li>
</ul>
<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252242219.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252242219.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<p>counts——转录组的原始表达矩阵，里面对应的基因表达量又被称作raw_count，行名为基因symbol，列名为样本名（也是病人的id，可以将每一列看作一个病人）<br>fpkm——raw_count经过转换后的表达矩阵，其计算公式可参考一文了解Count、FPKM、RPKM、TPM<br>phenotype——病人的临床信息，包含分组信息，肿瘤分期，分级，年龄，性别等<br>survival——病人的生存信息，里面通常会有4列信息，两列的病人的id，另外两列：OS——生存状态（0表示存活，1表示死亡），OS.time——生存时间</p>
<h3 id="b-进行注释-x2F-FPKM"><a href="#b-进行注释-x2F-FPKM" class="headerlink" title="b.进行注释&#x2F;FPKM"></a>b.进行注释&#x2F;FPKM</h3><p>首先安装要用的包</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">BiocManager::install(c(<span class="string">&quot;tidyverse&quot;</span>, <span class="string">&quot;dplyr&quot;</span>, <span class="string">&quot;rtracklayer&quot;</span>))</span><br><span class="line">library(tidyverse)</span><br><span class="line">library(dplyr)</span><br><span class="line">library(rtracklayer)</span><br></pre></td></tr></table></figure>

<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">### 读入下载的数据</span></span><br><span class="line">tcga_count &lt;- read_tsv(file = <span class="string">&#x27;./TCGA-PRAD.htseq_counts.tsv&#x27;</span>) <span class="comment"># count数据</span></span><br><span class="line">tcga_fpkm &lt;- read_tsv(file = <span class="string">&quot;./TCGA-PRAD.htseq_fpkm.tsv&quot;</span>) <span class="comment"># fpkm数据</span></span><br><span class="line">tcga_survival &lt;- read_tsv(file = <span class="string">&quot;./TCGA-PRAD.survival.tsv&quot;</span>) <span class="comment"># 患者生存状态</span></span><br><span class="line">tcga_phenotype &lt;- read_tsv(file = <span class="string">&quot;./TCGA-PRAD.GDC_phenotype.tsv&quot;</span>) <span class="comment"># 患者临床信息</span></span><br><span class="line"><span class="comment">### count数据处理</span></span><br><span class="line">tcga_count &lt;- as.data.frame(tcga_count) <span class="comment"># 将导入的文件转成数据框格式</span></span><br><span class="line">tcga_count &lt;- column_to_rownames(tcga_count, var = <span class="string">&#x27;Ensembl_ID&#x27;</span>) <span class="comment"># 将文件第一列转为行名</span></span><br><span class="line">tcga_count &lt;- 2^tcga_count-1 <span class="comment"># xena下载的数据经过了log2+1转化，需要将其还原</span></span><br></pre></td></tr></table></figure>

<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252242079.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252242079.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<p>接下来要导入一个比较重要的文件<code>gencode.v26.annotation.gtf</code> 它是人类标准基因组注释文件(第27版GRCh38.p10)，可以看到其中的v26，是第26个版本，下载地址如下，选择网址里面的第一个，只含有人类基因的</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">[https://www.gencodegenes.org/human/release_26.html](https://www.gencodegenes.org/human/release_26.html)</span><br><span class="line">wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_26/gencode.v26.annotation.gtf.gz</span><br><span class="line">gzip -d gencode.v26.annotation.gtf.gz</span><br></pre></td></tr></table></figure>

<p>然后导入</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">gene_annotation &lt;- import(<span class="string">&#x27;./gencode.v26.annotation.gtf&#x27;</span>) </span><br><span class="line">gene_annotation &lt;- as.data.frame(gene_annotation)<span class="comment">#将文件转换为数据框格式</span></span><br><span class="line">gene_annotation &lt;- gene_annotation [gene_annotation<span class="variable">$type</span> == <span class="string">&#x27;gene&#x27;</span>, ] <span class="comment"># 筛选为gene的类型</span></span><br><span class="line">gene_annotation &lt;- dplyr::select(gene_annotation, gene_id, gene_name, seqnames, start, end, width, strand, gene_type)</span><br></pre></td></tr></table></figure>

<p>看看最后处理好的gene_annotation长啥样，点击Rstudio右上角的gene_annotation 如下图所示：</p>
<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252243582.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252243582.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<p>gene_id——基因的ensemble_ID<br>gene_name——基因名<br>seqnames——基因位于哪条染色体<br>start——该基因于染色体上的起始位置<br>end——该基因于染色体上的终止位置<br>width——该基因的长度（注意：凭借这个就可以计算FPKM）<br>strand——链的方向，+表示正链，-表示负链（关于链的方向的描述信息可以自行查找资料，在这里跟教程无关不做过多介绍）<br>gene_type——基因的类型，有的是miRNA，有的是lincRNA…</p>
<p>这里我们用到的是前两列gene_id和gene_name</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">gene_symbol &lt;- gene_annotation[,(1:2)] <span class="comment"># 筛选gene_annotation文件中的第一列和第二列</span></span><br><span class="line">colnames(gene_symbol) &lt;- c(<span class="string">&quot;ID&quot;</span>, <span class="string">&quot;symbol&quot;</span>) <span class="comment"># 将gene_symbol文件的列名改成ID和symbol</span></span><br></pre></td></tr></table></figure>

<p>准备好前面的count数据和基因注释文件后，运行下面代码，每句代码后都有注释信息，可以挨个查看，比对处理前后的变化</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">data_tcga &lt;- tcga_count</span><br><span class="line">data_tcga<span class="variable">$ID</span> &lt;- rownames(data_tcga) <span class="comment"># 给data_tcga添加一列，列名为ID</span></span><br><span class="line">data_tcga<span class="variable">$ID</span> &lt;- as.character(data_tcga<span class="variable">$ID</span>) <span class="comment"># 将data_tcga文件中ID列从数值类型数据转化为字符串类型数据 </span></span><br><span class="line">gene_symbol<span class="variable">$ID</span> &lt;- as.character(gene_symbol<span class="variable">$ID</span>) <span class="comment"># 将gene_symbol文件中ID列从数值类型数据转化为字符串类型数据</span></span><br><span class="line"></span><br><span class="line">data_tcga &lt;- inner_join(data_tcga, gene_symbol, by = <span class="string">&quot;ID&quot;</span>) <span class="comment"># 将data_tcga文件和gene_symbol文件根据ID列进行合并（这样就能获得ID对应的基因名，但是都排在最后）</span></span><br><span class="line">data_tcga &lt;- dplyr::select(data_tcga, -ID) <span class="comment"># 删除ID列</span></span><br><span class="line">data_tcga &lt;- dplyr::select(data_tcga, symbol, everything()) <span class="comment"># 重新排列,将最后一列的基因名放到第一列（如果不加everything只会选择symbol一列）</span></span><br><span class="line">data_tcga &lt;- mutate(data_tcga, rowMean=rowMeans(data_tcga[grep(<span class="string">&#x27;TCGA&#x27;</span>,names(data_tcga))])) <span class="comment"># 添加一列为表达量的平均值</span></span><br><span class="line">data_tcga &lt;- arrange(data_tcga, desc(rowMean)) <span class="comment"># 把表达量的平均值从大到小排序</span></span><br><span class="line">data_tcga &lt;- distinct(data_tcga, symbol, .keep_all = T) <span class="comment"># distinct函数被用于去重，.keep_all参数表示去重后返回数据框的所有列向量</span></span><br><span class="line">data_tcga &lt;- dplyr::select(data_tcga, -rowMean) <span class="comment"># 去除rowMean这一列</span></span><br><span class="line">data_tcga &lt;- tibble::column_to_rownames(data_tcga, var = <span class="string">&quot;symbol&quot;</span>)   <span class="comment">## 把第一列变成行名并删除</span></span><br><span class="line">dim(data_tcga)</span><br></pre></td></tr></table></figure>

<p>上面的代码说白了就是一个去重加基因注释，结果如下图所示，行为基因symbol，列为样本名，但是同样每个样本后有个01A和11A这个是与癌症相关的，01-09为肿瘤，10-19为正常对照。</p>
<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252243902.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252243902.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<p>接下来根据样本最后的这个数字初步筛选癌症和对照样本</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">### 筛选癌症和对照样本。01-09为肿瘤，10-19为正常对照</span></span><br><span class="line">mete &lt;- data.frame(colnames(data_tcga)) </span><br><span class="line"><span class="keyword">for</span> (i <span class="keyword">in</span> 1:length(mete[, 1])) &#123;</span><br><span class="line">  num &lt;- as.numeric(as.character(substring(mete[i, 1], 14, 15))) <span class="comment"># 提取每行第一列中第14，15字符</span></span><br><span class="line">  <span class="keyword">if</span>(num %<span class="keyword">in</span>% <span class="built_in">seq</span>(1, 9))&#123;mete[i, 2] = <span class="string">&quot;T&quot;</span>&#125; <span class="comment"># 判断：如果提取的数字在0-9之内就在每行第二列加上T表示肿瘤样本</span></span><br><span class="line">  <span class="keyword">if</span>(num %<span class="keyword">in</span>% <span class="built_in">seq</span>(10, 29))&#123;mete[i, 2] = <span class="string">&quot;N&quot;</span>&#125; <span class="comment"># 判断：如果提取的数字在10-29之内就在每行第二列加上N表示正常对照</span></span><br><span class="line">&#125;</span><br><span class="line">colnames(mete) &lt;- c(<span class="string">&quot;id&quot;</span>, <span class="string">&quot;group&quot;</span>)</span><br><span class="line">table(mete<span class="variable">$group</span>)</span><br><span class="line">mete<span class="variable">$group</span> &lt;- as.factor(mete<span class="variable">$group</span>) <span class="comment"># 将mete中group列转为因子模式</span></span><br><span class="line">mete &lt;- subset(mete, mete<span class="variable">$group</span> == <span class="string">&quot;T&quot;</span>) <span class="comment"># subset函数，从数据框中选择出group组为T的行</span></span><br><span class="line"></span><br><span class="line">exp_tumor &lt;- data_tcga[, <span class="built_in">which</span>(colnames(data_tcga) %<span class="keyword">in</span>% mete<span class="variable">$id</span>)] <span class="comment"># 从大样本中选出组为T的样本</span></span><br><span class="line">exp_tumor &lt;- as.data.frame(exp_tumor)</span><br><span class="line">exp_tumor &lt;- exp_tumor[, colnames(exp_tumor) %<span class="keyword">in</span>% tcga_survival<span class="variable">$sample</span>] <span class="comment"># 进一步筛选具有生存信息的样本</span></span><br><span class="line"></span><br><span class="line">exp_control &lt;- data_tcga[, <span class="built_in">which</span>(!colnames(data_tcga) %<span class="keyword">in</span>% mete<span class="variable">$id</span>)] <span class="comment"># 反向从大样本中选出组为N的样本</span></span><br><span class="line">exp_control &lt;- as.data.frame(exp_control)</span><br><span class="line">exp_control &lt;- exp_control[, colnames(exp_control) %<span class="keyword">in</span>% tcga_survival<span class="variable">$sample</span>] <span class="comment"># 进一步筛选具有生存信息的样本</span></span><br><span class="line"></span><br><span class="line">data_final &lt;- cbind(exp_control, exp_tumor) <span class="comment"># 将肿瘤样本文件和正常样本文件合并</span></span><br><span class="line">data_final &lt;- na.omit(data_final) <span class="comment"># 去除含有NA的行</span></span><br></pre></td></tr></table></figure>

<p>虽然已经根据样本ID可以区分肿瘤样本和正常样本，但是为了确保分组可靠，接下来依据前面导入的临床信息（tcga_phenotype）进一步确认分组</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">table(tcga_phenotype<span class="variable">$sample_type</span>.samples) <span class="comment"># 查看样本类型，一共四种：FFPE Scrolls，Primary Tumor，Recurrent Tumor，Solid Tissue Normal，这里面我们选择Primary Tumor——原发肿瘤和Solid Tissue Normal——正常组织</span></span><br><span class="line">table(tcga_phenotype<span class="variable">$primary_site</span>) <span class="comment"># 查看癌症发生的部位</span></span><br><span class="line"></span><br><span class="line">clinical_data &lt;- subset(tcga_phenotype, sample_type.samples == <span class="string">&#x27;Primary Tumor&#x27;</span> | sample_type.samples == <span class="string">&#x27;Solid Tissue Normal&#x27;</span>)</span><br><span class="line">clinical_data &lt;- clinical_data[clinical_data<span class="variable">$submitter_id</span>.samples %<span class="keyword">in</span>% colnames(data_final), ]</span><br><span class="line">clinical_data &lt;- clinical_data[order(clinical_data<span class="variable">$sample_type</span>.samples, decreasing = T), ]</span><br><span class="line"></span><br><span class="line">Group &lt;- data.frame(sample = clinical_data<span class="variable">$submitter_id</span>.samples, </span><br><span class="line">                    group = clinical_data<span class="variable">$sample_type</span>.samples)</span><br><span class="line">Group<span class="variable">$group</span> &lt;- ifelse(Group<span class="variable">$group</span> == <span class="string">&#x27;Solid Tissue Normal&#x27;</span>, <span class="string">&#x27;control&#x27;</span>, <span class="string">&#x27;disease&#x27;</span>)</span><br><span class="line">data_final &lt;- subset(data_final, select = Group<span class="variable">$sample</span>) <span class="comment"># 确保前面整理的count数据与整理的分组信息的样本是一致的</span></span><br><span class="line"></span><br><span class="line">table(Group<span class="variable">$group</span>)   <span class="comment">##control:52   disease:498</span></span><br></pre></td></tr></table></figure>

<p>如下图所示是整理好的分组信息，第一列是样本的ID，需要和前面整理好的count数据集中的样本ID对应，第二列为样本所属的分组</p>
<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252243320.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252243320.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<p>目前我们已经得到了整理好的count数据以及样本的分组信息，目前还差样本对应的生存信息和fpkm表达矩阵，样本的生存信息比较好处理，如下图所示为导入的样本生存信息，第一列和第三列都是样本ID，第二列OS是患者的生存状态，0表示存活，1表示死亡，第四列OS.time是患者的生存时间。</p>
<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252243836.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252243836.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<p>此时只需要去除掉第三行并且让生存信息表中患者的ID和前面整理好的分组信息表中的患者ID做个匹配即可，最后将整理好的count，group，以及生存信息表保存成csv格式即可。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">tcga_survival &lt;- tcga_survival[, -3]</span><br><span class="line">tcga_survival &lt;- tcga_survival[tcga_survival<span class="variable">$sample</span> %<span class="keyword">in</span>% Group<span class="variable">$sample</span>, ]</span><br><span class="line"></span><br><span class="line">write.csv(tcga_survival, <span class="string">&#x27;./data_survival.csv&#x27;</span>) </span><br><span class="line">write.csv(clinical_data, <span class="string">&#x27;./data_clinical.csv&#x27;</span>) </span><br><span class="line">write.csv(data_final, file = <span class="string">&#x27;./data_count.csv&#x27;</span>)</span><br><span class="line">write.csv(Group, file = <span class="string">&#x27;./data_group.csv&#x27;</span>)</span><br></pre></td></tr></table></figure>

<p>最后一步就是处理fpkm数据，这是因为在后续分析中除了差异表达分析会用到count数据，其余很多情况用的都是fpkm数据。处理的代码如下，其实整体思路和前面处理count时差不多，最后将结果保存为.csv形式即可。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">### fpkm数据整理</span></span><br><span class="line">tcga_fpkm &lt;- as.data.frame(tcga_fpkm)</span><br><span class="line">tcga_fpkm &lt;- column_to_rownames(tcga_fpkm, var = <span class="string">&#x27;Ensembl_ID&#x27;</span>)</span><br><span class="line">tcga_fpkm &lt;- 2^tcga_fpkm-1</span><br><span class="line"></span><br><span class="line">dat_fpkm &lt;- tcga_fpkm</span><br><span class="line">dat_fpkm<span class="variable">$ID</span> &lt;- rownames(dat_fpkm)</span><br><span class="line">dat_fpkm<span class="variable">$ID</span> &lt;- as.character(dat_fpkm<span class="variable">$ID</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">#gene_symbol$Eesembl_ID &lt;- as.character(gene_symbol$Eesembl_ID)</span></span><br><span class="line"></span><br><span class="line">dat_fpkm&lt;-dat_fpkm %&gt;%</span><br><span class="line">  inner_join(gene_symbol,by=<span class="string">&#x27;ID&#x27;</span>)%&gt;% </span><br><span class="line">  select(-ID)%&gt;%     <span class="comment">## 去除多余信息</span></span><br><span class="line">  select(symbol,everything())%&gt;%     <span class="comment">## 重新排列</span></span><br><span class="line">  mutate(rowMean=rowMeans(.[grep(<span class="string">&#x27;TCGA&#x27;</span>,names(.))]))%&gt;%    <span class="comment">## 求出平均数</span></span><br><span class="line">  arrange(desc(rowMean))%&gt;%       <span class="comment">## 把表达量的平均值从大到小排序</span></span><br><span class="line">  distinct(symbol,.keep_all = T)%&gt;%      <span class="comment">## symbol留下第一个</span></span><br><span class="line">  select(-rowMean)%&gt;%     <span class="comment">## 反向选择去除rowMean这一列</span></span><br><span class="line">  tibble::column_to_rownames(colnames(.)[1])   <span class="comment">## 把第一列变成行名并删除</span></span><br><span class="line">dim(dat_fpkm)</span><br><span class="line"><span class="comment"># dat_fpkm&lt;-dat_fpkm[pcg$gene_name,]</span></span><br><span class="line">dat_fpkm &lt;- dat_fpkm[,colnames(data_final)]</span><br><span class="line">dat_fpkm &lt;- na.omit(dat_fpkm)</span><br><span class="line"></span><br><span class="line">write.csv(dat_fpkm, file = <span class="string">&#x27;./data_fpkm.csv&#x27;</span>)</span><br></pre></td></tr></table></figure>

<p>下课！！！</p>
<p>最后我们提取到36829个基因与原文基本类似（原文处理了一些坏数据导致比原文稍多34102）</p>
<h3 id="c-差异表达基因筛选-x2F-火山图-edgeR-x2F-热图-gplots"><a href="#c-差异表达基因筛选-x2F-火山图-edgeR-x2F-热图-gplots" class="headerlink" title="c 差异表达基因筛选&#x2F;火山图(edgeR)&#x2F;热图(gplots)"></a>c 差异表达基因筛选&#x2F;火山图(edgeR)&#x2F;热图(gplots)</h3><p>教程来源<a target="_blank" rel="noopener" href="https://www.jianshu.com/p/0e1ad0cc4ce6">https://www.jianshu.com/p/0e1ad0cc4ce6</a></p>
<p>安装包</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">BiocManager::install(<span class="string">&#x27;edgeR&#x27;</span>)</span><br><span class="line">library(edgeR)</span><br></pre></td></tr></table></figure>

<p>edgeR使用经验贝叶斯估计和基于负二项模型的精确检验来确定差异基因，通过在基因之间来调节跨基因的过度离散程度，使用类似于Fisher精确检验但适应过度分散数据的精确检验用于评估每个基因的差异表达。</p>
<p>以下是edgeR分析差异表达基因的一般过程。</p>
<p>把处理后的值进行分析，并且依照我们的Group进行分组</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#读取基因表达矩阵</span></span><br><span class="line">data_final</span><br><span class="line"><span class="comment">#Group我们先前的分组</span></span><br><span class="line">fdr=0.05 <span class="comment"># 指定调整后p的阈值</span></span><br><span class="line">outname=<span class="string">&quot;outfile&quot;</span></span><br><span class="line"><span class="comment">#数据预处理</span></span><br><span class="line">group &lt;- rep(c(rep(<span class="string">&#x27;control&#x27;</span>,52),rep(<span class="string">&#x27;disease&#x27;</span>,498)))</span><br><span class="line"><span class="comment">#我们把处理好的数据和分组情况一起导入</span></span><br><span class="line"><span class="comment">#（1）构建 DGEList 对象</span></span><br><span class="line">dgelist &lt;- DGEList(counts = data_final,group=group)</span><br><span class="line"></span><br><span class="line"><span class="comment">#（2）过滤 low count 数据，例如 CPM 标准化（推荐）</span></span><br><span class="line">keep &lt;- rowSums(cpm(dgelist) &gt; 1 ) &gt;= 2</span><br><span class="line">dgelist &lt;- dgelist[keep, , keep.lib.sizes = FALSE]</span><br><span class="line"></span><br><span class="line"><span class="comment">#（3）标准化，以 TMM 标准化为例</span></span><br><span class="line">dgelist_norm &lt;- calcNormFactors(dgelist, method = <span class="string">&#x27;TMM&#x27;</span>)</span><br><span class="line"><span class="comment">#差异表达基因分析</span></span><br><span class="line">design &lt;- model.matrix(~group)</span><br><span class="line"><span class="comment">#（1）估算基因表达值的离散度</span></span><br><span class="line">dge &lt;- estimateDisp(dgelist_norm, design, robust = TRUE)</span><br><span class="line"><span class="comment">#（2）模型拟合，edgeR 提供了多种拟合算法</span></span><br><span class="line">1. )</span><br><span class="line">   <span class="comment">#负二项广义对数线性模型</span></span><br><span class="line">   fit &lt;- glmFit(dge, design, robust = TRUE)</span><br><span class="line">   lrt &lt;- topTags(glmLRT(fit), n = nrow(dgelist<span class="variable">$counts</span>))</span><br><span class="line">2. )</span><br><span class="line">   mode=<span class="string">&#x27;glmQLFTest&#x27;</span></span><br><span class="line">   <span class="comment">#拟似然负二项广义对数线性模型</span></span><br><span class="line">   fit &lt;- glmQLFit(dge, design, robust = TRUE)</span><br><span class="line">   lrt &lt;- topTags(glmQLFTest(fit), n = nrow(dgelist<span class="variable">$counts</span>))</span><br><span class="line"><span class="comment">#排序并标出UP/down</span></span><br><span class="line">lrt&lt;-lrt<span class="variable">$table</span></span><br><span class="line">gene_diff &lt;- lrt[order(lrt<span class="variable">$FDR</span>, lrt<span class="variable">$logFC</span>, decreasing = c(FALSE, TRUE)), ]</span><br><span class="line">gene_diff[<span class="built_in">which</span>(gene_diff<span class="variable">$logFC</span> &gt;= 1 &amp; gene_diff<span class="variable">$FDR</span> &lt; fdr),<span class="string">&#x27;sig&#x27;</span>] &lt;- <span class="string">&#x27;up&#x27;</span></span><br><span class="line">gene_diff[<span class="built_in">which</span>(gene_diff<span class="variable">$logFC</span> &lt;= -1 &amp; gene_diff<span class="variable">$FDR</span> &lt; fdr),<span class="string">&#x27;sig&#x27;</span>] &lt;- <span class="string">&#x27;down&#x27;</span></span><br><span class="line">gene_diff[<span class="built_in">which</span>(abs(gene_diff<span class="variable">$logFC</span>) &lt;= 1 | gene_diff<span class="variable">$FDR</span> &gt;= fdr),<span class="string">&#x27;sig&#x27;</span>] &lt;- <span class="string">&#x27;non-significant&#x27;</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#输出总表</span></span><br><span class="line">write.table(gene_diff, <span class="built_in">paste</span>(outname,<span class="string">&#x27;.txt&#x27;</span>,sep=<span class="string">&quot;&quot;</span>), sep = <span class="string">&#x27;\t&#x27;</span>, col.names = NA, quote = FALSE)      </span><br><span class="line"><span class="comment">#输出选择的差异基因总表</span></span><br><span class="line">gene_diff_select &lt;- subset(gene_diff, sig %<span class="keyword">in</span>% c(<span class="string">&#x27;up&#x27;</span>, <span class="string">&#x27;down&#x27;</span>))</span><br><span class="line">write.table(gene_diff_select, file = <span class="built_in">paste</span>(outname,<span class="string">&#x27;.select.txt&#x27;</span>,sep=<span class="string">&quot;&quot;</span>), sep = <span class="string">&#x27;\t&#x27;</span>, col.names = NA, quote = FALSE)</span><br></pre></td></tr></table></figure>

<p>上述便可得到edgeR差异分析的基本结果了。下面可以使用ggplot绘制相应的火山图。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">outname = <span class="string">&quot;edgeRdiff&quot;</span> <span class="comment">#指定输出名字</span></span><br><span class="line"><span class="comment">##ggplot2 差异火山图</span></span><br><span class="line">library(ggplot2)</span><br><span class="line">data_res1=as.data.frame(gene_diff)</span><br><span class="line"></span><br><span class="line"><span class="comment">#默认情况下，横轴展示 log2FoldChange，纵轴展示 -log10 转化后的 padj</span></span><br><span class="line">p &lt;- ggplot(data = gene_diff, aes(x = logFC, y = -log10(FDR), color = sig)) +</span><br><span class="line">geom_point(size = 1) +  <span class="comment">#绘制散点图</span></span><br><span class="line">scale_color_manual(values = c(<span class="string">&#x27;red&#x27;</span>, <span class="string">&#x27;gray&#x27;</span>, <span class="string">&#x27;green&#x27;</span>), limits = c(<span class="string">&#x27;up&#x27;</span>, <span class="string">&#x27;non-significant&#x27;</span>, <span class="string">&#x27;down&#x27;</span>)) +  <span class="comment">#自定义点的颜色</span></span><br><span class="line">labs(x = <span class="string">&#x27;log2 Fold Change&#x27;</span>, y = <span class="string">&#x27;-log10 adjust p-value&#x27;</span>, title = <span class="string">&#x27;&#x27;</span>, color = <span class="string">&#x27;&#x27;</span>) +  <span class="comment">#坐标轴标题</span></span><br><span class="line">theme(plot.title = element_text(hjust = 0.5, size = 14), panel.grid = element_blank(), <span class="comment">#背景色、网格线&gt;、图例等主题修改</span></span><br><span class="line">   panel.background = element_rect(color = <span class="string">&#x27;black&#x27;</span>, fill = <span class="string">&#x27;transparent&#x27;</span>),</span><br><span class="line">   legend.key = element_rect(fill = <span class="string">&#x27;transparent&#x27;</span>)) +</span><br><span class="line">geom_vline(xintercept = c(-1, 1), lty = 3, color = <span class="string">&#x27;black&#x27;</span>) +  <span class="comment">#添加阈值线</span></span><br><span class="line">geom_hline(yintercept = -log10(fdr), lty = 3, color = <span class="string">&#x27;black&#x27;</span>) +</span><br><span class="line"><span class="comment">#xlim(-12, 12) + ylim(0, 35)</span></span><br><span class="line">pdf(<span class="built_in">paste</span>(outname, <span class="string">&quot;.volcano.pdf&quot;</span>,sep=<span class="string">&quot;&quot;</span>))</span><br><span class="line">p</span><br><span class="line">dev.off()</span><br></pre></td></tr></table></figure>

<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252243407.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252243407.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line">rt_sam_m &lt;- data.frame(cbind(colnames(data_final), c(rep(<span class="string">&#x27;control&#x27;</span>, 52),rep(<span class="string">&#x27;disease&#x27;</span>,498))), stringsAsFactors = FALSE)</span><br><span class="line">colnames(rt_sam_m) &lt;- c(<span class="string">&#x27;samples_id&#x27;</span>, <span class="string">&#x27;group&#x27;</span>)</span><br><span class="line">m_group = <span class="built_in">factor</span>(rt_sam_m<span class="variable">$group</span>, levels=c(<span class="string">&#x27;control&#x27;</span>, <span class="string">&#x27;disease&#x27;</span>))</span><br><span class="line"></span><br><span class="line">design_m = model.matrix(~0 + m_group)</span><br><span class="line">row.names(design_m) &lt;- rt_sam_m<span class="variable">$samples_id</span></span><br><span class="line">colnames(design_m) &lt;- c(<span class="string">&#x27;control&#x27;</span>,<span class="string">&#x27;disease&#x27;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 差异分析</span></span><br><span class="line">m_fit &lt;- lmFit(data_final, design_m)</span><br><span class="line">cont.matrix &lt;- makeContrasts(ControlDisease = control-disease, levels = design_m)</span><br><span class="line">m_fit2 &lt;- contrasts.fit(m_fit, cont.matrix)</span><br><span class="line">m_fit3 &lt;- eBayes(m_fit2)</span><br><span class="line">rt_diff &lt;- topTreat(m_fit3, number = length(row.names(data_final)))</span><br><span class="line">rt_diff2 &lt;- rt_diff[<span class="built_in">which</span>(abs(rt_diff<span class="variable">$logFC</span>) &gt; 1 &amp; rt_diff<span class="variable">$P</span>.Value &lt; 0.05), ]</span><br><span class="line">dim(rt_diff2)<span class="comment">#1677 6, 有1683个差异基因</span></span><br><span class="line"></span><br><span class="line">library(gplots)</span><br><span class="line">rt_ht &lt;- data_final[row.names(rt_diff2)[order(abs(rt_diff2<span class="variable">$logFC</span>), decreasing = TRUE)], ]</span><br><span class="line">mat_ht &lt;- as.matrix(apply(rt_ht, 2, <span class="keyword">function</span>(x)&#123;as.numeric(x)&#125;))</span><br><span class="line">mat_ht_z &lt;- apply(mat_ht, 2, FUN = <span class="keyword">function</span>(x)&#123;(x-median(x))/sd(x)&#125;)</span><br><span class="line">row.names(mat_ht_z) &lt;- row.names(rt_ht)</span><br><span class="line">heatmap.cols &lt;- c(rep(<span class="string">&#x27;blue&#x27;</span>, 52), rep(<span class="string">&#x27;red&#x27;</span>, 498))</span><br><span class="line"><span class="comment">#打开PDF</span></span><br><span class="line">pdf(<span class="built_in">paste</span>(<span class="string">&quot;heatmap.pdf&quot;</span>,sep=<span class="string">&quot;&quot;</span>))</span><br><span class="line"><span class="comment">#开始画图，并存储到PDF里面</span></span><br><span class="line">heatmap.2(mat_ht_z, col = colorpanel(99, <span class="string">&quot;blue&quot;</span>, <span class="string">&quot;black&quot;</span>, <span class="string">&quot;red&quot;</span>), dendrogram = <span class="string">&quot;both&quot;</span>, keysize = 1,</span><br><span class="line">         hclustfun = <span class="keyword">function</span>(x)&#123;hclust(x, method = <span class="string">&#x27;ward.D2&#x27;</span>)&#125;, trace = <span class="string">&quot;none&quot;</span>, density.info = <span class="string">&quot;none&quot;</span>,</span><br><span class="line">				 ColSideColors = heatmap.cols)</span><br><span class="line"><span class="comment">#pdf(&quot;heatmap.pdf&quot;,width=10,height = 10)</span></span><br><span class="line">dev.off()</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<h1 id="二、WGCNA网络搭建"><a href="#二、WGCNA网络搭建" class="headerlink" title="二、WGCNA网络搭建"></a>二、WGCNA网络搭建</h1><p>参考文章：<a target="_blank" rel="noopener" href="https://rstudio-pubs-static.s3.amazonaws.com/554864_6e0f44cdc1474b8ba2c5f8326559bbe4.html">https://rstudio-pubs-static.s3.amazonaws.com/554864_6e0f44cdc1474b8ba2c5f8326559bbe4.html</a></p>
<h2 id="1-基本概念"><a href="#1-基本概念" class="headerlink" title="1.基本概念"></a>1.基本概念</h2><p>加权基因共表达网络分析 (WGCNA, Weighted correlation network analysis)是用来描述不同样品之间基因关联模式的系统生物学方法，可以用来鉴定高度<strong>协同变化</strong>的基因集, 并根据基因集的内连性和基因集与表型之间的关联鉴定候补生物标记基因或治疗靶点。</p>
<p>相比于只关注差异表达的基因，WGCNA利用数千或近万个变化最大的基因或全部基因的信息识别感兴趣的基因集，并与表型进行显著性关联分析。一是充分利用了信息， 二是把数千个基因与表型的关联转换为数个基因集与表型的关联，免去了多重假设检验校正的问题。</p>
<p>理解WGCNA，需要先理解下面几个术语和它们在WGCNA中的定义。</p>
<ul>
<li><p>共表达网络：定义为加权基因网络。点代表基因，边代表基因表达相关性。加权是指对<strong>相关性值进行冥次运算</strong> (冥次的值也就是<strong>软阈值</strong> (power, pickSoftThreshold这个函数所做的就是确定合适的power))。无向网络的边属性计算方式为 <code>abs(cor(genex, geney)) ^ power</code>；有向网络的边属性计算方式为 <code>(1+cor(genex, geney)/2) ^ power</code>; sign hybrid的边属性计算方式为<code>cor(genex, geney)^power if cor&gt;0 else 0</code>。这种处理方式强化了强相关，弱化了弱相关或负相关，使得相关性数值更符合<code>无标度网络</code>特征，更具有生物意义。如果没有合适的<strong>power</strong>，一般是由于部分样品与其它样品因为某种原因差别太大导致的，可根据具体问题移除部分 样品或查看后面的<code>经验值</code>。</p>
</li>
<li><p>Module(模块)：高度內连的基因集。在无向网络中，模块内是高度<strong>相关</strong>的基因。在有向网络中，模块内是高度<strong>正相关</strong>的基因。把基因聚类成模块后，可以对每个模块进行三个层次的分析：<code>1</code>. 功能富集分析查看其功能特征是否与研究目的相符；<code>2</code>. 模块与性状进行关联分析，找出与关注性状相关度最高的模块；<code>3</code>. 模块 与样本进行关联分析，找到样品特异高表达的模块。</p>
<p>  基因富集相关文章 <a target="_blank" rel="noopener" href="https://mp.weixin.qq.com/s/l6j2encDfEQkt2UeNCMFhg">去东方，最好用的在线GO富集分析工具</a>；<a target="_blank" rel="noopener" href="http://mp.weixin.qq.com/s/d1KCETQZ88yaOLGwAtpWYg">GO、GSEA富集分析一网打进</a>；<a target="_blank" rel="noopener" href="http://mp.weixin.qq.com/s/3Nd3urhfRGkw-F0LGZrlZQ">GSEA富集分析 - 界面操作</a>。其它关联后面都会提及。</p>
</li>
<li><p>Connectivity (连接度)：类似于网络中 “度” (degree)的概念。每个基因的连接度是与其相连的基因的<code>边属性之和</code>。</p>
</li>
<li><p>Module eigengene E: 给定模型的第一主成分，代表<strong>整个模型的基因表达谱</strong>。这个是个很巧妙的梳理，我们之前讲过<a target="_blank" rel="noopener" href="https://mp.weixin.qq.com/s/ZKvQieq_6KX6l6LZyUz7jA">PCA分析</a>的降维作用，之前主要是拿来做可视化，现在用到这个地方，很好的用一个向量代替了一个矩阵，方便后期计算。(降维除了PCA，还可以看看<a target="_blank" rel="noopener" href="http://mp.weixin.qq.com/s/alBfj3Y08qCnZoz5JwVdaw">tSNE</a>)</p>
</li>
<li><p>Intramodular connectivity: 给定基因与给定模型内其他基因的关联度，判断基因所属关系。</p>
</li>
<li><p>Module membership: 给定基因表达谱与给定模型的eigengene的相关性。</p>
</li>
<li><p>Hub gene: 关键基因 (连接度最多或连接多个模块的基因)。</p>
</li>
<li><p>Adjacency matrix (邻接矩阵)：基因和基因之间的加权相关性值构成的矩阵。</p>
</li>
<li><p>TOM (Topological overlap matrix)：把邻接矩阵转换为拓扑重叠矩阵，以降低噪音和假相关，获得的新距离矩阵，这个信息可拿来构建网络或绘制TOM图。其定义依 据是任何两个基因的相关性不近由它们自己的相关性决定，还依赖于与这两个基因存在相关性的其它基因的互作，把这些因素都考虑进来，才能更好地定义基因表达谱的相似性。</p>
</li>
</ul>
<p>$$<br>w_{i,j}&#x3D;\frac{\sum_u (a_{iu}a_{uj})+a_{ij}}{min(\sum_u a_{iu},\sum_u a_{ju})+1-a_{ij}}<br>$$</p>
<h2 id="2基本分析流程"><a href="#2基本分析流程" class="headerlink" title="2基本分析流程"></a><strong>2基本分析流程</strong></h2><ol>
<li>构建基因共表达网络：使用加权的表达相关性。</li>
<li>识别基因集：基于加权相关性，进行层级聚类分析，并根据设定标准切分聚类结果，获得不同的基因模块，用聚类树的分枝和不同颜色表示。</li>
<li>如果有表型信息，计算基因模块与表型的相关性，鉴定性状相关的模块。</li>
<li>研究模型之间的关系，从系统层面查看不同模型的互作网络。</li>
<li>从关键模型中选择感兴趣的驱动基因，或根据模型中已知基因的功能推测未知基因的功能。</li>
<li>导出TOM矩阵，绘制相关性图。</li>
</ol>
<h2 id="3-WGCNA包实战"><a href="#3-WGCNA包实战" class="headerlink" title="3 WGCNA包实战"></a>3 <strong>WGCNA包实战</strong></h2><p>R包<code>WGCNA</code>是用于计算各种加权关联分析的功能集合，可用于网络构建，基因筛选，基因簇鉴定，拓扑特征计算，数据模拟和可视化等。</p>
<h3 id="a-输入数据和参数选择"><a href="#a-输入数据和参数选择" class="headerlink" title="a. 输入数据和参数选择"></a><strong>a. 输入数据和参数选择</strong></h3><ol>
<li>WGCNA本质是基于相关系数的网络分析方法，适用于多样品数据模式，一般要求样本数多于<strong>15</strong>个。样本数多于<strong>20</strong>时效果更好，样本越多，结果越稳定。</li>
<li>基因表达矩阵: 常规表达矩阵即可，即基因在行，样品在列，进入分析前做一个转置。RPKM、FPKM或其它标准化方法影响不大，推荐使用Deseq2的<code>varianceStabilizingTransformation</code>或<code>log2(x+1)</code>对标准化后的数据做个转换。如果数据来自不同的批次，需要先移除批次效应 (记得上次转录组培训课讲过如何操作)。如果数据存在系统偏移，需要做下<code>quantile normalization</code>。</li>
<li>性状矩阵：用于关联分析的性状必须是<strong>数值型</strong>特征 (如下面示例中的<code>Height</code>, <code>Weight</code>, <code>Diameter</code>)。如果是区域或分类变量，需要转换为<strong>0-1</strong>矩阵的形 式(1表示属于此组或有此属性，0表示不属于此组或无此属性，如样品分组信息WT, KO, OE)。</li>
<li>推荐使用<code>Signed network</code>和<code>Robust correlation (bicor)</code>。(这个根据自己的需要，看看上面写的每个网络怎么计算的，更知道怎么选择)</li>
<li>无向网络在power小于<code>15</code>或有向网络power小于<code>30</code>内，没有一个<strong>power</strong>值可以使无标度网络图谱结构<strong>R^2达到0.8</strong>或平均连接度降到<strong>100</strong>以下，可能是由 于部分样品与其他样品差别太大造成的。这可能由<code>批次效应</code>、<code>样品异质性</code>或<code>实验条件对表达影响太大</code>等造成, 可以通过绘制样品聚类查看分组信息、关联批次信息、处理信息和有无异常样品 (可以使用之前讲过的<a target="_blank" rel="noopener" href="https://mp.weixin.qq.com/s/_9LKs6t6rcjzokF_0gneSA">热图简化，增加行或列属性</a>)。如果这确实是由有意义的生物变化引起的，也可以使用后面程序中的经验<strong>power</strong>值。</li>
</ol>
<table>
<thead>
<tr>
<th>ID</th>
<th>WT</th>
<th>KO</th>
<th>OE</th>
<th>Height</th>
<th>Weight</th>
<th>Diameter</th>
</tr>
</thead>
<tbody><tr>
<td>samp1</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>2</td>
<td>3</td>
</tr>
<tr>
<td>samp1</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>2</td>
<td>4</td>
<td>6</td>
</tr>
<tr>
<td>samp1</td>
<td>0</td>
<td>1</td>
<td>0</td>
<td>10</td>
<td>20</td>
<td>50</td>
</tr>
<tr>
<td>samp1</td>
<td>0</td>
<td>1</td>
<td>1</td>
<td>15</td>
<td>30</td>
<td>80</td>
</tr>
<tr>
<td>samp1</td>
<td>0</td>
<td>1</td>
<td>1</td>
<td>NA</td>
<td>9</td>
<td>8</td>
</tr>
</tbody></table>
<h3 id="b-安装导入依赖的包及数据"><a href="#b-安装导入依赖的包及数据" class="headerlink" title="b 安装导入依赖的包及数据"></a>b <strong>安装导入依赖的包及数据</strong></h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line">library(<span class="string">&quot;BiocManager&quot;</span>)</span><br><span class="line">install(c(<span class="string">&quot;AnnotationDbi&quot;</span>, <span class="string">&quot;impute&quot;</span>,<span class="string">&quot;GO.db&quot;</span>, <span class="string">&quot;preprocessCore&quot;</span>))</span><br><span class="line">site=<span class="string">&quot;https://mirrors.tuna.tsinghua.edu.cn/CRAN&quot;</span></span><br><span class="line">install.packages(c(<span class="string">&quot;WGCNA&quot;</span>, <span class="string">&quot;stringr&quot;</span>, <span class="string">&quot;reshape2&quot;</span>, <span class="string">&quot;pheatmap&quot;</span>), repos=site)</span><br><span class="line">library(WGCNA)</span><br><span class="line">library(reshape2)</span><br><span class="line">library(stringr)</span><br><span class="line">options(stringsAsFactors = FALSE)</span><br><span class="line"><span class="comment"># 打开多线程</span></span><br><span class="line">enableWGCNAThreads()</span><br><span class="line"><span class="built_in">type</span> = <span class="string">&quot;unsigned&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 相关性计算</span></span><br><span class="line"><span class="comment"># 官方推荐 biweight mid-correlation &amp; bicor</span></span><br><span class="line"><span class="comment"># corType: pearson or bicor</span></span><br><span class="line">corType = <span class="string">&quot;bicor&quot;</span></span><br><span class="line"></span><br><span class="line">corFnc = ifelse(corType==<span class="string">&quot;pearson&quot;</span>, cor, bicor)</span><br><span class="line"><span class="comment"># 对二元变量，如样本性状信息计算相关性时，</span></span><br><span class="line"><span class="comment"># 或基因表达严重依赖于疾病状态时，需设置下面参数</span></span><br><span class="line">maxPOutliers = ifelse(corType==<span class="string">&quot;pearson&quot;</span>,1,0.05)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 关联样品性状的二元变量时，设置</span></span><br><span class="line">robustY = ifelse(corType==<span class="string">&quot;pearson&quot;</span>,T,F)</span><br><span class="line"></span><br><span class="line"><span class="comment">##导入数据##</span></span><br><span class="line"><span class="comment">#dataExpr &lt;- read.table(exprMat, sep=&#x27;\t&#x27;, row.names=1, header=T, </span></span><br><span class="line"><span class="comment">#                     quote=&quot;&quot;, comment=&quot;&quot;, check.names=F)</span></span><br><span class="line"></span><br><span class="line">dat_fpkm &lt;- read_tsv(file = <span class="string">&#x27;./data_fpkm.csv&#x27;</span>)</span><br><span class="line">dim(dat_fpkm)</span><br><span class="line"><span class="comment">#[1] 36829   550</span></span><br></pre></td></tr></table></figure>

<p>规范数据如图所示</p>
<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252243305.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252243305.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<h3 id="c-数据基本处理"><a href="#c-数据基本处理" class="headerlink" title="c 数据基本处理"></a>c 数据基本处理</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## ---- eval=T-------------------------------------------------------------</span></span><br><span class="line"><span class="comment">## 筛选中位绝对偏差前75%的基因，至少MAD大于0.01</span></span><br><span class="line"><span class="comment">## 筛选后会降低运算量，也会失去部分信息</span></span><br><span class="line"><span class="comment">## 也可不做筛选，使MAD大于0即可</span></span><br><span class="line">m.mad &lt;- apply(dat_fpkm ,1,mad)</span><br><span class="line">dataExprVar &lt;- dat_fpkm [<span class="built_in">which</span>(m.mad &gt; </span><br><span class="line">                 max(quantile(m.mad, probs=<span class="built_in">seq</span>(0, 1, 0.25))[2],0.01)),]</span><br><span class="line"></span><br><span class="line"><span class="comment">## 转换为样品在行，基因在列的矩阵</span></span><br><span class="line">dataExpr &lt;- as.data.frame(t(dataExprVar))</span><br><span class="line"></span><br><span class="line"><span class="comment">## 检测缺失值</span></span><br><span class="line">gsg = goodSamplesGenes(dat_fpkm , verbose = 3)</span><br><span class="line"><span class="comment">##  Flagging genes and samples with too many missing values...</span></span><br><span class="line"><span class="comment">##   ..step 1</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> (!gsg<span class="variable">$allOK</span>)&#123;</span><br><span class="line">  <span class="comment"># Optionally, print the gene and sample names that were removed:</span></span><br><span class="line">  <span class="keyword">if</span> (<span class="built_in">sum</span>(!gsg<span class="variable">$goodGenes</span>)&gt;0) </span><br><span class="line">    printFlush(<span class="built_in">paste</span>(<span class="string">&quot;Removing genes:&quot;</span>, </span><br><span class="line">                     <span class="built_in">paste</span>(names(dataExpr)[!gsg<span class="variable">$goodGenes</span>], collapse = <span class="string">&quot;,&quot;</span>)));</span><br><span class="line">  <span class="keyword">if</span> (<span class="built_in">sum</span>(!gsg<span class="variable">$goodSamples</span>)&gt;0) </span><br><span class="line">    printFlush(<span class="built_in">paste</span>(<span class="string">&quot;Removing samples:&quot;</span>, </span><br><span class="line">                     <span class="built_in">paste</span>(rownames(dataExpr)[!gsg<span class="variable">$goodSamples</span>], collapse = <span class="string">&quot;,&quot;</span>)));</span><br><span class="line">  <span class="comment"># Remove the offending genes and samples from the data:</span></span><br><span class="line">  dataExpr = dataExpr[gsg<span class="variable">$goodSamples</span>, gsg<span class="variable">$goodGenes</span>]</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">nGenes = ncol(dataExpr)</span><br><span class="line">nSamples = nrow(dataExpr)</span><br><span class="line"></span><br><span class="line">dim(dataExpr)</span><br><span class="line"><span class="comment">#[1]   550 13755</span></span><br></pre></td></tr></table></figure>

<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252243592.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252243592.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<h3 id="d-样本层级聚类，查看有无离群值"><a href="#d-样本层级聚类，查看有无离群值" class="headerlink" title="d 样本层级聚类，查看有无离群值"></a>d <strong>样本层级聚类，查看有无离群值</strong></h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## ---- echo=T, fig.cap=&quot;查看是否有离群样品&quot;, fig.width=20-----------------</span></span><br><span class="line"><span class="comment">## 查看是否有离群样品</span></span><br><span class="line">sampleTree = hclust(dist(dataExpr), method = <span class="string">&quot;average&quot;</span>)</span><br><span class="line"><span class="comment">#pdf(paste(&quot;LiverFemaleClean.pdf&quot;,sep=&quot;&quot;))</span></span><br><span class="line">pdf(file=<span class="string">&quot;LiverFemaleClean.txt.pdf&quot;</span>, onefile=F, paper=<span class="string">&quot;special&quot;</span>, </span><br><span class="line">    width=80, height=14, <span class="built_in">bg</span>=<span class="string">&quot;white&quot;</span>, pointsize=6)</span><br><span class="line">plot(sampleTree, main = <span class="string">&quot;Sample clustering to detect outliers&quot;</span>, sub=<span class="string">&quot;&quot;</span>, xlab=<span class="string">&quot;&quot;</span>)</span><br><span class="line">dev.off()</span><br><span class="line"><span class="comment">#RStudioGD </span></span><br><span class="line"><span class="comment">#       2 </span></span><br></pre></td></tr></table></figure>

<p>(部分,原图太大了)</p>
<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252243449.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252243449.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># sample network based on squared Euclidean distance note that we</span></span><br><span class="line"><span class="comment"># transpose the data</span></span><br><span class="line">A = adjacency(t(dataExpr), <span class="built_in">type</span> = <span class="string">&quot;distance&quot;</span>)</span><br><span class="line"><span class="comment"># this calculates the whole network connectivity</span></span><br><span class="line">k = as.numeric(apply(A, 2, <span class="built_in">sum</span>)) - 1</span><br><span class="line"><span class="comment"># standardized connectivity</span></span><br><span class="line">Z.k = scale(k)</span><br><span class="line"><span class="comment"># Designate samples as outlying if their Z.k value is below the threshold</span></span><br><span class="line">thresholdZ.k = -5  <span class="comment"># often -2.5</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># the color vector indicates outlyingness (red)</span></span><br><span class="line">outlierColor = ifelse(Z.k &lt; thresholdZ.k, <span class="string">&quot;red&quot;</span>, <span class="string">&quot;black&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># calculate the cluster tree using flahsClust or hclust</span></span><br><span class="line">sampleTree = hclust(as.dist(1 - A), method = <span class="string">&quot;average&quot;</span>)</span><br><span class="line"><span class="comment"># Convert traits to a color representation: where red indicates high</span></span><br><span class="line"><span class="comment"># values</span></span><br><span class="line"><span class="comment"># traitColors = data.frame(numbers2colors(datTraits, signed = FALSE))</span></span><br><span class="line"><span class="comment"># dimnames(traitColors)[[2]] = paste(names(datTraits), &quot;C&quot;, sep = &quot;&quot;)</span></span><br><span class="line"><span class="comment"># datColors = data.frame(outlierC = outlierColor, traitColors)</span></span><br><span class="line"><span class="comment"># Plot the sample dendrogram and the colors underneath.</span></span><br><span class="line">plotDendroAndColors(sampleTree, groupLabels = names(outlierColor), </span><br><span class="line">                    colors = outlierColor, </span><br><span class="line">                    main = <span class="string">&quot;Sample dendrogram and trait heatmap&quot;</span>)</span><br><span class="line">pdf(file = <span class="string">&quot;plotDendroAndColorsAndTree.pdf&quot;</span>,width=80, height=14)</span><br><span class="line">plotDendroAndColors(sampleTree, groupLabels = names(outlierColor), </span><br><span class="line">                    colors = outlierColor, </span><br><span class="line">                    main = <span class="string">&quot;Sample dendrogram and trait heatmap&quot;</span>)</span><br><span class="line">dev.off()</span><br></pre></td></tr></table></figure>

<p>(部分)</p>
<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252243250.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252243250.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<h3 id="e-确定软阈值"><a href="#e-确定软阈值" class="headerlink" title="e 确定软阈值"></a>e <strong>确定软阈值</strong></h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## ---- echo=T-------------------------------------------------------------</span></span><br><span class="line"><span class="comment">## 软阈值筛选</span></span><br><span class="line"><span class="comment">## 软阈值的筛选原则是使构建的网络更符合无标度网络特征。</span></span><br><span class="line">powers = c(c(1:10), <span class="built_in">seq</span>(from = 12, to=30, by=2))</span><br><span class="line">sft = pickSoftThreshold(dataExpr, powerVector=powers, </span><br><span class="line">                        networkType=<span class="built_in">type</span>, verbose=5)</span><br></pre></td></tr></table></figure>

<p><img src="/%E9%80%9A%E8%BF%87%E5%85%B1%E8%A1%A8%E8%BE%BE%E7%BD%91%E7%BB%9C%E5%88%86%E6%9E%90%EF%BC%88WGCNA%EF%BC%89%205201c79a5b3e49c5a4f38f90d7c724ba/Untitled%2019.png" class="lazyload" data-srcset="/%E9%80%9A%E8%BF%87%E5%85%B1%E8%A1%A8%E8%BE%BE%E7%BD%91%E7%BB%9C%E5%88%86%E6%9E%90%EF%BC%88WGCNA%EF%BC%89%205201c79a5b3e49c5a4f38f90d7c724ba/Untitled%2019.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 图展示在文档中</span></span><br><span class="line">par(mfrow = c(1,2))</span><br><span class="line">cex1 = 0.9</span><br><span class="line"><span class="comment"># 横轴是Soft threshold (power)，纵轴是无标度网络的评估参数，数值越高，</span></span><br><span class="line"><span class="comment"># 网络越符合无标度特征 (non-scale)</span></span><br><span class="line">plot(sft<span class="variable">$fitIndices</span>[,1], -sign(sft<span class="variable">$fitIndices</span>[,3])*sft<span class="variable">$fitIndices</span>[,2],</span><br><span class="line">     xlab=<span class="string">&quot;Soft Threshold (power)&quot;</span>,</span><br><span class="line">     ylab=<span class="string">&quot;Scale Free Topology Model Fit,signed R^2&quot;</span>,<span class="built_in">type</span>=<span class="string">&quot;n&quot;</span>,</span><br><span class="line">     main = <span class="built_in">paste</span>(<span class="string">&quot;Scale independence&quot;</span>))</span><br><span class="line">text(sft<span class="variable">$fitIndices</span>[,1], -sign(sft<span class="variable">$fitIndices</span>[,3])*sft<span class="variable">$fitIndices</span>[,2],</span><br><span class="line">     labels=powers,cex=cex1,col=<span class="string">&quot;red&quot;</span>)</span><br><span class="line"><span class="comment"># 筛选标准。R-square=0.85</span></span><br><span class="line">abline(h=0.85,col=<span class="string">&quot;red&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Soft threshold与平均连通性</span></span><br><span class="line">plot(sft<span class="variable">$fitIndices</span>[,1], sft<span class="variable">$fitIndices</span>[,5],</span><br><span class="line">     xlab=<span class="string">&quot;Soft Threshold (power)&quot;</span>,ylab=<span class="string">&quot;Mean Connectivity&quot;</span>, <span class="built_in">type</span>=<span class="string">&quot;n&quot;</span>,</span><br><span class="line">     main = <span class="built_in">paste</span>(<span class="string">&quot;Mean connectivity&quot;</span>))</span><br><span class="line">text(sft<span class="variable">$fitIndices</span>[,1], sft<span class="variable">$fitIndices</span>[,5], labels=powers, </span><br><span class="line">     cex=cex1, col=<span class="string">&quot;red&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># ---从pdf到dev.off() 表示把图存储为PDF文件--------------------------------------</span></span><br><span class="line"></span><br><span class="line">pdf(file=<span class="string">&quot;LiverFemaleClean.txt.softpower.pdf&quot;</span>, onefile=F, paper=<span class="string">&quot;special&quot;</span>, </span><br><span class="line">    <span class="built_in">bg</span>=<span class="string">&quot;white&quot;</span>, pointsize=6)</span><br><span class="line">par(mfrow = c(1,2))</span><br><span class="line">cex1 = 0.9</span><br><span class="line"><span class="comment"># 横轴是Soft threshold (power)，纵轴是无标度网络的评估参数，数值越高，</span></span><br><span class="line"><span class="comment"># 网络越符合无标度特征 (non-scale)</span></span><br><span class="line">plot(sft<span class="variable">$fitIndices</span>[,1], -sign(sft<span class="variable">$fitIndices</span>[,3])*sft<span class="variable">$fitIndices</span>[,2],</span><br><span class="line">     xlab=<span class="string">&quot;Soft Threshold (power)&quot;</span>,</span><br><span class="line">     ylab=<span class="string">&quot;Scale Free Topology Model Fit,signed R^2&quot;</span>,<span class="built_in">type</span>=<span class="string">&quot;n&quot;</span>,</span><br><span class="line">     main = <span class="built_in">paste</span>(<span class="string">&quot;Scale independence&quot;</span>))</span><br><span class="line">text(sft<span class="variable">$fitIndices</span>[,1], -sign(sft<span class="variable">$fitIndices</span>[,3])*sft<span class="variable">$fitIndices</span>[,2],</span><br><span class="line">     labels=powers,cex=cex1,col=<span class="string">&quot;red&quot;</span>)</span><br><span class="line"><span class="comment"># 筛选标准。R-square=0.85</span></span><br><span class="line">abline(h=0.85,col=<span class="string">&quot;red&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Soft threshold与平均连通性</span></span><br><span class="line">plot(sft<span class="variable">$fitIndices</span>[,1], sft<span class="variable">$fitIndices</span>[,5],</span><br><span class="line">     xlab=<span class="string">&quot;Soft Threshold (power)&quot;</span>,ylab=<span class="string">&quot;Mean Connectivity&quot;</span>, <span class="built_in">type</span>=<span class="string">&quot;n&quot;</span>,</span><br><span class="line">     main = <span class="built_in">paste</span>(<span class="string">&quot;Mean connectivity&quot;</span>))</span><br><span class="line">text(sft<span class="variable">$fitIndices</span>[,1], sft<span class="variable">$fitIndices</span>[,5], labels=powers, </span><br><span class="line">     cex=cex1, col=<span class="string">&quot;red&quot;</span>)</span><br><span class="line">dev.off()</span><br><span class="line"></span><br><span class="line">power = sft<span class="variable">$powerEstimate</span></span><br><span class="line">power</span><br></pre></td></tr></table></figure>

<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252243073.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252243073.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## ---- echo=T-------------------------------------------------------------</span></span><br><span class="line"><span class="comment"># 无向网络在power小于15或有向网络power小于30内，没有一个power值可以使</span></span><br><span class="line"><span class="comment"># 无标度网络图谱结构R^2达到0.8，平均连接度较高如在100以上，可能是由于</span></span><br><span class="line"><span class="comment"># 部分样品与其他样品差别太大。这可能由批次效应、样品异质性或实验条件对</span></span><br><span class="line"><span class="comment"># 表达影响太大等造成。可以通过绘制样品聚类查看分组信息和有无异常样品。</span></span><br><span class="line"><span class="comment"># 如果这确实是由有意义的生物变化引起的，也可以使用下面的经验power值。</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> (is.na(power))&#123;</span><br><span class="line">  <span class="built_in">print</span>(<span class="string">&quot;Using experience power since no suitable power found.&quot;</span>)</span><br><span class="line">  power = ifelse(nSamples&lt;20, ifelse(<span class="built_in">type</span> == <span class="string">&quot;unsigned&quot;</span>, 9, 18),</span><br><span class="line">          ifelse(nSamples&lt;30, ifelse(<span class="built_in">type</span> == <span class="string">&quot;unsigned&quot;</span>, 8, 16),</span><br><span class="line">          ifelse(nSamples&lt;40, ifelse(<span class="built_in">type</span> == <span class="string">&quot;unsigned&quot;</span>, 7, 14),</span><br><span class="line">          ifelse(<span class="built_in">type</span> == <span class="string">&quot;unsigned&quot;</span>, 6, 12))       </span><br><span class="line">          )</span><br><span class="line">          )</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="built_in">print</span>(<span class="built_in">paste</span>(<span class="string">&quot;Finally chooosed power is :&quot;</span>, power))</span><br></pre></td></tr></table></figure>

<h3 id="f-构建共表达网络"><a href="#f-构建共表达网络" class="headerlink" title="f 构建共表达网络"></a>f <strong>构建共表达网络</strong></h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## ---- echo=T-------------------------------------------------------------</span></span><br><span class="line"><span class="comment">##一步法网络构建：One-step network construction and module detection##</span></span><br><span class="line"><span class="comment"># power: 上一步计算的软阈值</span></span><br><span class="line"><span class="comment"># maxBlockSize: 计算机能处理的最大模块的基因数量 (默认5000)；</span></span><br><span class="line"><span class="comment">#  4G内存电脑可处理8000-10000个，16G内存电脑可以处理2万个，32G内存电脑可</span></span><br><span class="line"><span class="comment">#  以处理3万个</span></span><br><span class="line"><span class="comment">#  计算资源允许的情况下最好放在一个block里面。</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># corType: pearson or bicor</span></span><br><span class="line"><span class="comment"># numericLabels: 返回数字而不是颜色作为模块的名字，后面可以再转换为颜色</span></span><br><span class="line"><span class="comment"># saveTOMs：最耗费时间的计算，存储起来，供后续使用</span></span><br><span class="line"><span class="comment"># mergeCutHeight: 合并模块的阈值，越大模块越少；越小模块越多，冗余度越大；</span></span><br><span class="line"><span class="comment"># 一般在0.15-0.3之间</span></span><br><span class="line"><span class="comment"># loadTOMs: 避免重复计算</span></span><br><span class="line"></span><br><span class="line">net = blockwiseModules(dataExpr, power = power, maxBlockSize = nGenes,</span><br><span class="line">                       TOMType = <span class="built_in">type</span>, minModuleSize = 25,</span><br><span class="line">                       networkType = <span class="built_in">type</span>,</span><br><span class="line">                       reassignThreshold = 0, mergeCutHeight = 0.2,</span><br><span class="line">                       numericLabels = TRUE, pamRespectsDendro = FALSE,</span><br><span class="line">                       saveTOMs=TRUE, corType = corType, </span><br><span class="line">                       maxPOutliers=maxPOutliers, loadTOM=TRUE,</span><br><span class="line">                       TOMDenom = <span class="string">&quot;min&quot;</span>,  deepSplit = 1,</span><br><span class="line">                       stabilityCriterion = <span class="string">&quot;Individual fraction&quot;</span>, </span><br><span class="line">                       saveTOMFileBase = paste0(<span class="string">&quot;net.mid&quot;</span> , <span class="string">&quot;.tom&quot;</span>),</span><br><span class="line">                       verbose = 3, randomSeed=1117)</span><br></pre></td></tr></table></figure>

<p><img src="/%E9%80%9A%E8%BF%87%E5%85%B1%E8%A1%A8%E8%BE%BE%E7%BD%91%E7%BB%9C%E5%88%86%E6%9E%90%EF%BC%88WGCNA%EF%BC%89%205201c79a5b3e49c5a4f38f90d7c724ba/Untitled%2021.png" class="lazyload" data-srcset="/%E9%80%9A%E8%BF%87%E5%85%B1%E8%A1%A8%E8%BE%BE%E7%BD%91%E7%BB%9C%E5%88%86%E6%9E%90%EF%BC%88WGCNA%EF%BC%89%205201c79a5b3e49c5a4f38f90d7c724ba/Untitled%2021.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## ---- echo=T, fig.cap=&quot;层级聚类树展示各个模块&quot;---------------------------</span></span><br><span class="line"><span class="comment">## 灰色的为**未分类**到模块的基因。</span></span><br><span class="line"><span class="comment"># Convert labels to colors for plotting</span></span><br><span class="line">moduleLabels = net<span class="variable">$colors</span></span><br><span class="line">moduleColors = labels2colors(moduleLabels)</span><br><span class="line"><span class="comment"># Plot the dendrogram and the module colors underneath</span></span><br><span class="line"><span class="comment"># 如果对结果不满意，还可以recutBlockwiseTrees，节省计算时间</span></span><br><span class="line">pdf(file=<span class="string">&quot;plotDendroAndColors.pdf&quot;</span>, onefile=F, paper=<span class="string">&quot;special&quot;</span>, </span><br><span class="line">    <span class="built_in">bg</span>=<span class="string">&quot;white&quot;</span>, pointsize=6)</span><br><span class="line">dynamicColors &lt;- labels2colors(net<span class="variable">$unmergedColors</span>)</span><br><span class="line">plotDendroAndColors(net<span class="variable">$dendrograms</span>[[1]], cbind(dynamicColors,moduleColors),</span><br><span class="line">                    c(<span class="string">&quot;Dynamic Tree Cut&quot;</span>, <span class="string">&quot;Module colors&quot;</span>),</span><br><span class="line">                    dendroLabels = FALSE, hang = 0.5,</span><br><span class="line">                    addGuide = TRUE, guideHang = 0.05)</span><br><span class="line">dev.off()</span><br></pre></td></tr></table></figure>

<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252243496.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252243496.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<h3 id="g-共表达网络结果输出"><a href="#g-共表达网络结果输出" class="headerlink" title="g 共表达网络结果输出"></a>g <strong>共表达网络结果输出</strong></h3><ol>
<li>输出基因及其所在模块信息，方便对模块进行富集分析。</li>
<li>输出模块的主成分信息 (ME)，代表模块整体基因表达量</li>
</ol>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## ---- echo=T, fig.cap=&quot;模块之间的相关性&quot;, fig.width=5, fig.height=8------</span></span><br><span class="line"><span class="comment">### 基因和所在模块信息</span></span><br><span class="line">gene_module &lt;- data.frame(ID=colnames(dataExpr), module=moduleColors)</span><br><span class="line">gene_module = gene_module[order(gene_module<span class="variable">$module</span>),]</span><br><span class="line">write.table(gene_module,file=paste0(<span class="string">&quot;finall.data&quot;</span>,<span class="string">&quot;.gene_module.xls&quot;</span>),</span><br><span class="line">            sep=<span class="string">&quot;\t&quot;</span>,quote=F,row.names=F)</span><br><span class="line"></span><br><span class="line"><span class="comment"># module eigengene, 可以绘制线图，作为每个模块的基因表达趋势的展示</span></span><br><span class="line">MEs = net<span class="variable">$MEs</span></span><br><span class="line"></span><br><span class="line"><span class="comment">### 不需要重新计算，改下列名字就好</span></span><br><span class="line"><span class="comment">### 官方教程是重新计算的，其实可以不用这么麻烦</span></span><br><span class="line">MEs_col = MEs</span><br><span class="line">colnames(MEs_col) = paste0(<span class="string">&quot;ME&quot;</span>, labels2colors(</span><br><span class="line">  as.numeric(str_replace_all(colnames(MEs),<span class="string">&quot;ME&quot;</span>,<span class="string">&quot;&quot;</span>))))</span><br><span class="line">MEs_col = orderMEs(MEs_col)</span><br><span class="line"></span><br><span class="line"><span class="comment">## 保存模块代表性信息</span></span><br><span class="line">MEs_colt = as.data.frame(t(MEs_col))</span><br><span class="line">colnames(MEs_colt) = rownames(dataExpr)</span><br><span class="line">write.table(MEs_colt,file=paste0(<span class="string">&quot;finall.data&quot;</span>,<span class="string">&quot;.module_eipgengene.xls&quot;</span>),</span><br><span class="line">            sep=<span class="string">&quot;\t&quot;</span>,quote=F)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 根据基因间表达量进行聚类所得到的各模块间的相关性图</span></span><br><span class="line"><span class="comment"># marDendro/marHeatmap 设置下、左、上、右的边距</span></span><br><span class="line"></span><br><span class="line">pdf(file=<span class="string">&quot;plotEigengeneNetworks.pdf&quot;</span>, onefile=F, paper=<span class="string">&quot;special&quot;</span>, </span><br><span class="line">  <span class="built_in">bg</span>=<span class="string">&quot;white&quot;</span>, pointsize=6)</span><br><span class="line">plotEigengeneNetworks(MEs_col, <span class="string">&quot;Eigengene adjacency heatmap&quot;</span>, </span><br><span class="line">                      marDendro = c(3,3,2,4),</span><br><span class="line">                      marHeatmap = c(3,4,2,2), plotDendrograms = T, </span><br><span class="line">                      xLabelsAngle = 90)</span><br><span class="line">dev.off()</span><br><span class="line"></span><br><span class="line">R.devices::toPDF(<span class="string">&quot;plotEigengeneNetworks.pdf&quot;</span>, &#123;</span><br><span class="line">  plotEigengeneNetworks(MEs_col, <span class="string">&quot;Eigengene adjacency heatmap&quot;</span>, </span><br><span class="line">                      marDendro = c(3,3,2,4),</span><br><span class="line">                      marHeatmap = c(3,4,2,2), plotDendrograms = T, </span><br><span class="line">                      xLabelsAngle = 90)</span><br><span class="line">&#125;)</span><br></pre></td></tr></table></figure>

<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252243545.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252243545.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<h3 id="h-筛选Hub基因"><a href="#h-筛选Hub基因" class="headerlink" title="h 筛选Hub基因"></a>h <strong>筛选Hub基因</strong></h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">hubs = chooseTopHubInEachModule(dataExpr, colorh=moduleColors, power=power, <span class="built_in">type</span>=<span class="built_in">type</span>)</span><br><span class="line">hubs</span><br><span class="line">con &lt;- nearestNeighborConnectivity(dataExpr, nNeighbors=50, power=power,</span><br><span class="line">                                   <span class="built_in">type</span>=<span class="built_in">type</span>, corFnc = corType)</span><br></pre></td></tr></table></figure>

<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252244643.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252244643.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<h3 id="i-获取TOM矩阵，导出Cytoscape可用的数据方便网络图绘制"><a href="#i-获取TOM矩阵，导出Cytoscape可用的数据方便网络图绘制" class="headerlink" title="i 获取TOM矩阵，导出Cytoscape可用的数据方便网络图绘制"></a>i <strong>获取TOM矩阵，导出Cytoscape可用的数据方便网络图绘制</strong></h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line">load(net<span class="variable">$TOMFiles</span>[1], verbose=T)</span><br><span class="line">TOM &lt;- as.matrix(TOM)</span><br><span class="line"></span><br><span class="line">dissTOM = 1-TOM</span><br><span class="line"><span class="comment"># Transform dissTOM with a power to make moderately strong </span></span><br><span class="line"><span class="comment"># connections more visible in the heatmap</span></span><br><span class="line">plotTOM = dissTOM^power</span><br><span class="line"><span class="comment"># Set diagonal to NA for a nicer plot</span></span><br><span class="line">diag(plotTOM) = NA</span><br><span class="line"><span class="comment"># Call the plot function</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## ---- echo=T, eval=F, fig.cap=&quot;TOM plot&quot;---------------------------------</span></span><br><span class="line"><span class="comment">## # 这一部分特别耗时，行列同时做层级聚类</span></span><br><span class="line"><span class="comment">## TOMplot(plotTOM, net$dendrograms, moduleColors,</span></span><br><span class="line"><span class="comment">##         main = &quot;Network heatmap plot, all genes&quot;)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## ---- fig.cap=&quot;TOMplot&quot;--------------------------------------------------</span></span><br><span class="line"><span class="comment">#knitr::include_graphics(&quot;images/TOMplot.png&quot;)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## ---- echo=T-------------------------------------------------------------</span></span><br><span class="line">probes = colnames(dataExpr)</span><br><span class="line">dimnames(TOM) &lt;- list(probes, probes)</span><br><span class="line">exprMat &lt;- <span class="string">&quot;finall.data&quot;</span></span><br><span class="line"><span class="comment"># Export the network into edge and node list files Cytoscape can read</span></span><br><span class="line"><span class="comment"># threshold 默认为0.5, 可以根据自己的需要调整，也可以都导出后在</span></span><br><span class="line"><span class="comment"># cytoscape中再调整</span></span><br><span class="line">cyt = exportNetworkToCytoscape(TOM,</span><br><span class="line">             edgeFile = <span class="built_in">paste</span>(<span class="string">&quot;finall.data&quot;</span>, <span class="string">&quot;.edges.txt&quot;</span>, sep=<span class="string">&quot;&quot;</span>),</span><br><span class="line">             nodeFile = <span class="built_in">paste</span>(exprMat, <span class="string">&quot;.nodes.txt&quot;</span>, sep=<span class="string">&quot;&quot;</span>),</span><br><span class="line">             weighted = TRUE, threshold = 0.01,</span><br><span class="line">             nodeNames = probes, nodeAttr = moduleColors)</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<h3 id="j-筛选Hub基因2"><a href="#j-筛选Hub基因2" class="headerlink" title="j 筛选Hub基因2"></a>j <strong>筛选Hub基因2</strong></h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">library(dplyr)</span><br><span class="line">edgeData1 &lt;- cyt<span class="variable">$edgeData</span>[,c(1,2,3)]</span><br><span class="line">edgeData2 &lt;- cyt<span class="variable">$edgeData</span>[,c(2,1,3)]</span><br><span class="line">nodeattrib &lt;- cyt<span class="variable">$nodeData</span>[,c(1,3)]</span><br><span class="line">colnames(nodeattrib) &lt;- c(<span class="string">&quot;nodeName&quot;</span>, <span class="string">&quot;Module&quot;</span>)</span><br><span class="line">colnames(edgeData1) &lt;- c(<span class="string">&quot;Node1&quot;</span>,<span class="string">&quot;Node2&quot;</span>,<span class="string">&quot;Weight&quot;</span>)</span><br><span class="line">colnames(edgeData2) &lt;- c(<span class="string">&quot;Node1&quot;</span>,<span class="string">&quot;Node2&quot;</span>,<span class="string">&quot;Weight&quot;</span>)</span><br><span class="line">edgeData &lt;- rbind(edgeData1, edgeData2)</span><br><span class="line">edgeData<span class="variable">$Module1</span> &lt;- nodeattrib[match(edgeData<span class="variable">$Node1</span>, nodeattrib<span class="variable">$nodeName</span>), 2]</span><br><span class="line">edgeData<span class="variable">$Module2</span> &lt;- nodeattrib[match(edgeData<span class="variable">$Node2</span>, nodeattrib<span class="variable">$nodeName</span>), 2]</span><br><span class="line">edgeData &lt;- edgeData[edgeData<span class="variable">$Module1</span>==edgeData<span class="variable">$Module2</span>,c(1,3,4)]</span><br><span class="line"><span class="built_in">head</span>(edgeData)</span><br></pre></td></tr></table></figure>

<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252244814.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252244814.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">nodeTotalWeight &lt;- edgeData %&gt;% group_by(Node1, Module1) %&gt;% summarise(weight=<span class="built_in">sum</span>(Weight))</span><br><span class="line"></span><br><span class="line">nodeTotalWeight &lt;- nodeTotalWeight[with(nodeTotalWeight, order(Module1, -weight)),]</span><br><span class="line"><span class="built_in">head</span>(nodeTotalWeight)</span><br></pre></td></tr></table></figure>

<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252244400.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252244400.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">nodeTotalWeightTop20 = nodeTotalWeight %&gt;% group_by(Module1) %&gt;% top_n(2, weight)</span><br><span class="line">nodeTotalWeightTop20</span><br></pre></td></tr></table></figure>

<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252244227.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252244227.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<h3 id="k表型关联分析-有表型数据可做"><a href="#k表型关联分析-有表型数据可做" class="headerlink" title="k表型关联分析 (有表型数据可做)"></a>k<strong>表型关联分析 (有表型数据可做)</strong></h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 读入表型数据，不是必须的</span></span><br><span class="line"><span class="keyword">if</span>(trait != <span class="string">&quot;&quot;</span>) &#123;</span><br><span class="line">  <span class="comment">#traitData &lt;- read.table(file=trait, sep=&#x27;\t&#x27;, header=T, row.names=1,</span></span><br><span class="line">  <span class="comment">#                        check.names=FALSE, comment=&#x27;&#x27;,quote=&quot;&quot;)</span></span><br><span class="line">  sampleName = rownames(dataExpr)</span><br><span class="line">  traitData = traitData[match(sampleName, rownames(traitData)), ]</span><br><span class="line"></span><br><span class="line">  <span class="comment">#sampleTree2 = hclust(dist(dataExpr), method = &quot;average&quot;)</span></span><br><span class="line">  <span class="comment"># Convert traits to a color representation: white means low, red means high, grey means missing entry</span></span><br><span class="line">  traitColors = numbers2colors(traitData, signed = FALSE);</span><br><span class="line">  <span class="comment"># Plot the sample dendrogram and the colors underneath.</span></span><br><span class="line">  plotDendroAndColors(sampleTree, traitColors,</span><br><span class="line">                    groupLabels = names(traitData), </span><br><span class="line">                    main = <span class="string">&quot;Sample dendrogram and trait heatmap&quot;</span>)</span><br><span class="line">  </span><br><span class="line">  </span><br><span class="line">  <span class="comment">## 如果有表型数据，也可以跟ME数据放一起，一起出图</span></span><br><span class="line">  MEs_colpheno = orderMEs(cbind(MEs_col, traitData))</span><br><span class="line">  plotEigengeneNetworks(MEs_colpheno, <span class="string">&quot;Eigengene adjacency heatmap&quot;</span>, </span><br><span class="line">                    marDendro = c(3,3,2,4), marHeatmap = c(3,4,2,2),</span><br><span class="line">                    plotDendrograms = T, xLabelsAngle = 90)</span><br><span class="line">  </span><br><span class="line">  </span><br><span class="line">    <span class="comment">## ---- echo=T, fig.height=8, fig,width=8, fig.cap=&quot;模块与表型的关联&quot;------</span></span><br><span class="line">    <span class="comment">### 模块与表型数据关联</span></span><br><span class="line">    <span class="keyword">if</span> (corType==<span class="string">&quot;pearsoon&quot;</span>) &#123;</span><br><span class="line">      modTraitCor = cor(MEs_col, traitData, use = <span class="string">&quot;p&quot;</span>)</span><br><span class="line">      modTraitP = corPvalueStudent(modTraitCor, nSamples)</span><br><span class="line">    &#125; <span class="keyword">else</span> &#123;</span><br><span class="line">      modTraitCorP = bicorAndPvalue(MEs_col, traitData, robustY=robustY)</span><br><span class="line">      modTraitCor = modTraitCorP<span class="variable">$bicor</span></span><br><span class="line">      modTraitP   = modTraitCorP<span class="variable">$p</span></span><br><span class="line">    &#125;</span><br><span class="line">    <span class="comment"># signif表示保留几位小数</span></span><br><span class="line">    textMatrix = <span class="built_in">paste</span>(signif(modTraitCor, 2), <span class="string">&quot;\n(&quot;</span>, signif(modTraitP, 1), <span class="string">&quot;)&quot;</span>, sep = <span class="string">&quot;&quot;</span>)</span><br><span class="line">    dim(textMatrix) = dim(modTraitCor)</span><br><span class="line"></span><br><span class="line">    pdf(file=<span class="string">&quot;labeledHeatmap.pdf&quot;</span>, onefile=F, paper=<span class="string">&quot;special&quot;</span>, </span><br><span class="line">        pointsize=6)</span><br><span class="line">    labeledHeatmap(Matrix = modTraitCor, xLabels = colnames(traitData), </span><br><span class="line">                   yLabels = colnames(MEs_col), </span><br><span class="line">                   cex.lab = 0.5, </span><br><span class="line">                   ySymbols = colnames(MEs_col), colorLabels = FALSE, </span><br><span class="line">                   colors = blueWhiteRed(50), </span><br><span class="line">                   textMatrix = textMatrix, setStdMargins = FALSE, </span><br><span class="line">                   cex.text = 0.5, zlim = c(-1,1),</span><br><span class="line">                   main = <span class="built_in">paste</span>(<span class="string">&quot;Module-trait relationships&quot;</span>))</span><br><span class="line">    dev.off()</span><br><span class="line">    </span><br><span class="line">    labeledHeatmap(Matrix = modTraitCor, xLabels = colnames(traitData), </span><br><span class="line">                   yLabels = colnames(MEs_col), </span><br><span class="line">                   cex.lab = 0.5, </span><br><span class="line">                   ySymbols = colnames(MEs_col), colorLabels = FALSE, </span><br><span class="line">                   colors = blueWhiteRed(50), </span><br><span class="line">                   textMatrix = textMatrix, setStdMargins = FALSE, </span><br><span class="line">                   cex.text = 0.5, zlim = c(-1,1),</span><br><span class="line">                   main = <span class="built_in">paste</span>(<span class="string">&quot;Module-trait relationships&quot;</span>))</span><br><span class="line"></span><br><span class="line">    modTraitCorMelt = as.data.frame(modTraitCor)</span><br><span class="line">    write.table(modTraitCorMelt,file=paste0(exprMat,<span class="string">&quot;.module_trait_correlation.xls&quot;</span>),</span><br><span class="line">                sep=<span class="string">&quot;\t&quot;</span>,quote=F)</span><br><span class="line">    modTraitCorMelt<span class="variable">$ID</span> = rownames(modTraitCor)</span><br><span class="line">    modTraitCorMelt = melt(modTraitCorMelt)</span><br><span class="line">    colnames(modTraitCorMelt) &lt;- c(<span class="string">&quot;Module&quot;</span>,<span class="string">&quot;Trait&quot;</span>,<span class="string">&quot;PersonCorrelationValue&quot;</span>)</span><br><span class="line">    modTraitPMelt = as.data.frame(modTraitP)</span><br><span class="line">    write.table(modTraitPMelt,file=paste0(exprMat,<span class="string">&quot;.module_trait_correlationPvalue.xls&quot;</span>),</span><br><span class="line">                sep=<span class="string">&quot;\t&quot;</span>,quote=F)</span><br><span class="line">    modTraitPMelt<span class="variable">$ID</span> = rownames(modTraitP)</span><br><span class="line">    modTraitPMelt = melt(modTraitPMelt)</span><br><span class="line">    colnames(modTraitPMelt) &lt;- c(<span class="string">&quot;Module&quot;</span>,<span class="string">&quot;Trait&quot;</span>,<span class="string">&quot;Pvalue&quot;</span>)</span><br><span class="line">    <span class="comment">#modTraitCorP = cbind(modTraitCorMelt, Pvalue=modTraitPMelt$Pvalue)</span></span><br><span class="line">    modTraitCorP = merge(modTraitCorMelt, modTraitPMelt, by=c(<span class="string">&quot;Module&quot;</span>,<span class="string">&quot;Trait&quot;</span>))</span><br><span class="line">    write.table(modTraitCorP,file=paste0(exprMat,<span class="string">&quot;.module_trait_correlationPvalueMelt.xls&quot;</span>),</span><br><span class="line">                sep=<span class="string">&quot;\t&quot;</span>,quote=F,row.names=F)</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<h1 id="三、WGCNA原理解析"><a href="#三、WGCNA原理解析" class="headerlink" title="三、WGCNA原理解析"></a>三、WGCNA原理解析</h1><p>该解析来源于</p>
<p><strong>WGCAN原理的提出者文章：</strong></p>
<p>Zhang, Bin, and Steve Horvath. “A general framework for weighted gene co-expression network analysis.” <em>Statistical applications in genetics and molecular biology</em> 4.1 (2005).<br><strong>WGCAN R语言实现文章</strong></p>
<p>Langfelder, Peter, and Steve Horvath. “WGCNA: an R package for weighted correlation network analysis.” <em>BMC bioinformatics</em> 9 (2008): 1-13.</p>
<p><strong>WGCAN处理流程</strong></p>
<p>宋长新,雷萍,王婷.基于WGCNA算法的基因共表达网络构建理论及其R软件实现[J].基因组学与应用生物学,2013,32(01):135-141.</p>
<h2 id="1-WGCAN流程如下图"><a href="#1-WGCAN流程如下图" class="headerlink" title="1.WGCAN流程如下图"></a>1.WGCAN流程如下图</h2><p>以下对上两部进行讲解</p>
<p><img src="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252244184.png" class="lazyload" data-srcset="https://my-bloag-image-xiaowu.oss-cn-beijing.aliyuncs.com/image/202403252244184.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" alt="Untitled"></p>
<h2 id="2-构建共表达网络"><a href="#2-构建共表达网络" class="headerlink" title="2.构建共表达网络"></a>2.构建共表达网络</h2><p>这一步是利用了利用基因之间的相互作用模式，来得到他们之间的关系。并且以相关性作为共表达的度量。</p>

  </div>
  
  
    
    <div class='footer'>
       <!-- 参考资料、相关资料等 -->
      
       <!-- 相关文章 -->
      
      <!-- 版权声明组件 -->
      
      <!-- 打赏组件 -->
      
    </div>
  
  
    


  <div class='article-meta' id="bottom">
    <div class='new-meta-box'>
      
        
          <div class="new-meta-item date" itemprop="dateModified" datetime="2024-03-25T22:51:16+08:00">
  <a class='notlink'>
    <i class="fa-solid fa-edit fa-fw" aria-hidden="true"></i>
    <p>更新于：2024年3月25日</p>
  </a>
</div>

        
      
        
          
  <div class="new-meta-item wordcount">
    <a class='notlink'>
      <i class="fa-solid fa-keyboard fa-fw" aria-hidden="true"></i>
      <p>字数：9.6k 字</p>
    </a>
  </div>
  <div class="new-meta-item readtime">
    <a class='notlink'>
      <i class="fa-solid fa-hourglass-half fa-fw" aria-hidden="true"></i>
      <p>时长：43 分钟</p>
    </a>
  </div>


        
      
        
          
  
  
  
  <div class="new-meta-item meta-tags"><a class="tag" href="/tags/WGCAN%E7%BD%91%E7%BB%9C/" rel="nofollow"><i class="fa-solid fa-hashtag fa-fw" aria-hidden="true"></i><p>WGCAN网络</p></a></div>
  <span hidden itemprop="keywords">WGCAN网络</span>
  


        
      
        
          
  <div class="new-meta-item share -mob-share-list">
  <div class="-mob-share-list share-body">
    
      
        <a class="-mob-share-qq" title="" rel="external nofollow noopener noreferrer noopener"
          
          target="_blank" href="http://connect.qq.com/widget/shareqq/index.html?url=https://snowman12137.github.io/2024/03/25/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90-%E9%80%9A%E8%BF%87%E5%85%B1%E8%A1%A8%E8%BE%BE%E7%BD%91%E7%BB%9C%E5%88%86%E6%9E%90%EF%BC%88WGCNA%EF%BC%89/&title=通过共表达网络分析（WGCNA） - 小吴の博客&summary="
          
          >
          
            <img src="https://unpkg.com/volantis-static@0.0.1660614606622/media/org.volantis/logo/128/qq.png" class="lazyload" data-srcset="https://unpkg.com/volantis-static@0.0.1660614606622/media/org.volantis/logo/128/qq.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==">
          
        </a>
      
    
      
        <a class="-mob-share-qzone" title="" rel="external nofollow noopener noreferrer noopener"
          
          target="_blank" href="https://sns.qzone.qq.com/cgi-bin/qzshare/cgi_qzshare_onekey?url=https://snowman12137.github.io/2024/03/25/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90-%E9%80%9A%E8%BF%87%E5%85%B1%E8%A1%A8%E8%BE%BE%E7%BD%91%E7%BB%9C%E5%88%86%E6%9E%90%EF%BC%88WGCNA%EF%BC%89/&title=通过共表达网络分析（WGCNA） - 小吴の博客&summary="
          
          >
          
            <img src="https://unpkg.com/volantis-static@0.0.1660614606622/media/org.volantis/logo/128/qzone.png" class="lazyload" data-srcset="https://unpkg.com/volantis-static@0.0.1660614606622/media/org.volantis/logo/128/qzone.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==">
          
        </a>
      
    
      
        <a class="-mob-share-weibo" title="" rel="external nofollow noopener noreferrer noopener"
          
          target="_blank" href="http://service.weibo.com/share/share.php?url=https://snowman12137.github.io/2024/03/25/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90-%E9%80%9A%E8%BF%87%E5%85%B1%E8%A1%A8%E8%BE%BE%E7%BD%91%E7%BB%9C%E5%88%86%E6%9E%90%EF%BC%88WGCNA%EF%BC%89/&title=通过共表达网络分析（WGCNA） - 小吴の博客&summary="
          
          >
          
            <img src="https://unpkg.com/volantis-static@0.0.1660614606622/media/org.volantis/logo/128/weibo.png" class="lazyload" data-srcset="https://unpkg.com/volantis-static@0.0.1660614606622/media/org.volantis/logo/128/weibo.png" srcset="data:image/gif;base64,R0lGODlhAQABAIAAAP///////yH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==">
          
        </a>
      
    
      
    
      
    
  </div>
</div>



        
      
    </div>
    <!-- Custom Files bottomMeta begin -->
    
    <!-- Custom Files bottomMeta end -->
  </div>


  
  

  
    <div class="prev-next">
      
        <a class='prev' href='/2024/03/25/%E5%AF%86%E7%A0%81%E5%AD%A6-%E6%95%B0%E8%AE%BA%E5%9F%BA%E7%A1%80-%E4%BA%8C-%E2%80%94%E6%A4%AD%E5%9C%86%E6%9B%B2%E7%BA%BF/'>
          <p class='title'><i class="fa-solid fa-chevron-left" aria-hidden="true"></i>数论基础(二)—椭圆曲线</p>
          <p class='content'>数论基础(二)—椭圆曲线参考文章：
https://accu.cc/content/cryptography/ecc/
https://blog.csdn.net/qq_44925830&#x2...</p>
        </a>
      
      
        <a class='next' href='/2024/03/16/Chatgpt-api%E5%85%85%E5%80%BC-%E5%8D%87%E7%BA%A7plus-%E6%95%99%E7%A8%8B-%E8%99%9A%E6%8B%9F%E4%BF%A1%E7%94%A8%E5%8D%A1/'>
          <p class='title'>Chatgpt api充值 / 升级plus 教程 /虚拟信用卡<i class="fa-solid fa-chevron-right" aria-hidden="true"></i></p>
          <p class='content'>Chatgpt api充值 &#x2F; 升级plus 教程 &#x2F;虚拟信用卡
系列文章：
科学上网工具
国内手机号注册Google账号（2024亲测有效）
Chatgpt api充值 &...</p>
        </a>
      
    </div>
  
  <!-- 文章推荐 -->
  
  
  <!-- Custom Files postEnd begin-->
  
  <!-- Custom Files postEnd end-->
</article>


  


  <article class="post white-box shadow floatable blur" id="comments">
    <span hidden>
      <meta itemprop="discussionUrl" content="/2024/03/25/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90-%E9%80%9A%E8%BF%87%E5%85%B1%E8%A1%A8%E8%BE%BE%E7%BD%91%E7%BB%9C%E5%88%86%E6%9E%90%EF%BC%88WGCNA%EF%BC%89/index.html#comments">
    </span>
    <p ct><i class='fa-solid fa-comments'></i> 评论</p>
    

    <div id="layoutHelper-comments"></div>

  </article>






</div>
<aside id='l_side' itemscope itemtype="http://schema.org/WPSideBar">
  

  
    
    
      
    
  


<div class="widget-sticky">

  
  


<section class="widget toc-wrapper desktop mobile" id="toc-div" >
  
  <header>
    
      <i class="fa-solid fa-list fa-fw" aria-hidden="true"></i><span class='name'>本文目录</span>
    
  </header>


  <div class='content'>
      <ol class="toc"><li class="toc-item toc-level-2"><a class="toc-link" href="#1-%E6%BA%90%E7%A0%81%E4%B8%8B%E8%BD%BD"><span class="toc-text">1.源码下载</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#a-%E6%96%87%E4%BB%B6%E5%87%86%E5%A4%87"><span class="toc-text">a.文件准备</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#b-R%E5%8C%85%E5%87%86%E5%A4%87"><span class="toc-text">b.R包准备</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#2-%E4%B8%8B%E8%BD%BD%E6%95%B0%E6%8D%AE"><span class="toc-text">2.下载数据</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#a-%E4%B8%8B%E8%BD%BD%E7%94%9F%E7%89%A9%E6%A0%B7%E6%9C%AC%E4%B8%B4%E5%BA%8A%E6%95%B0%E6%8D%AE"><span class="toc-text">a.下载生物样本临床数据</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#b-%E4%B8%8B%E8%BD%BD%E6%8B%B7%E8%B4%9D%E6%95%B0%E6%8D%AE"><span class="toc-text">b 下载拷贝数据</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#3-%E6%95%B0%E6%8D%AE%E9%A2%84%E5%A4%84%E7%90%86"><span class="toc-text">3.数据预处理</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#a-%E8%BF%9B%E8%A1%8CFPKM%E8%BF%9B%E8%A1%8C%E6%95%B0%E6%8D%AE%E5%BD%92%E4%B8%80%E5%8C%96"><span class="toc-text">a. 进行FPKM进行数据归一化</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#b-%E8%BF%9B%E8%A1%8C%E6%B3%A8%E9%87%8A-x2F-FPKM"><span class="toc-text">b.进行注释&#x2F;FPKM</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#c-%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%9F%BA%E5%9B%A0%E7%AD%9B%E9%80%89-x2F-%E7%81%AB%E5%B1%B1%E5%9B%BE-edgeR-x2F-%E7%83%AD%E5%9B%BE-gplots"><span class="toc-text">c 差异表达基因筛选&#x2F;火山图(edgeR)&#x2F;热图(gplots)</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#1-%E5%9F%BA%E6%9C%AC%E6%A6%82%E5%BF%B5"><span class="toc-text">1.基本概念</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#2%E5%9F%BA%E6%9C%AC%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B"><span class="toc-text">2基本分析流程</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#3-WGCNA%E5%8C%85%E5%AE%9E%E6%88%98"><span class="toc-text">3 WGCNA包实战</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#a-%E8%BE%93%E5%85%A5%E6%95%B0%E6%8D%AE%E5%92%8C%E5%8F%82%E6%95%B0%E9%80%89%E6%8B%A9"><span class="toc-text">a. 输入数据和参数选择</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#b-%E5%AE%89%E8%A3%85%E5%AF%BC%E5%85%A5%E4%BE%9D%E8%B5%96%E7%9A%84%E5%8C%85%E5%8F%8A%E6%95%B0%E6%8D%AE"><span class="toc-text">b 安装导入依赖的包及数据</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#c-%E6%95%B0%E6%8D%AE%E5%9F%BA%E6%9C%AC%E5%A4%84%E7%90%86"><span class="toc-text">c 数据基本处理</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#d-%E6%A0%B7%E6%9C%AC%E5%B1%82%E7%BA%A7%E8%81%9A%E7%B1%BB%EF%BC%8C%E6%9F%A5%E7%9C%8B%E6%9C%89%E6%97%A0%E7%A6%BB%E7%BE%A4%E5%80%BC"><span class="toc-text">d 样本层级聚类，查看有无离群值</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#e-%E7%A1%AE%E5%AE%9A%E8%BD%AF%E9%98%88%E5%80%BC"><span class="toc-text">e 确定软阈值</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#f-%E6%9E%84%E5%BB%BA%E5%85%B1%E8%A1%A8%E8%BE%BE%E7%BD%91%E7%BB%9C"><span class="toc-text">f 构建共表达网络</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#g-%E5%85%B1%E8%A1%A8%E8%BE%BE%E7%BD%91%E7%BB%9C%E7%BB%93%E6%9E%9C%E8%BE%93%E5%87%BA"><span class="toc-text">g 共表达网络结果输出</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#h-%E7%AD%9B%E9%80%89Hub%E5%9F%BA%E5%9B%A0"><span class="toc-text">h 筛选Hub基因</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#i-%E8%8E%B7%E5%8F%96TOM%E7%9F%A9%E9%98%B5%EF%BC%8C%E5%AF%BC%E5%87%BACytoscape%E5%8F%AF%E7%94%A8%E7%9A%84%E6%95%B0%E6%8D%AE%E6%96%B9%E4%BE%BF%E7%BD%91%E7%BB%9C%E5%9B%BE%E7%BB%98%E5%88%B6"><span class="toc-text">i 获取TOM矩阵，导出Cytoscape可用的数据方便网络图绘制</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#j-%E7%AD%9B%E9%80%89Hub%E5%9F%BA%E5%9B%A02"><span class="toc-text">j 筛选Hub基因2</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#k%E8%A1%A8%E5%9E%8B%E5%85%B3%E8%81%94%E5%88%86%E6%9E%90-%E6%9C%89%E8%A1%A8%E5%9E%8B%E6%95%B0%E6%8D%AE%E5%8F%AF%E5%81%9A"><span class="toc-text">k表型关联分析 (有表型数据可做)</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#1-WGCAN%E6%B5%81%E7%A8%8B%E5%A6%82%E4%B8%8B%E5%9B%BE"><span class="toc-text">1.WGCAN流程如下图</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#2-%E6%9E%84%E5%BB%BA%E5%85%B1%E8%A1%A8%E8%BE%BE%E7%BD%91%E7%BB%9C"><span class="toc-text">2.构建共表达网络</span></a></li></ol>
  </div>
</section>

  

</div>

  <!-- Custom Files side begin -->
  
  <!-- Custom Files side end -->
</aside>



        </div>
        
  
  <footer class="footer clearfix"  itemscope itemtype="http://schema.org/WPFooter">
    <br><br>
    
      
        <div class="aplayer-container">
          


        </div>
      
    
      
        <br>
        <div class="social-wrapper" itemprop="about" itemscope itemtype="http://schema.org/Thing">
          
            
          
            
          
            
          
        </div>
      
    
      
        <div><p>博客内容遵循 <a target="_blank" rel="noopener" href="https://creativecommons.org/licenses/by-nc-sa/4.0/deed.zh">署名-非商业性使用-相同方式共享 4.0 国际 (CC BY-NC-SA 4.0) 协议</a></p>
</div>
      
    
      
        本站使用
        <a href="https://github.com/volantis-x/hexo-theme-volantis/#6.0.0-alpha.1" target="_blank" class="codename">Volantis</a>
        作为主题
      
    
      
        <div class='copyright'>
        <p><a href="/">Copyright © since 2017 XXX</a></p>

        </div>
      
    
    <!-- Custom Files footer begin-->
    
    <!-- Custom Files footer end-->
  </footer>


        <a id="s-top" class="fa-solid fa-arrow-up fa-fw" title="top"></a>
      </div>
    </div>
    <div>
      <script>
  /******************** volantis.dom ********************************/
  // 页面选择器 将dom对象缓存起来 see: /source/js/app.js etc.
  volantis.dom.bodyAnchor = volantis.dom.$(document.getElementById("safearea")); // 页面主体
  volantis.dom.topBtn = volantis.dom.$(document.getElementById('s-top')); // 向上
  volantis.dom.wrapper = volantis.dom.$(document.getElementById('wrapper')); // 整个导航栏
  volantis.dom.coverAnchor = volantis.dom.$(document.querySelector('#l_cover .cover-wrapper')); // 1个
  volantis.dom.switcher = volantis.dom.$(document.querySelector('#l_header .switcher .s-search')); // 搜索按钮   移动端 1个
  volantis.dom.header = volantis.dom.$(document.getElementById('l_header')); // 移动端导航栏
  volantis.dom.search = volantis.dom.$(document.querySelector('#l_header .m_search')); // 搜索框 桌面端 移动端 1个
  volantis.dom.mPhoneList = volantis.dom.$(document.querySelectorAll('#l_header .m-phone .list-v')); //  手机端 子菜单 多个
</script>

<script>
  
  volantis.css("https://cdn.bootcdn.net/ajax/libs/font-awesome/6.1.1/css/all.min.css");
  
  
  
</script>

<!-- required -->


<!-- internal -->

<script src="/js/app.js"></script>



  













<div id="rightmenu-wrapper">
  <ul class="list-v rightmenu" id="rightmenu-content">
    
  <li class='navigation menuNavigation-Content'>


    <a class="nav icon-only fix-cursor-default" onclick="history.back()"><i class="fa-solid fa-arrow-left fa-fw"></i></a>



    <a class="nav icon-only fix-cursor-default" onclick="history.forward()"><i class="fa-solid fa-arrow-right fa-fw"></i></a>



    <a class="nav icon-only fix-cursor-default" onclick="window.location.reload()"><i class="fa-solid fa-redo fa-fw"></i></a>



    <a class="nav icon-only fix-cursor-default" onclick="VolantisApp.scrolltoElement(volantis.dom.bodyAnchor)"><i class="fa-solid fa-arrow-up fa-fw"></i></a>


  </li>


    <hr class="menuLoad-Content" >



  <li class="menuLoad-Content">
    <span class="vlts-menu fix-cursor-default event" id="copyPaste" data-event="copyPaste" data-group="inputBox">
      <i class="fa-solid fa-paste fa-fw"></i>
      粘贴文本
    </span>
  </li>



  <li class="menuLoad-Content">
    <span class="vlts-menu fix-cursor-default event" id="copyAll" data-event="copyAll" data-group="inputBox">
      <i class="fa-solid fa-object-ungroup fa-fw"></i>
      全选文本
    </span>
  </li>



  <li class="menuLoad-Content">
    <span class="vlts-menu fix-cursor-default event" id="copyCut" data-event="copyCut" data-group="inputBox">
      <i class="fa-solid fa-cut fa-fw"></i>
      剪切文本
    </span>
  </li>



  <li class="menuLoad-Content">
    <span class="vlts-menu fix-cursor-default event" id="copyText" data-event="copyText" data-group="seletctText">
      <i class="fa-solid fa-copy fa-fw"></i>
      复制文本
    </span>
  </li>



  <li class="menuLoad-Content">
    <span class="vlts-menu fix-cursor-default event" id="searchWord" data-event="OpenSearch(__text__)" data-group="seletctText">
      <i class="fa-solid fa-search fa-fw"></i>
      站内搜索
    </span>
  </li>



  <li class="menuLoad-Content">
    <span class="vlts-menu fix-cursor-default event" id="bingSearch" data-event="window.open(`https://cn.bing.com/search?q=${__text__}`)" data-group="seletctText">
      <i class="fa-solid fa-search fa-fw"></i>
      必应搜索
    </span>
  </li>



  <li class="menuLoad-Content">
    <span class="vlts-menu fix-cursor-default event" id="openTab" data-event="window.open(__link__)" data-group="elementCheck">
      <i class="fa-solid fa-external-link-square-alt fa-fw"></i>
      新标签页打开
    </span>
  </li>



  <li class="menuLoad-Content">
    <span class="vlts-menu fix-cursor-default event" id="copyLink" data-event="copyLink" data-group="elementCheck">
      <i class="fa-solid fa-link fa-fw"></i>
      复制链接地址
    </span>
  </li>



  <li class="menuLoad-Content">
    <span class="vlts-menu fix-cursor-default event" id="copyImg" data-event="copyImg" data-group="elementImage">
      <i class="fa-solid fa-image fa-fw"></i>
      复制图片
    </span>
  </li>



  <li class="menuLoad-Content">
    <span class="vlts-menu fix-cursor-default event" id="googleImg" data-event="window.open(`https://www.google.com.hk/searchbyimage?image_url=${__link__}`)" data-group="elementImage">
      <i class="fa-solid fa-images fa-fw"></i>
      谷歌识图
    </span>
  </li>



  <li class="menuLoad-Content">
    <a class="vlts-menu fix-cursor-default" id="help" target="_blank" rel="noopener" href="https://volantis.js.org/faqs/" data-group="link">
      <i class="fa-solid fa-question fa-fw"></i>
      常见问题
    </a>
  </li>



  <li class="menuLoad-Content">
    <a class="vlts-menu fix-cursor-default" id="examples" target="_blank" rel="noopener" href="https://volantis.js.org/examples/" data-group="link">
      <i class="fa-solid fa-rss fa-fw"></i>
      示例博客
    </a>
  </li>



  <li class="menuLoad-Content">
    <a class="vlts-menu fix-cursor-default" id="contributors" target="_blank" rel="noopener" href="https://volantis.js.org/contributors/" data-group="link">
      <i class="fa-solid fa-fan fa-fw"></i>
      加入社区
    </a>
  </li>



    <hr class="menuLoad-Content" >



  <li class="menuLoad-Content">
    <a class="vlts-menu fix-cursor-default" id="source_docs" target="_blank" rel="noopener" href="https://github.com/volantis-x/volantis-docs/" data-group="link">
      <i class="fa-solid fa-code-branch fa-fw"></i>
      本站源码
    </a>
  </li>



  <li class="menuLoad-Content">
    <a class="vlts-menu fix-cursor-default" id="source_theme" target="_blank" rel="noopener" href="https://github.com/volantis-x/hexo-theme-volantis/" data-group="link">
      <i class="fa-solid fa-code-branch fa-fw"></i>
      主题源码
    </a>
  </li>



    <hr class="menuLoad-Content" >



  <li class="menuLoad-Content">
    <span class="vlts-menu fix-cursor-default event" id="darkMode" data-event="volantis.dark.toggle()" data-group="darkMode">
      <i class="fa-solid fa-moon fa-fw"></i>
      暗黑模式
    </span>
  </li>



  <li class="menuLoad-Content">
    <span class="vlts-menu fix-cursor-default event" id="printMode" data-event="printMode" data-group="articlePage">
      <i class="fa-solid fa-print fa-fw"></i>
      打印页面
    </span>
  </li>



  <li class="menuLoad-Content">
    <span class="vlts-menu fix-cursor-default event" id="readMode" data-event="readMode" data-group="articlePage">
      <i class="fa-solid fa-book-open fa-fw"></i>
      阅读模式
    </span>
  </li>


<div id="menuMusic">
  <li class='music name menuOption-Content'>
    <p class='nav music-title fix-cursor-default'></p>
  </li>
  <li class='music ctrl'>
    <a class='nav icon-only backward fix-cursor-default'  title="backward">
      <i class='fa-solid fa-step-backward fa-fw'></i>
    </a>
    <a class='nav icon-only toggle fix-cursor-default'  title="toggle">
      <i class='fa-solid fa-play fa-fw'></i>
    </a>
    <a class='nav icon-only forward fix-cursor-default'  title="forward">
      <i class='fa-solid fa-step-forward fa-fw'></i>
    </a>
  </li>
  <li class='music volume'>
    <div class='nav volume'>
      <div class="aplayer-volume-bar-wrap">
        <div class="aplayer-volume-bar fix-cursor-pointer">
          <div class="aplayer-volume"></div>
          <i class='left fa-solid fa-volume-off fa-fw'></i>
          <i class='right fa-solid fa-volume-up fa-fw'></i>
        </div>
      </div>
    </div>
  </li>
</div>

  </ul>
</div>
<script src="/js/plugins/rightMenus.js"></script>
<script>
  const RightMenusFunction = {};
  












  //RightMenusFunction['copyPaste'] = (fun) => {fun()}





  //RightMenusFunction['copyAll'] = (fun) => {fun()}





  //RightMenusFunction['copyCut'] = (fun) => {fun()}





  //RightMenusFunction['copyText'] = (fun) => {fun()}





  RightMenusFunction['searchWord'] = (__text__) => {OpenSearch(__text__)}





  RightMenusFunction['bingSearch'] = (__text__) => {window.open(`https://cn.bing.com/search?q=${__text__}`)}





  RightMenusFunction['openTab'] = (__link__) => {window.open(__link__)}





  //RightMenusFunction['copyLink'] = (fun) => {fun()}





  //RightMenusFunction['copyImg'] = (fun) => {fun()}





  RightMenusFunction['googleImg'] = (__link__) => {window.open(`https://www.google.com.hk/searchbyimage?image_url=${__link__}`)}



















  RightMenusFunction['darkMode'] = () => {volantis.dark.toggle()}





  //RightMenusFunction['printMode'] = (fun) => {fun()}





  //RightMenusFunction['readMode'] = (fun) => {fun()}





</script>



<!-- rightmenu要在darkmode之前（ToggleButton） darkmode要在comments之前（volantis.dark.push）-->



<script>
  function loadIssuesJS() {
    
      const sites_api = document.getElementById('sites-api');
      if (sites_api != undefined && typeof SitesJS === 'undefined') {
        volantis.js("/js/plugins/tags/sites.js")
      }
    
    
      const friends_api = document.getElementById('friends-api');
      if (friends_api != undefined && typeof FriendsJS === 'undefined') {
        volantis.js("/js/plugins/tags/friends.js")
      }
    
    
      const contributors_api = document.getElementById('contributors-api');
      if (contributors_api != undefined && typeof ContributorsJS === 'undefined') {
        volantis.js("/js/plugins/tags/contributors.js")
      }
    
  };
  loadIssuesJS()
</script>




  <script defer src="https://cdn.bootcdn.net/ajax/libs/vanilla-lazyload/17.1.0/lazyload.min.js"></script>
<script>
  // https://www.npmjs.com/package/vanilla-lazyload
  // Set the options globally
  // to make LazyLoad self-initialize
  window.lazyLoadOptions = {
    elements_selector: ".lazyload",
    threshold: 0
  };
  // Listen to the initialization event
  // and get the instance of LazyLoad
  window.addEventListener(
    "LazyLoad::Initialized",
    function (event) {
      window.lazyLoadInstance = event.detail.instance;
    },
    false
  );
  document.addEventListener('DOMContentLoaded', function () {
    lazyLoadInstance.update();
  });
</script>




  

<script>
  window.FPConfig = {
	delay: 0,
	ignoreKeywords: ["#"],
	maxRPS: 6,
	hoverDelay: 0
  };
</script>
<script defer src="https://unpkg.com/volantis-static@0.0.1660614606622/libs/flying-pages/flying-pages.min.js"></script>









      <script>
  volantis.layoutHelper("comments",`<div id="giscus_container"></div>`)

  volantis.giscus = {};

  function check_giscus() {
    if (volantis.dark.mode === "dark") {
      volantis.giscus.Theme = 'dark';
    } else {
      volantis.giscus.Theme = 'light';
    }

    return document.getElementById("giscus_container");
  }

  function p_giscus() {
    const HEAD = check_giscus();
    if (!HEAD) return;
    let cfg = Object.assign({"repo":"Snowman12137/blog-comments","repo-id":"R_kgDOLKoZ0Q","category":"Announcements","category-id":"DIC_kwDOLKoZ0c4CcwFX","mapping":"pathname","reactions-enabled":"1","emit-metadata":"0","lang":"zh-CN","theme":{"light":"light","dark":"dark"}},pdata.commentConfig)
    const script = document.createElement('script');
    script.setAttribute('src', 'https://giscus.app/client.js');
    Object.keys(cfg).forEach(k=>{
      if (k != "theme") {
        script.setAttribute('data-'+k, cfg[k]);
      }
    })
    script.setAttribute('data-theme', volantis.giscus.Theme);
    script.setAttribute('crossorigin', "anonymous");
    HEAD.appendChild(script);
  }

  function dark_giscus() {
    const HEAD = check_giscus();
    if (!HEAD) return;

    const message = {
      setConfig: {
        theme: volantis.giscus.Theme
      }
    };
    const giscusIframe = document.querySelector('iframe.giscus-frame');
    giscusIframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
  }
  p_giscus();
  volantis.dark.push(dark_giscus);
</script>

    





<!-- optional -->

  <script>
  async function loadSearchScript() {
    
        
    
    return volantis.js("/js/search/hexo.js");
  }

  function loadSearchService() {
    loadSearchScript();
    document.querySelectorAll(".input.u-search-input").forEach((e) => {
      e.removeEventListener("focus", loadSearchService, false);
    });

    document.querySelectorAll(".u-search-form").forEach((e) => {
      e.addEventListener("submit", (event) => {
        event.preventDefault();
      }, false);
    });
  }

  // 打开并搜索 字符串 s
  function OpenSearch(s) {
    if (typeof SearchService === 'undefined')
      loadSearchScript().then(() => {
        SearchService.setQueryText(s);
        SearchService.search();
      });
    else {
      SearchService.setQueryText(s);
      SearchService.search();
    }
  }

  // 访问含有 ?s=xxx  的链接时打开搜索 // 与搜索引擎 structured data 相关: /scripts/helpers/structured-data/lib/config.js
  if (window.location.search && /^\?s=/g.test(window.location.search)) {
    let queryText = decodeURI(window.location.search)
      .replace(/\ /g, "-")
      .replace(/^\?s=/g, "");
    OpenSearch(queryText);
  }

  // 搜索输入框获取焦点时加载搜索
  document.querySelectorAll(".input.u-search-input").forEach((e) => {
    e.addEventListener("focus", loadSearchService, false);
  });
</script>







  <script>



  function p_highlightjs_copyCode(){
    if (!(document.querySelector(".highlight .code pre") ||
      document.querySelector(".article pre code"))) {
      return;
    }
    VolantisApp.utilCopyCode(".highlight .code pre, .article pre code")
  }
  volantis.requestAnimationFrame(p_highlightjs_copyCode)

</script>








  <script>
  let imgs = ["https://unpkg.com/volantis-static@0.0.1660614606622/media/wallpaper/minimalist/2020/012.webp","https://unpkg.com/volantis-static@0.0.1660614606622/media/wallpaper/minimalist/2020/039.webp","https://unpkg.com/volantis-static@0.0.1660614606622/media/wallpaper/minimalist/2020/002.webp","https://unpkg.com/volantis-static@0.0.1660614606622/media/wallpaper/minimalist/2020/034.webp","https://unpkg.com/volantis-static@0.0.1660614606622/media/wallpaper/minimalist/2020/001.webp","https://unpkg.com/volantis-static@0.0.1660614606622/media/wallpaper/minimalist/2020/019.webp","https://unpkg.com/volantis-static@0.0.1660614606622/media/wallpaper/minimalist/2020/004.webp","https://unpkg.com/volantis-static@0.0.1660614606622/media/wallpaper/minimalist/2020/033.webp","https://unpkg.com/volantis-static@0.0.1660614606622/media/wallpaper/minimalist/2020/016.webp","https://unpkg.com/volantis-static@0.0.1660614606622/media/wallpaper/minimalist/2020/054.webp","https://unpkg.com/volantis-static@0.0.1660614606622/media/wallpaper/minimalist/2020/042.webp","https://unpkg.com/volantis-static@0.0.1660614606622/media/wallpaper/minimalist/2020/038.webp","https://unpkg.com/volantis-static@0.0.1660614606622/media/wallpaper/minimalist/2020/025.webp","https://unpkg.com/volantis-static@0.0.1660614606622/media/wallpaper/minimalist/2020/035.webp","https://unpkg.com/volantis-static@0.0.1660614606622/media/wallpaper/minimalist/2020/006.webp","https://unpkg.com/volantis-static@0.0.1660614606622/media/wallpaper/minimalist/2020/005.webp","https://unpkg.com/volantis-static@0.0.1660614606622/media/wallpaper/minimalist/2020/052.webp","https://unpkg.com/volantis-static@0.0.1660614606622/media/wallpaper/minimalist/2020/046.webp","https://unpkg.com/volantis-static@0.0.1660614606622/media/wallpaper/minimalist/2020/003.webp","https://unpkg.com/volantis-static@0.0.1660614606622/media/wallpaper/minimalist/2020/056.webp","https://unpkg.com/volantis-static@0.0.1660614606622/media/wallpaper/minimalist/2020/051.webp"];
  let index = 0;
  let IntervalParallax = null;

  function parallax(){
    let ParallaxWindow = document.querySelector("#parallax-window");
    
    Parallax.window = ParallaxWindow;
    Parallax.options.fade = 1500;
    Parallax.cache = 1;
    next_parallax();
    Parallax.init();
    if (imgs.length>1) {
      IntervalParallax = setInterval(function () {
        next_parallax();
      }, '10000');
    }
  }

  function next_parallax() {
    if (typeof Parallax == "undefined") {
      return
    }
    
      if (!document.querySelector("#full")&&!document.querySelector("#half")) {
        return
      }
    
    if (imgs.length>=1) {
      Parallax.options.src = imgs[index % imgs.length];
      Parallax.start();
      index++;
      if (Parallax.cache) {
        fetch(imgs[index % imgs.length] +"?t=" + new Date().getTime());
        if (index == imgs.length) {
          Parallax.cache = 0;
        }
      }
    }
  }
  var runningOnBrowser = typeof window !== "undefined";
  var isBot = runningOnBrowser && !("onscroll" in window) || typeof navigator !== "undefined" && /(gle|ing|ro|msn)bot|crawl|spider|yand|duckgo/i.test(navigator.userAgent);
  if (!isBot) {
    volantis.js('/js/plugins/parallax.js').then(()=>{
      parallax()
    })
  }
</script>




  <script>
  function load_swiper() {
    if (!document.querySelectorAll(".swiper-container")[0]) return;
    volantis.css("https://cdn.bootcdn.net/ajax/libs/Swiper/8.3.2/swiper-bundle.min.css");
    volantis.js("https://cdn.bootcdn.net/ajax/libs/Swiper/8.3.2/swiper-bundle.min.js").then(() => {
      p_swiper();
    });
  }

  load_swiper();

  function p_swiper() {
    volantis.swiper = new Swiper('.swiper-container', {
      slidesPerView: 'auto',
      spaceBetween: 8,
      centeredSlides: true,
      loop: true,
      pagination: {
        el: '.swiper-pagination',
        clickable: true,
      },
      navigation: {
        nextEl: '.swiper-button-next',
        prevEl: '.swiper-button-prev',
      },
    });
  }

</script>



  <script>
  volantis.css("https://cdn.bootcdn.net/ajax/libs/pace/1.2.4/themes/blue/pace-theme-minimal.min.css");
  volantis.js("https://cdn.bootcdn.net/ajax/libs/pace/1.2.4/pace.min.js").then(() => {
    Pace.options.restartOnPushState = false;
    // 阅读进度
    volantis.scroll.push(() => {
      Pace.stop();
      Pace.trigger('start');
      Pace.bar.update(volantis.scroll.progress() * 100);
      document.body.className = document.body.className.replaceAll("pace-running ","");
    },"阅读进度")
    // todo：配置项开关, 哪些页面添加阅读进度
  })
</script>




<script>
  function listenSidebarTOC() {
    const navItems = document.querySelectorAll(".toc li");
    if (!navItems.length) return;
    const targets = []
    const sections = Array.from(navItems).forEach((element) => {
      const link = element.querySelector(".toc-link");
      const target = document.getElementById(
        link.getAttribute("href") ? decodeURI(link.getAttribute("href")).replace("#", "") : link.getAttribute("toc-action").split("toc-")[1] // 兼容 hexo-blog-encrypt
      );
      targets.push(target)
      // 解除 a 标签 href 的 锚点定位, a 标签 href 的 锚点定位 会随机启用?? 产生错位???
      if (link.getAttribute("href")) {
        link.setAttribute("toc-action", "toc-" + decodeURI(link.getAttribute("href")).replace("#", ""))
        link.removeAttribute("href")
      }
      // 配置 点击 触发新的锚点定位
      if (target && target.id) {
        link.addEventListener("click", (event) => {
          event.preventDefault();
          // 这里的 addTop 是通过错位使得 toc 自动展开.
          volantis.scroll.to(target, { addTop: 5, observer: true })
          // Anchor id
          history.pushState(null, document.title, "#" + target.id);
        });
      }
      return target;
    });

    function activateNavByIndex(target) {
      if (target.classList.contains("active-current")) return;

      document.querySelectorAll(".toc .active").forEach((element) => {
        element.classList.remove("active", "active-current");
      });
      target.classList.add("active", "active-current");
      let parent = target.parentNode;
      while (!parent.matches(".toc")) {
        if (parent.matches("li")) parent.classList.add("active");
        parent = parent.parentNode;
      }
    }

    // 方案一：
    volantis.activateNavIndex = 0
    activateNavByIndex(navItems[volantis.activateNavIndex])
    if (targets[0]) {
      volantis.scroll.push(() => {
        if (targets[0].getBoundingClientRect().top >= 0) {
          volantis.activateNavIndex = 0
        } else if (targets[targets.length - 1].getBoundingClientRect().top < 0) {
          volantis.activateNavIndex = targets.length - 1
        } else {
          for (let index = 0; index < targets.length; index++) {
            const target0 = targets[index];
            const target1 = targets[(index + 1) % targets.length];
            if (target0.getBoundingClientRect().top < 0 && target1.getBoundingClientRect().top >= 0) {
              volantis.activateNavIndex = index
              break;
            }
          }
        }
        activateNavByIndex(navItems[volantis.activateNavIndex])
      })
    }

    // 方案二：
    // IntersectionObserver 不是完美精确到像素级别 也不是低延时性的
    // function findIndex(entries) {
    //   let index = 0;
    //   let entry = entries[index];
    //   if (entry.boundingClientRect.top > 0) {
    //     index = sections.indexOf(entry.target);
    //     return index === 0 ? 0 : index - 1;
    //   }
    //   for (; index < entries.length; index++) {
    //     if (entries[index].boundingClientRect.top <= 0) {
    //       entry = entries[index];
    //     } else {
    //       return sections.indexOf(entry.target);
    //     }
    //   }
    //   return sections.indexOf(entry.target);
    // }
    // function createIntersectionObserver(marginTop) {
    //   marginTop = Math.floor(marginTop + 10000);
    //   let intersectionObserver = new IntersectionObserver(
    //     (entries, observe) => {
    //       let scrollHeight = document.documentElement.scrollHeight;
    //       if (scrollHeight > marginTop) {
    //         observe.disconnect();
    //         createIntersectionObserver(scrollHeight);
    //         return;
    //       }
    //       let index = findIndex(entries);
    //       activateNavByIndex(navItems[index]);
    //     }, {
    //       rootMargin: marginTop + "px 0px -100% 0px",
    //       threshold: 0,
    //     }
    //   );
    //   sections.forEach((element) => {
    //     element && intersectionObserver.observe(element);
    //   });
    // }
    // createIntersectionObserver(document.documentElement.scrollHeight);
  }

  document.addEventListener("DOMContentLoaded", () => {
    volantis.requestAnimationFrame(listenSidebarTOC)
  });
</script>





<script>
  document.onreadystatechange = function () {
    if (document.readyState == 'complete') {
      // 页面加载完毕 样式加载失败，或是当前网速慢，或是开启了省流模式
      const { saveData, effectiveType } = navigator.connection || navigator.mozConnection || navigator.webkitConnection || {}
      if (getComputedStyle(document.querySelector("#safearea"), null)["display"] == "none" || saveData || /2g/.test(effectiveType)) {
        document.querySelectorAll(".reveal").forEach(function (e) {
          e.style["opacity"] = "1";
        });
        document.querySelector("#safearea").style["display"] = "block";
      }
    }
  }
</script>


  <script type="application/ld+json">[{"@context":"http://schema.org","@type":"Organization","name":"小吴の博客","url":"https://snowman12137.github.io/","logo":{"@type":"ImageObject","url":"https://unpkg.com/volantis-static@0.0.1660614606622/media/org.volantis/blog/favicon/android-chrome-192x192.png","width":192,"height":192}},{"@context":"http://schema.org","@type":"Person","name":"Snowman","image":{"@type":"ImageObject","url":"https://unpkg.com/volantis-static@0.0.1660614606622/media/org.volantis/blog/favicon/android-chrome-192x192.png"},"url":"https://snowman12137.github.io/","sameAs":["https://github.com/volantis-x"],"description":"我的博客"},{"@context":"http://schema.org","@type":"BreadcrumbList","itemListElement":[{"@type":"ListItem","position":1,"item":{"@id":"https://snowman12137.github.io/","name":"小吴の博客"}},{"@type":"ListItem","position":2,"item":{"@id":"https://snowman12137.github.io/categories/生物计算机科学/","name":"生物计算机科学"}},{"@type":"ListItem","position":3,"item":{"@id":"https://snowman12137.github.io/2024/03/25/生信分析-通过共表达网络分析（WGCNA）/","name":"通过共表达网络分析（WGCNA）"}}]},{"@context":"http://schema.org","@type":"WebSite","name":"小吴の博客","url":"https://snowman12137.github.io/","keywords":null,"description":"我的博客","author":{"@type":"Person","name":"Snowman","image":{"@type":"ImageObject","url":"https://unpkg.com/volantis-static@0.0.1660614606622/media/org.volantis/blog/favicon/android-chrome-192x192.png"},"url":"https://snowman12137.github.io/","description":"我的博客"},"publisher":{"@type":"Organization","name":"小吴の博客","url":"https://snowman12137.github.io/","logo":{"@type":"ImageObject","url":"https://unpkg.com/volantis-static@0.0.1660614606622/media/org.volantis/blog/favicon/android-chrome-192x192.png","width":192,"height":192}},"potentialAction":{"@type":"SearchAction","name":"Site Search","target":{"@type":"EntryPoint","urlTemplate":"https://snowman12137.github.io?s={search_term_string}"},"query-input":"required name=search_term_string"}},{"@context":"http://schema.org","@type":"BlogPosting","headline":"通过共表达网络分析（WGCNA）","description":"我的博客","inLanguage":["zh-CN","en","zh-TW","default"],"mainEntityOfPage":{"@type":"WebPage","@id":"https://snowman12137.github.io/2024/03/25/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90-%E9%80%9A%E8%BF%87%E5%85%B1%E8%A1%A8%E8%BE%BE%E7%BD%91%E7%BB%9C%E5%88%86%E6%9E%90%EF%BC%88WGCNA%EF%BC%89/"},"author":{"@type":"Person","name":"Snowman","image":{"@type":"ImageObject","url":"https://unpkg.com/volantis-static@0.0.1660614606622/media/org.volantis/blog/favicon/android-chrome-192x192.png"},"url":"https://snowman12137.github.io/"},"publisher":{"@type":"Organization","name":"小吴の博客","logo":{"@type":"ImageObject","url":"https://unpkg.com/volantis-static@0.0.1660614606622/media/org.volantis/blog/favicon/android-chrome-192x192.png","width":192,"height":192}},"url":"https://snowman12137.github.io/2024/03/25/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90-%E9%80%9A%E8%BF%87%E5%85%B1%E8%A1%A8%E8%BE%BE%E7%BD%91%E7%BB%9C%E5%88%86%E6%9E%90%EF%BC%88WGCNA%EF%BC%89/","wordCount":0,"datePublished":"2024-03-24T16:00:00.000Z","dateModified":"2024-03-25T14:51:16.314Z","articleSection":"生物计算机科学","keywords":"WGCAN网络","image":{"@type":"ImageObject","url":"https://unpkg.com/volantis-static@0.0.1660614606622/media/org.volantis/blog/favicon/android-chrome-192x192.png","width":192,"height":192}}]</script>



    </div>
    <!-- import body_end begin-->
    <!-- import body_end end-->
    <!-- Custom Files bodyEnd begin-->
    
    <!-- Custom Files bodyEnd end-->
  </body>
</html>
<script type="text/javascript" src="/js/love.js"></script>
<script type="text/javascript" src="\js\FunnyTitle.js"></script>